{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "13d4ac73",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T16:04:09.940994Z",
     "start_time": "2023-06-05T16:03:58.298158Z"
    }
   },
   "outputs": [],
   "source": [
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "from Bio.Seq import reverse_complement, Seq\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "import config_readwrite as crw\n",
    "\n",
    "import glob\n",
    "import gzip\n",
    "import numpy as np\n",
    "import os, sys\n",
    "import pandas as pd\n",
    "from random import shuffle\n",
    "import subprocess as sp\n",
    "import time"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3682616c",
   "metadata": {},
   "source": [
    "# find coordinates of kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e889df9d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T16:04:10.123611Z",
     "start_time": "2023-06-05T16:04:09.947824Z"
    }
   },
   "outputs": [],
   "source": [
    "CL, MER, NMUTS, BUILD, FO_ONLY = \"common\", \"15\", \"2\", \"hg38\", True\n",
    "config, cfn = crw.read(os.path.join(os.path.dirname(os.getcwd()), \"config.ini\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f5b23b8c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T16:04:17.779591Z",
     "start_time": "2023-06-05T16:04:17.761726Z"
    }
   },
   "outputs": [],
   "source": [
    "# read\n",
    "\n",
    "# concatenated nullomers\n",
    "CONCAT= config[f\"{CL}.{MER}mer.{NMUTS}mut\"][\"concat-13\"]\n",
    "\n",
    "# randomly sampled nullomers for single inserts\n",
    "NULLS = config[f\"seed\"]['pruned_tsv']\n",
    "\n",
    "# bed file of common MPRA coordinates\n",
    "BED = config[CL][f\"bed_{BUILD}\"]\n",
    "\n",
    "# extended 4096bp fasta file for common MPRA coordinates\n",
    "FA_EXT_BED = config[CL][f\"fa_extended_{BUILD}\"]\n",
    "\n",
    "# extended 4096 bed file for common MPRA coordinates\n",
    "EXTENDED = config[CL][f\"bed_extended_{BUILD}\"]\n",
    "\n",
    "# fast file for common MPRA coordinates\n",
    "FA = config[CL][f\"fasta_{BUILD}\"]\n",
    "\n",
    "# scaffolds for inserting single nullomers. \n",
    "# Chosen because active (top 97.5% Agarwal) | inactive (IQR Agarwal) across 3 cell lines.\n",
    "# Randomly sample 150 elements from these two categories. \n",
    "# Predict scaffold activity with Sei\n",
    "# Pick scaffolds w/ predicted activity in Sei. (Are these sequences more active than not predicted?)\n",
    "JOINT_LIBRARY=config[\"MPRA_AGARWAL\"][\"joint_library\"]\n",
    "SCAFFOLDS = config[\"scaffold\"][\"sampled\"]\n",
    "\n",
    "ROOT = \"/wynton/home/ahituv/fongsl/nullomers/results/lock/library_design\"\n",
    "\n",
    "#write\n",
    "\n",
    "LIBRARY_FILE = os.path.join(ROOT, f'{MER}mer.fo.pam.scaffold'+\n",
    "              f\".ext200.library.tsv\")\n",
    "\n",
    "LIBRARY_CTRL = os.path.splitext(LIBRARY_FILE)[0] + \".ctrl.tsv\"\n",
    "\n",
    "LIBRARY_TWIST = os.path.splitext(LIBRARY_FILE)[0] + \".TWIST.tsv\"\n",
    "\n",
    "out200 = os.path.join(ROOT, f'{MER}mer.fo.pam.scaffold'+\n",
    "              f\".ext200.fa\")\n",
    "out4096 = os.path.join(ROOT, f'{MER}mer.fo.pam.scaffold'+\n",
    "              f\".ext4096.fa\")\n",
    "\n",
    "# write library file to config\n",
    "config[f\"{CL}.{MER}mer.{NMUTS}mut\"][\"200mers\"] = out200\n",
    "config[f\"{CL}.{MER}mer.{NMUTS}mut\"][\"4096mers\"] = out4096\n",
    "config[f\"{CL}.{MER}mer.{NMUTS}mut\"][\"library\"] = LIBRARY_FILE\n",
    "config[f\"{CL}.{MER}mer.{NMUTS}mut\"][\"library_ctrl\"] = LIBRARY_CTRL\n",
    "config[f\"{CL}.{MER}mer.{NMUTS}mut\"][\"library_twist\"] = LIBRARY_TWIST\n",
    "crw.write(config, cfn)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32574b56",
   "metadata": {},
   "source": [
    "# functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "34f1b8c0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T16:04:21.004435Z",
     "start_time": "2023-06-05T16:04:20.966789Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def getNullSeqs(file):\n",
    "    \"\"\"\n",
    "    read nullomer/concat sequence .tsv\n",
    "\n",
    "    input\n",
    "        file(str) - .txt or .tsv file to nullomer sequences\n",
    "\n",
    "    method\n",
    "        1. make empty nulls, kmers list to collect data\n",
    "        2. iterate through file lines\n",
    "        3. concatmer - append null, kmer sequences\n",
    "        4. single nullomers - append just null sequences. \n",
    "    return list of nullomers, matching kmers (for concatemers only)\n",
    "    \"\"\"\n",
    "\n",
    "    # read file\n",
    "    df = pd.read_csv(file, sep='\\t')\n",
    "\n",
    "    # reformat the concat library\n",
    "    if \"nullConcat\" in list(df):\n",
    "        if \"id\" not in list(df):\n",
    "            # add id number to concats\n",
    "            df['id'] = \"concat-\" + df.index.map(str)\n",
    "\n",
    "            # save the id in the file\n",
    "            df.to_csv(file, sep='\\t', index=False)\n",
    "\n",
    "        df = df.rename(columns={\"nullConcat\": \"seq\"})\n",
    "\n",
    "    return df[[\"id\", \"seq\"]]\n",
    "\n",
    "\n",
    "def getScaffolds(scaffold_file):\n",
    "\n",
    "    df = pd.read_csv(scaffold_file, sep='\\t')\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def getStrand(scaffold_df):\n",
    "    \"\"\"\n",
    "    return strand of the scaffold sequences\n",
    "\n",
    "    input\n",
    "        scaffold_seq_list (list)- list of scaffold coordinates\n",
    "        lib (pd dataframe) - vikram's 60k common library as pandas dataframe\n",
    "\n",
    "    method\n",
    "        1. iterate through seq coordinate list\n",
    "        1.1 get sequence from coordinate. \n",
    "            Because library was in hg38, but scaffolds are in hs1 (i lifted), need to harmonize on sequence using 150 bases\n",
    "\n",
    "        2. get strand for sequence\n",
    "            2.1 check for strand conflicts (where sequence matches both to a + and - strand)\n",
    "        3. if strand is negative, get reverse complement (fwd), else return coor and fwd seq\n",
    "\n",
    "    return\n",
    "        scaffold_coor_fa (list) - tuple of coordinates and forward strand sequences\n",
    "\n",
    "    \"\"\"\n",
    "    plus = scaffold_df.loc[scaffold_df.strand == \"+\"].copy()\n",
    "    minus = scaffold_df.loc[scaffold_df.strand == \"-\"].copy()\n",
    "\n",
    "    plus[\"fwd.nt\"] = plus[\"200nt\"]\n",
    "    plus[\"rev.nt\"] = plus[\"200nt\"].apply(\n",
    "        lambda x: \"\".join(reverse_complement(Seq(x))))\n",
    "\n",
    "    minus[\"rev.nt\"] = minus[\"200nt\"]\n",
    "    minus[\"fwd.nt\"] = minus[\"200nt\"].apply(\n",
    "        lambda x: \"\".join(reverse_complement(Seq(x))))\n",
    "\n",
    "    return pd.concat([plus, minus])\n",
    "\n",
    "\n",
    "def faToDict(fa_file):\n",
    "    \"\"\"\n",
    "    turn all the fa coordinates into a dictionary. \n",
    "    \"\"\"\n",
    "\n",
    "    fa_dict = {}\n",
    "    with open(fa_file, \"r\") as fasta_reader:\n",
    "        for values in SimpleFastaParser(fasta_reader):\n",
    "\n",
    "            name, seq = values\n",
    "            fa_dict[name] = seq\n",
    "\n",
    "    return fa_dict\n",
    "\n",
    "\n",
    "def dissectFaId(line):\n",
    "    \"\"\"\n",
    "    dissect mutated fa id line for kmer, nullomer information, and get the coordinate file\n",
    "    \"\"\"\n",
    "    # process the line\n",
    "    kmer, chr_, coor, strand, id_ = line.strip('\\n').strip(\">\").split(\".\")\n",
    "\n",
    "    # get key and value sequence to look up\n",
    "    key, value_seq = kmer[:5], kmer[5:]\n",
    "\n",
    "    bed_coor = chr_ + \":\" + coor\n",
    "\n",
    "    # TACCCATTCGAGTC.chr1.100333924-100334124.fwd\n",
    "    region_str = \".\".join([kmer, bed_coor, strand, id_])\n",
    "\n",
    "    return kmer, key, value_seq, bed_coor, region_str\n",
    "\n",
    "\n",
    "def seqInsert(seq, insert_start, insert_seq):\n",
    "    \"\"\"\n",
    "    insert sequence fragment (insert_seq) at position (insert_start) within full sequence (seq)\n",
    "\n",
    "    return inserted sequence. \n",
    "    \"\"\"\n",
    "\n",
    "    insert_size = len(insert_seq)\n",
    "\n",
    "    insert_end = insert_start + insert_size  # find center end\n",
    "\n",
    "    return seq[:int(insert_start)] + insert_seq + seq[int(insert_end):]\n",
    "\n",
    "\n",
    "def MPRAInsertNull(null_sequence, scaffold_sequence, scaffold_activity):\n",
    "    \"\"\"\n",
    "    return positions where the nullomer is in fa sequence\n",
    "    \"\"\"\n",
    "    if scaffold_activity == \"active\":\n",
    "        insert_start = 200 - len(null_sequence)  # place nullomer at the 3' end\n",
    "    else:\n",
    "        insert_start = (len(scaffold_sequence) - len(null_sequence))/2\n",
    "\n",
    "    # insert the nullomer sequence into the sequence\n",
    "    null_insert = seqInsert(scaffold_sequence.lower(),\n",
    "                            insert_start, null_sequence.upper())\n",
    "\n",
    "    return null_insert\n",
    "\n",
    "\n",
    "def mapBed(bedfile, expandedfile):\n",
    "\n",
    "    bed = pd.read_csv(bedfile, sep='\\t')\n",
    "    bed.columns = names = [\"#chr\", \"start\", \"end\", \"id\",  \"strand\"]\n",
    "    bed[\"coor\"] = bed[\"#chr\"] + \":\" + \\\n",
    "        bed[\"start\"].map(str) + \"-\" + bed[\"end\"].map(str)\n",
    "\n",
    "    ext = pd.read_csv(expandedfile, sep='\\t', header=None,\n",
    "                      names=[\"#chr_e\", \"start_e\", \"end_e\", \"id\"])\n",
    "    ext[\"coor_e\"] = ext[\"#chr_e\"] + \":\" + \\\n",
    "        ext[\"start_e\"].map(str) + \"-\" + ext[\"end_e\"].map(str)\n",
    "\n",
    "    joint = pd.merge(ext[[\"id\", \"coor_e\"]], bed[[\"id\", \"coor\"]])\n",
    "\n",
    "    return joint\n",
    "\n",
    "\n",
    "def insert4096(null_seq, joint_bedex, fa_ex_dict, coor):\n",
    "\n",
    "    # extend the coordinates to 4096\n",
    "    extended_coor = joint_bedex.loc[joint_bedex[\"coor\"]\n",
    "                                    == coor, \"coor_e\"].iloc[0]\n",
    "\n",
    "    # get the fasta file linked to the extended coordinates\n",
    "    extended_seq = fa_ex_dict[extended_coor]\n",
    "\n",
    "    # get the start to insert  # figure out where to insert the sequence\n",
    "    insert_start = len(extended_seq)/2 - 100\n",
    "\n",
    "    # insert the 200bp MPRA tile w/ nullomer variants into sequence\n",
    "    null4096 = seqInsert(extended_seq.lower(), insert_start, null_seq.upper())\n",
    "\n",
    "    return null4096, extended_seq\n",
    "\n",
    "\n",
    "def prettifySeq(original, mut):\n",
    "    \"\"\"\n",
    "    prettify sequence. \n",
    "        All matching bases will be written as \".\"\n",
    "        All non matching bases will be written w mutated base identity. \n",
    "    \"\"\"\n",
    "    prettyseq = \"\"\n",
    "    for o, m in zip(original, mut):\n",
    "        if o != m:\n",
    "            prettyseq += m\n",
    "        else:\n",
    "            prettyseq += \".\"\n",
    "\n",
    "    return prettyseq\n",
    "\n",
    "\n",
    "def visualInspection(kmer, null, null_ins200, inserts4096, original_seq):\n",
    "    \"\"\"\n",
    "    print a bunch of information related to how this script is \n",
    "    inserting kmers into 200bp regions, \n",
    "    inserting those 200bp regions into 4096mers\n",
    "    \"\"\"\n",
    "\n",
    "    print(\"\\nkmer\\n\", kmer)\n",
    "\n",
    "    print(\"\\nthe nullomer sequence\\n\",\n",
    "          null, \"\\n\\ninserted\\n\", null_ins200)\n",
    "\n",
    "    print(\"\\n4096 w/ null sequence\\n\", inserts4096)\n",
    "\n",
    "    print(\"\\n pretty 4096\", prettifySeq(\n",
    "        original_seq.upper(), list(inserts4096.values())[0].upper()))\n",
    "\n",
    "\n",
    "def addAdaptor(seq):\n",
    "    adaptor5, adaptor3 = \"AGGACCGGATCAACT\", \"CATTGCGTGAACCGA\"\n",
    "    adapted = adaptor5 + seq + adaptor3\n",
    "    return adapted.upper()\n",
    "\n",
    "\n",
    "def writeRow(file, row):\n",
    "    \"\"\"\n",
    "    append line to file\n",
    "\n",
    "    input\n",
    "        file (str) - file to write to\n",
    "        row (str) - row to write\n",
    "    method\n",
    "        1. open file\n",
    "        2. write\n",
    "        3. close\n",
    "    \"\"\"\n",
    "    # 1\n",
    "    with open(file, \"a\") as writer:\n",
    "        # 2\n",
    "        writer.write(row)\n",
    "    # 3\n",
    "    writer.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b346b4b",
   "metadata": {},
   "source": [
    "# scaffold selection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0ee7f1e",
   "metadata": {},
   "source": [
    "Scaffold selection? \n",
    "\n",
    "See: ./nullomers/bin-lock/Analysis/mpra_corr_scaffold.ipynb\n",
    "    ./nullomers/bin-lock/Sei_seqeunce_class-scaffold-minmax-selection.ipynb\n",
    "\n",
    "    Basically, 60K sequences commonly active|inactive \n",
    "    across 3 cell lines from Agarwal 2023 were evaluated \n",
    "    for MPRA activity (log(rna/dna)). I quantified the mean \n",
    "    activity across these three cell lines, then randomly sampled from  \n",
    "    the top 97.5%, IQR, and bottom 2.5% active sequences \n",
    "    \n",
    "    Then, I predicted the sequence class from each of those \n",
    "    common sequences. Of the sequences with interpretable predictions in Sei\n",
    "    (i.e. sei predicted active sequences as promoter/ enhancer)\n",
    "    I sampled 50 of the active and 50 of the IQR (marginal activity)\n",
    "    to use as scaffolds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fd5277c",
   "metadata": {},
   "source": [
    "# Main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7c00ed40",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T16:04:24.182596Z",
     "start_time": "2023-06-05T16:04:23.252815Z"
    }
   },
   "outputs": [],
   "source": [
    "def mapBedGenomeBuild(from_, to_, config, cl):\n",
    "    \n",
    "    \n",
    "    for i, build in enumerate([from_, to_]):\n",
    "        BED = config[cl][f\"bed_{build}\"]\n",
    "        cols =[f\"chr.{build}\", f\"start.{build}\", f\"end.{build}\", \"id\", f\"str.{build}\"]\n",
    "    \n",
    "        bed = pd.read_csv(BED, sep='\\t', header=None)\n",
    "        bed.columns = cols\n",
    "        bed[f\"{build}.coor\"] = bed[f\"chr.{build}\"]+\":\"+ bed[f\"start.{build}\"].map(str) +\"-\"+ bed[f\"end.{build}\"].map(str)\n",
    "        \n",
    "        if i ==0:\n",
    "            df = bed.copy()\n",
    "        else:\n",
    "            merged = pd.merge(df, bed, how='left')\n",
    "            \n",
    "    return merged.iloc[1:]\n",
    "    \n",
    "\n",
    "build_map = mapBedGenomeBuild(\"hg38\", 'hs1', config, CL)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afc8f7db",
   "metadata": {},
   "source": [
    "## set up scaffold FA resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "33e57a43",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T16:04:30.238583Z",
     "start_time": "2023-06-05T16:04:26.106212Z"
    }
   },
   "outputs": [],
   "source": [
    "fa_dict = faToDict(FA) # in hs1\n",
    "fa_ex_dict = faToDict(FA_EXT_BED) # in hs1\n",
    "\n",
    "joint_bedex = mapBed(BED, EXTENDED) # in hg38\n",
    "\n",
    "lib = pd.read_csv(JOINT_LIBRARY, header=1) # in hg38"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d97e5045",
   "metadata": {},
   "source": [
    "## get scaffolds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7d66eeb6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T16:04:49.547455Z",
     "start_time": "2023-06-05T16:04:49.518600Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "strand\n",
      "+    399\n",
      "-      1\n",
      "Name: name, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "scaffolds_ = getScaffolds(SCAFFOLDS)\n",
    "\n",
    "print(scaffolds_.groupby(\"strand\")[\"name\"].count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "12fe3a2e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T16:04:50.482073Z",
     "start_time": "2023-06-05T16:04:50.464175Z"
    }
   },
   "outputs": [],
   "source": [
    "# make sure all the scaffold sequences are in the forward direction\n",
    "scaffolds = getStrand(scaffolds_)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abf7cabd",
   "metadata": {},
   "source": [
    "## insert nulls into scaffold and write\n",
    "    - takes 12 minutes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "41e6eb9a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T16:04:53.972972Z",
     "start_time": "2023-06-05T16:04:52.655802Z"
    }
   },
   "outputs": [],
   "source": [
    "if os.path.exists(LIBRARY_FILE) is False:\n",
    "\n",
    "    out_tsv_dict = {}  # collect data for tsv\n",
    "\n",
    "    # do single nullomers first\n",
    "    nulls = getNullSeqs(NULLS)\n",
    "    concatnulls = getNullSeqs(CONCAT)\n",
    "    i = 0\n",
    "\n",
    "    datas = [\n",
    "             (\"single-null\", nulls, scaffolds),\n",
    "             (\"concat-null\", concatnulls, \n",
    "              scaffolds.loc[scaffolds[\"name\"] ==\"WTC11_seq8625_F\"].copy()),\n",
    "             ]\n",
    "\n",
    "    # per dataset\n",
    "    for datatype, seq_list, scaffold_list in datas:\n",
    "        print(datatype, len(seq_list), len(scaffold_list))\n",
    "\n",
    "        # per nullomer\n",
    "        for null_row in seq_list.iterrows():\n",
    "            null_id, null_seq = null_row[1]\n",
    "            print(i)\n",
    "            # per scaffold\n",
    "            for scaf_row in scaffold_list[[\"name\", \"fwd.nt\", \"scaffold_activity\", \"hg38.coor\"]].iterrows():\n",
    "                scaf_id, scaf_seq, scaf_act, scaf_coor = scaf_row[1]\n",
    "\n",
    "                # get hg38, hs1 coordinates (in case)\n",
    "                coor_hg38, coor_hs1 = list(build_map.loc[build_map[\"id\"] == scaf_id, \n",
    "                                                         [\"hg38.coor\", \"hs1.coor\"]].iloc[0])\n",
    "\n",
    "\n",
    "                # get fa sequence for the coordinate, make id from coordinate and the sequence index\n",
    "                region_id  = null_id + \"|\" + scaf_id + \"|\" + scaf_act\n",
    "                region_fa_id = scaf_coor + \"|\" + region_id\n",
    "                #print(region_fa_id)\n",
    "\n",
    "                # insert nulls, kmers into center of 200bp MPRA sequences.\n",
    "                seq_ins200 = MPRAInsertNull(\n",
    "                    null_seq, scaf_seq, scaf_act)  # insert null into scaffold seq\n",
    "\n",
    "                # insert 200bp null sequence into the 4096 sequence\n",
    "                seq_ins4096, scaf_seq_ext = insert4096(seq_ins200, joint_bedex, fa_ex_dict, scaf_coor)\n",
    "\n",
    "                if i == 0:  # if this is the first sequence\n",
    "\n",
    "                    # write the original sequences\n",
    "                    writeRow(out200, f\">{scaf_coor}_endog_\\n{scaf_seq}\\n\")\n",
    "                    writeRow(out4096, f\">{scaf_coor}_endog_\\n{scaf_seq_ext}\\n\")\n",
    "\n",
    "                    out_tsv_dict[f\"endog-{scaf_id}\"] = [f\"endog|{scaf_id}|{scaf_act}\", \"endog\", \n",
    "                                                  None, scaf_id, scaf_coor, \"+\", scaf_act,\n",
    "                                                  None, scaf_seq, \n",
    "                                                  addAdaptor(scaf_seq)\n",
    "                                                 ]\n",
    "\n",
    "                # write 200mers\n",
    "                writeRow(out200, f\">{region_fa_id}\\n{seq_ins200.upper()}\\n\")\n",
    "\n",
    "                # write 4096 nullomers\n",
    "                writeRow(out4096, f\">{region_fa_id}\\n{seq_ins4096.upper()}\\n\")\n",
    "\n",
    "                out_tsv_dict[region_fa_id] = [region_id, f\"{datatype}\", \n",
    "                                              null_id, scaf_id, scaf_coor, \"+\", scaf_act, \n",
    "                                              null_seq.upper(), seq_ins200.upper(), \n",
    "                                              addAdaptor(seq_ins200)\n",
    "                                             ]\n",
    "            i +=1\n",
    "\n",
    "\n",
    "    a = [np.array(i) for i in out_tsv_dict.values()]\n",
    "\n",
    "    df = pd.DataFrame(a)\n",
    "    before = df.shape[0]\n",
    "    df= df.drop_duplicates()\n",
    "    print(before - df.shape[0])\n",
    "\n",
    "    df.columns=[\"null_id|scaffold_id|activity\",\"exp.id\", \"null.id\",\"scaffold.id\", \"hg38.scaffold.coor\", \n",
    "                \"hg38.str\", \"scaffold.activity\", \"seq_frag\", \"200bp_insert\", \"200bp_insert+adaptors\"]\n",
    "    df[\"gc_frac\"] = df[\"200bp_insert\"].apply(lambda x:gc_fraction(x))\n",
    "    df.to_csv(LIBRARY_FILE, sep='\\t', index=False)\n",
    "\n",
    "else:\n",
    "    df = pd.read_csv(LIBRARY_FILE, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b889c63a",
   "metadata": {},
   "source": [
    "# add control sequences. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0d791d2",
   "metadata": {},
   "source": [
    "## functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "80eccdca",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T16:09:46.976423Z",
     "start_time": "2023-06-05T16:09:46.954052Z"
    }
   },
   "outputs": [],
   "source": [
    "def dinucScramble(seq, verbose):\n",
    "\n",
    "    n=2 # pick size 2\n",
    "    \n",
    "    splits = [seq[i:i+n] for i in range(0, len(seq), n)]\n",
    "    shuf_splits = splits[::]\n",
    "    \n",
    "    shuffle(shuf_splits)\n",
    "    \n",
    "    if verbose is True:\n",
    "        print(np.unique(splits, return_counts=True))\n",
    "        print(np.unique(shuf_splits, return_counts=True))\n",
    "    \n",
    "\n",
    "    return \"\".join(shuf_splits)\n",
    "\n",
    "def checkLength(ctrl):\n",
    "    ctrl[\"len\"]= ctrl[\"200bp_insert+adaptors\"].apply(lambda x: len(x))\n",
    "\n",
    "    keep = ctrl.loc[ctrl[\"len\"]==230].copy()\n",
    "    trim = ctrl.loc[ctrl[\"len\"]==260].copy() # double adaptor!\n",
    "    \n",
    "    # trim double adaptor to single adaptor\n",
    "    trim[\"200bp_insert+adaptors\"] = trim[\"200bp_insert+adaptors\"].apply(lambda x: x[15:-15])\n",
    "    trim[\"len\"]= trim[\"200bp_insert+adaptors\"].apply(lambda x: len(x))\n",
    "\n",
    "    ctrl = pd.concat([keep, trim])\n",
    " \n",
    "    return ctrl\n",
    "\n",
    "def getAgarwalCtrls(lib):\n",
    "\n",
    "# get control\n",
    "    ctrl = lib.loc[\n",
    "                    (lib[\"category\"].str.contains(\"negative\")) |\n",
    "                    (lib[\"category\"].str.contains(\"positive\"))\n",
    "                     ].drop_duplicates()\n",
    "\n",
    "    ctrl = ctrl.loc[ctrl[\"category\"]!=\"negative, shuffled\"]\n",
    "    ctrl = ctrl[['name',\n",
    "                 'category',\n",
    "                 'chr.hg38',\n",
    "                 'start.hg38',\n",
    "                 'stop.hg38',\n",
    "                 'str.hg38',\n",
    "                 \"230nt sequence (15nt 5' adaptor - 200nt element - 15nt 3' adaptor)\"]]\n",
    "\n",
    "    # fillna\n",
    "    ctrl = ctrl.fillna(-1)\n",
    "\n",
    "    # build the coordinate field\n",
    "    ctrl[\"hg38.scaffold.coor\"] = \"chr\" + ctrl[\"chr.hg38\"].map(str) +\\\n",
    "                                \":\" + ctrl[\"start.hg38\"].map(int).map(str) +\\\n",
    "                                \"-\" + ctrl[\"stop.hg38\"].map(int).map(str)\n",
    "\n",
    "    # rename columns\n",
    "    ctrl = ctrl.rename(columns={\"name\":\"scaffold.id\", \n",
    "                                \"category\":\"exp.id\",\n",
    "                                \"str.hg38\": \"hg38.str\", \n",
    "                                \"230nt sequence (15nt 5' adaptor - 200nt element - 15nt 3' adaptor)\":'200bp_insert+adaptors'\n",
    "                        })\n",
    "\n",
    "    # trim adaptors\n",
    "    ctrl[\"200bp_insert\"]  = ctrl[\"200bp_insert+adaptors\"].apply(lambda x: x[15:-15])\n",
    "    ctrl[\"gc_frac\"]= ctrl[\"200bp_insert\"].apply(lambda x: gc_fraction(x))\n",
    "\n",
    "    # add in empty fields. \n",
    "    ctrl['null.id'] = None\n",
    "    ctrl[\"null_id|scaffold_id|activity\"] = ctrl['null.id'].map(str) + \"|\" + ctrl['scaffold.id'] + \"|\" + ctrl['scaffold.activity']\n",
    "    ctrl['scaffold.activity'] = ctrl[\"exp.id\"].apply(lambda x: \"inactive\" if \"negative\" in x else \"active\")\n",
    "    ctrl[\"seq_frag\"]=None\n",
    "\n",
    "    return ctrl\n",
    "\n",
    "def getEndogSeqNScramble(df, n_scramble):\n",
    "\n",
    "    endog_dict = {}\n",
    "\n",
    "    # per scramble\n",
    "    for i in np.arange(n_scramble):\n",
    "\n",
    "        # make copy of endog dataframe rows\n",
    "        endog_seqs = df.loc[df[\"exp.id\"] == \"endog\"].copy()\n",
    "\n",
    "        # change a bunch of vals\n",
    "        # id - update w/ scramble iter\n",
    "        endog_seqs['exp.id']=\"endog-scramble\"\n",
    "        endog_seqs['null.id'] = f\"endog-scramble-{i}\"\n",
    "        endog_seqs[\"null_id|scaffold_id|activity\"] = f\"endog-scramble-{i}\" + \\\n",
    "            \"|\" + endog_seqs[\"scaffold.id\"]\n",
    "\n",
    "        # scramble!\n",
    "        endog_seqs[\"200bp_insert\"] = endog_seqs[\"200bp_insert\"].apply(\n",
    "            lambda x: dinucScramble(x, False))\n",
    "\n",
    "        # add adaptors to scramble!\n",
    "        endog_seqs[\"200bp_insert+adaptors\"] = endog_seqs[\"200bp_insert\"].apply(\n",
    "            lambda x: addAdaptor(x))\n",
    "\n",
    "        # quantify GC fraction of scramble! (Should be the same!)\n",
    "        endog_seqs[\"gc_frac\"] = endog_seqs[\"200bp_insert\"].apply(\n",
    "            lambda x: gc_fraction(x))\n",
    "\n",
    "        # add scramble to dictionary\n",
    "        endog_dict[i] = endog_seqs.copy()\n",
    "\n",
    "    endog_seqs = pd.concat(endog_dict.values())\n",
    "\n",
    "    return endog_seqs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71408f4a",
   "metadata": {},
   "source": [
    "## get positive controls from Agarwal MPRA\n",
    "## scramble scaffold sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0deb4a86",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T19:09:29.185971Z",
     "start_time": "2023-06-05T19:09:27.066147Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp.id</th>\n",
       "      <th>null.id</th>\n",
       "      <th>scaffold.id</th>\n",
       "      <th>null_id+scaffold_id</th>\n",
       "      <th>hg38.scaffold.coor</th>\n",
       "      <th>hg38.str</th>\n",
       "      <th>gc_frac</th>\n",
       "      <th>len</th>\n",
       "      <th>scaffold.activity</th>\n",
       "      <th>200bp_insert+adaptors</th>\n",
       "      <th>insert_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>concat-null</td>\n",
       "      <td>concat-0</td>\n",
       "      <td>WTC11_seq8625_F</td>\n",
       "      <td>concat-0|WTC11_seq8625_F|inactive</td>\n",
       "      <td>chr10:37754790-37754990</td>\n",
       "      <td>+</td>\n",
       "      <td>0.570</td>\n",
       "      <td>230</td>\n",
       "      <td>inactive</td>\n",
       "      <td>AGGACCGGATCAACTGATATTCGCTCAACCGGAACGAATGTCGACG...</td>\n",
       "      <td>TATTCGCTCAACCGGAACGAATGTCGACGGTAACGTCGAGTGCGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>concat-null</td>\n",
       "      <td>concat-1</td>\n",
       "      <td>WTC11_seq8625_F</td>\n",
       "      <td>concat-1|WTC11_seq8625_F|inactive</td>\n",
       "      <td>chr10:37754790-37754990</td>\n",
       "      <td>+</td>\n",
       "      <td>0.560</td>\n",
       "      <td>230</td>\n",
       "      <td>inactive</td>\n",
       "      <td>AGGACCGGATCAACTGATTCGAACGCAAGCGGTAATTACGTCGACG...</td>\n",
       "      <td>TTCGAACGCAAGCGGTAATTACGTCGACGGGTTGCGAGTAGTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>concat-null</td>\n",
       "      <td>concat-2</td>\n",
       "      <td>WTC11_seq8625_F</td>\n",
       "      <td>concat-2|WTC11_seq8625_F|inactive</td>\n",
       "      <td>chr10:37754790-37754990</td>\n",
       "      <td>+</td>\n",
       "      <td>0.580</td>\n",
       "      <td>230</td>\n",
       "      <td>inactive</td>\n",
       "      <td>AGGACCGGATCAACTGACCATCGATCGAACGGCCGGTATTGGATCG...</td>\n",
       "      <td>CCATCGATCGAACGGCCGGTATTGGATCGGCTAGATCGACGGAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>concat-null</td>\n",
       "      <td>concat-3</td>\n",
       "      <td>WTC11_seq8625_F</td>\n",
       "      <td>concat-3|WTC11_seq8625_F|inactive</td>\n",
       "      <td>chr10:37754790-37754990</td>\n",
       "      <td>+</td>\n",
       "      <td>0.585</td>\n",
       "      <td>230</td>\n",
       "      <td>inactive</td>\n",
       "      <td>AGGACCGGATCAACTGAACCGCTATATTGCGGTTATTGCGCGAGCG...</td>\n",
       "      <td>ACCGCTATATTGCGGTTATTGCGCGAGCGGTATCGACGAAGGCGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>concat-null</td>\n",
       "      <td>concat-4</td>\n",
       "      <td>WTC11_seq8625_F</td>\n",
       "      <td>concat-4|WTC11_seq8625_F|inactive</td>\n",
       "      <td>chr10:37754790-37754990</td>\n",
       "      <td>+</td>\n",
       "      <td>0.580</td>\n",
       "      <td>230</td>\n",
       "      <td>inactive</td>\n",
       "      <td>AGGACCGGATCAACTGAAAGCGCGTCTATCGGCCTTAGTTCGGACG...</td>\n",
       "      <td>AAGCGCGTCTATCGGCCTTAGTTCGGACGGTTCGATACGGGTCGGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        exp.id   null.id      scaffold.id                null_id+scaffold_id  \\\n",
       "0  concat-null  concat-0  WTC11_seq8625_F  concat-0|WTC11_seq8625_F|inactive   \n",
       "1  concat-null  concat-1  WTC11_seq8625_F  concat-1|WTC11_seq8625_F|inactive   \n",
       "2  concat-null  concat-2  WTC11_seq8625_F  concat-2|WTC11_seq8625_F|inactive   \n",
       "3  concat-null  concat-3  WTC11_seq8625_F  concat-3|WTC11_seq8625_F|inactive   \n",
       "4  concat-null  concat-4  WTC11_seq8625_F  concat-4|WTC11_seq8625_F|inactive   \n",
       "\n",
       "        hg38.scaffold.coor hg38.str  gc_frac  len scaffold.activity  \\\n",
       "0  chr10:37754790-37754990        +    0.570  230          inactive   \n",
       "1  chr10:37754790-37754990        +    0.560  230          inactive   \n",
       "2  chr10:37754790-37754990        +    0.580  230          inactive   \n",
       "3  chr10:37754790-37754990        +    0.585  230          inactive   \n",
       "4  chr10:37754790-37754990        +    0.580  230          inactive   \n",
       "\n",
       "                               200bp_insert+adaptors  \\\n",
       "0  AGGACCGGATCAACTGATATTCGCTCAACCGGAACGAATGTCGACG...   \n",
       "1  AGGACCGGATCAACTGATTCGAACGCAAGCGGTAATTACGTCGACG...   \n",
       "2  AGGACCGGATCAACTGACCATCGATCGAACGGCCGGTATTGGATCG...   \n",
       "3  AGGACCGGATCAACTGAACCGCTATATTGCGGTTATTGCGCGAGCG...   \n",
       "4  AGGACCGGATCAACTGAAAGCGCGTCTATCGGCCTTAGTTCGGACG...   \n",
       "\n",
       "                                          insert_seq  \n",
       "0  TATTCGCTCAACCGGAACGAATGTCGACGGTAACGTCGAGTGCGGT...  \n",
       "1  TTCGAACGCAAGCGGTAATTACGTCGACGGGTTGCGAGTAGTCGGC...  \n",
       "2  CCATCGATCGAACGGCCGGTATTGGATCGGCTAGATCGACGGAGGC...  \n",
       "3  ACCGCTATATTGCGGTTATTGCGCGAGCGGTATCGACGAAGGCGGG...  \n",
       "4  AAGCGCGTCTATCGGCCTTAGTTCGGACGGTTCGATACGGGTCGGC...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if os.path.exists(LIBRARY_CTRL) is False:\n",
    "    # get controls from library\n",
    "    ctrl = getAgarwalCtrls(lib)\n",
    "    print(ctrl.shape)\n",
    "\n",
    "    # rearrange columns to be like df\n",
    "    ctrl = ctrl[list(df)]\n",
    "\n",
    "    N_SCRAMBLE = 3\n",
    "    endog_scram = getEndogSeqNScramble(df, N_SCRAMBLE)\n",
    "\n",
    "    # add positive controls, scaffold scrambles\n",
    "    ctrl = pd.concat([ctrl, endog_scram])\n",
    "    print(ctrl.shape)\n",
    "\n",
    "    # add the nullomer inserts\n",
    "    ctrl = pd.concat([ctrl, df])\n",
    "\n",
    "    # quantify length\n",
    "    ctrl = checkLength(ctrl)\n",
    "    print(ctrl.shape)\n",
    "\n",
    "    ctrl = ctrl.rename(columns={\"seg_frag\": \"seq_insert\"})\n",
    "\n",
    "    # reorganize columns\n",
    "    ctrl = ctrl[[\n",
    "        'exp.id', 'null.id', 'scaffold.id',\n",
    "        'null_id|scaffold_id|activity', 'hg38.scaffold.coor', 'hg38.str',\n",
    "        'gc_frac', 'len',\n",
    "        'scaffold.activity', \n",
    "         '200bp_insert+adaptors',\n",
    "        'seq_frag',\n",
    "        #'200bp_insert',\n",
    "\n",
    "    ]].sort_values(by=[\"exp.id\", \"scaffold.id\"])\n",
    "\n",
    "    # rename column \"seq_frag\" to \"insert_seq\" \n",
    "    ctrl=ctrl.rename(columns = {\"seq_frag\":\"insert_seq\"})\n",
    "\n",
    "    # SAVE\n",
    "    ctrl.to_csv(LIBRARY_CTRL, sep='\\t', index=False)\n",
    "\n",
    "else:\n",
    "    ctrl = pd.read_csv(LIBRARY_CTRL, sep='\\t')\n",
    "\n",
    "ctrl.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b1d0ecb",
   "metadata": {},
   "source": [
    "## double check \n",
    "### adaptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b500db0c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T16:09:33.329532Z",
     "start_time": "2023-06-05T16:09:33.160414Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "({'AGGACCGGATCAACT'}, {'CATTGCGTGAACCGA'})"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(ctrl[\"200bp_insert+adaptors\"].apply(lambda x: x[:15])),set(ctrl[\"200bp_insert+adaptors\"].apply(lambda x: x[-15:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2d566ee2",
   "metadata": {},
   "outputs": [],
   "source": [
    "for_twist = ctrl[['null_id|scaffold_id|activity', '200bp_insert+adaptors']].drop_duplicates().copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f648108c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for_twist.to_csv(LIBRARY_TWIST, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89a9b876",
   "metadata": {},
   "source": [
    "### check scrambles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ae2a819f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T16:12:26.378085Z",
     "start_time": "2023-06-05T16:12:26.308128Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp.id</th>\n",
       "      <th>null.id</th>\n",
       "      <th>scaffold.id</th>\n",
       "      <th>null_id+scaffold_id</th>\n",
       "      <th>hg38.scaffold.coor</th>\n",
       "      <th>hg38.str</th>\n",
       "      <th>gc_frac</th>\n",
       "      <th>len</th>\n",
       "      <th>scaffold.activity</th>\n",
       "      <th>200bp_insert+adaptors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>764</th>\n",
       "      <td>endog</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HepG2_DNasePeakNoPromoter20370</td>\n",
       "      <td>endog</td>\n",
       "      <td>chr15:72237529-72237729</td>\n",
       "      <td>+</td>\n",
       "      <td>0.525</td>\n",
       "      <td>230</td>\n",
       "      <td>active</td>\n",
       "      <td>AGGACCGGATCAACTAGCTGGTAAGAAACCCAGTAATGACTCAGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764</th>\n",
       "      <td>endog-scramble</td>\n",
       "      <td>endog-scramble-0</td>\n",
       "      <td>HepG2_DNasePeakNoPromoter20370</td>\n",
       "      <td>endog-scramble-0|HepG2_DNasePeakNoPromoter20370</td>\n",
       "      <td>chr15:72237529-72237729</td>\n",
       "      <td>+</td>\n",
       "      <td>0.525</td>\n",
       "      <td>230</td>\n",
       "      <td>active</td>\n",
       "      <td>AGGACCGGATCAACTGGGGCCTGCCCGGACTAGGGGACCATAAACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764</th>\n",
       "      <td>endog-scramble</td>\n",
       "      <td>endog-scramble-1</td>\n",
       "      <td>HepG2_DNasePeakNoPromoter20370</td>\n",
       "      <td>endog-scramble-1|HepG2_DNasePeakNoPromoter20370</td>\n",
       "      <td>chr15:72237529-72237729</td>\n",
       "      <td>+</td>\n",
       "      <td>0.525</td>\n",
       "      <td>230</td>\n",
       "      <td>active</td>\n",
       "      <td>AGGACCGGATCAACTTAAAAACACATCCAGAAAGTACATCACTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764</th>\n",
       "      <td>endog-scramble</td>\n",
       "      <td>endog-scramble-2</td>\n",
       "      <td>HepG2_DNasePeakNoPromoter20370</td>\n",
       "      <td>endog-scramble-2|HepG2_DNasePeakNoPromoter20370</td>\n",
       "      <td>chr15:72237529-72237729</td>\n",
       "      <td>+</td>\n",
       "      <td>0.525</td>\n",
       "      <td>230</td>\n",
       "      <td>active</td>\n",
       "      <td>AGGACCGGATCAACTGACTCCGGCCTCAGTCCACTGGACTCCAGAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             exp.id           null.id                     scaffold.id  \\\n",
       "764           endog               NaN  HepG2_DNasePeakNoPromoter20370   \n",
       "764  endog-scramble  endog-scramble-0  HepG2_DNasePeakNoPromoter20370   \n",
       "764  endog-scramble  endog-scramble-1  HepG2_DNasePeakNoPromoter20370   \n",
       "764  endog-scramble  endog-scramble-2  HepG2_DNasePeakNoPromoter20370   \n",
       "\n",
       "                                 null_id+scaffold_id       hg38.scaffold.coor  \\\n",
       "764                                            endog  chr15:72237529-72237729   \n",
       "764  endog-scramble-0|HepG2_DNasePeakNoPromoter20370  chr15:72237529-72237729   \n",
       "764  endog-scramble-1|HepG2_DNasePeakNoPromoter20370  chr15:72237529-72237729   \n",
       "764  endog-scramble-2|HepG2_DNasePeakNoPromoter20370  chr15:72237529-72237729   \n",
       "\n",
       "    hg38.str  gc_frac  len scaffold.activity  \\\n",
       "764        +    0.525  230            active   \n",
       "764        +    0.525  230            active   \n",
       "764        +    0.525  230            active   \n",
       "764        +    0.525  230            active   \n",
       "\n",
       "                                 200bp_insert+adaptors  \n",
       "764  AGGACCGGATCAACTAGCTGGTAAGAAACCCAGTAATGACTCAGCA...  \n",
       "764  AGGACCGGATCAACTGGGGCCTGCCCGGACTAGGGGACCATAAACA...  \n",
       "764  AGGACCGGATCAACTTAAAAACACATCCAGAAAGTACATCACTGGC...  \n",
       "764  AGGACCGGATCAACTGACTCCGGCCTCAGTCCACTGGACTCCAGAG...  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrl.loc[(ctrl['scaffold.id'] == \"HepG2_DNasePeakNoPromoter20370\")\n",
    "         & (ctrl[\"exp.id\"].str.contains(\"endog\")), ['exp.id',\n",
    "                                                    'null.id',\n",
    "                                                    'scaffold.id',\n",
    "                                                    'null_id|scaffold_id|activity',\n",
    "                                                    'hg38.scaffold.coor',\n",
    "                                                    'hg38.str',\n",
    "                                                    'gc_frac',\n",
    "                                                    'len',\n",
    "                                                    'scaffold.activity',\n",
    "                                                    # 'seq_frag',\n",
    "                                                    # '200bp_insert',\n",
    "                                                    '200bp_insert+adaptors']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "434aba0c",
   "metadata": {},
   "source": [
    "### test for k562 alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "9556424e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T19:05:40.772195Z",
     "start_time": "2023-06-05T19:05:40.690240Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "scaffold.id                                              K562_peak26939\n",
       "200bp_insert          TCAAAGTCAACCAGGCAGCTGCCTTACTTAAATTCATTTCCCTCAG...\n",
       "hg38.scaffold.coor                              chr15:40962392-40962592\n",
       "Name: 480, dtype: object"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_cols = [\"scaffold.id\", \"200bp_insert\", \"hg38.scaffold.coor\"]\n",
    "ctrl.loc[ctrl[\"scaffold.id\"].str.contains(\"K562\"), check_cols].iloc[271]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe7e40c2",
   "metadata": {},
   "source": [
    "### check one sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90b978d6",
   "metadata": {},
   "source": [
    "### align ucsc v. our seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "7f6d88a1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-05T19:04:32.741436Z",
     "start_time": "2023-06-05T19:04:32.736287Z"
    }
   },
   "outputs": [],
   "source": [
    "# ids we tested\n",
    "\n",
    "IDS = {\n",
    "    \"K562_peak10458\":\"CCCAGATCAATAAACTTGCTCATCTGCTCTTATGGCCCCCACCCAGTAACTGACTCAGTGCAAGAAGACAGCTGACTCACGGCCATTTCATCTCTGACCAATCAGCACCCTTGGCTCACTGGCTTACCGATTTCATCTCTGACCAATCAGCACTCCTGGCTCACTGGCTTACCAATTTCATCTCTGACCAATCAGCACTC\",\n",
    "    \"K562_peak49867\":\"GCGGGGGAGGTGAGATTGGAAGGATGCCAGGAATCTTGGGCCCAGCTGCTGCTTTATGACTTAGGGATATAACTGCTTCTGACTCATGCGTGCTTGGCTCTTTGTGATTATGTGCTGAGACACGTCAGCGGGAGGCTGCCCCGGAATGTCTCCCCATGCGAATCGAGGGTGGGGCCTCTGCTGAGTCACCTCTGGCTCCT\", \n",
    "    \"K562_peak26939\":\"AGCGCCCTGACAAAACAGGCCACTCGGCTCTACCAATCAGCAAGATGTGGGTGGGGCCAGATAAGAGAATAAAAGCAGGCTGCCCGAGCTTGCATTAGTAACTCGCTCGGGTCCCCTTCCGGATTGTGGAAACTTTGTTCTTTCGCTGTATGTAATAACTCTTGCTACTGTAGCTCGCTTTTTGGGTCTACGCTGCTTCT\"\n",
    "}\n",
    "\n",
    "# align our sequence to UCSC sequence\n",
    "\n",
    "for name, ucsc_seq in IDS.items():\n",
    "    \n",
    "\n",
    "    # check MPRA sequence\n",
    "    mpra_seq=ctrl.loc[ctrl[\"scaffold.id\"]==name, \"200bp_insert\"].iloc[0]\n",
    "    v=0\n",
    "\tfor i, j in zip(ucsc_seq, mpra_seq):\n",
    "\t    if i==j:  # if nucleotide aligns\n",
    "\t        continue\n",
    "\t    else:   # if nucleotide does not aligns\n",
    "\t        print(v, i, j)\n",
    "\t    v+=1  # count the index\n",
    "\t\n",
    "## all results returned were negative\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e84070b6",
   "metadata": {},
   "source": [
    "## counts of all the experimental categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "8694e2fc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-01T21:05:53.573640Z",
     "start_time": "2023-06-01T21:05:53.544654Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(exp.id\n",
       " concat-null                 325\n",
       " endog                       400\n",
       " endog-scramble             1200\n",
       " negative HepG2, Smith       169\n",
       " positive HepG2, Smith       181\n",
       " positive K562, Kellis        50\n",
       " single-null              107200\n",
       " Name: hg38.scaffold.coor, dtype: int64,\n",
       " 109525)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrl.groupby(\"exp.id\")[\"hg38.scaffold.coor\"].count(), ctrl.groupby(\"exp.id\")[\"hg38.scaffold.coor\"].count().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "8ea81cb3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-01T21:08:42.861912Z",
     "start_time": "2023-06-01T21:08:42.812968Z"
    }
   },
   "outputs": [],
   "source": [
    "frag, insert = ctrl.loc[(ctrl[\"scaffold.activity\"]==\"active\") \n",
    "         & (ctrl[\"exp.id\"]==\"single-null\"), \n",
    "         [\"seq_frag\", \"200bp_insert\"]].iloc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea37e162",
   "metadata": {},
   "source": [
    "# GC content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "e62c7141",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-01T22:20:21.822037Z",
     "start_time": "2023-06-01T22:20:21.805012Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    107925.000000\n",
       "mean          0.522703\n",
       "std           0.047965\n",
       "min           0.395000\n",
       "25%           0.490000\n",
       "50%           0.525000\n",
       "75%           0.565000\n",
       "max           0.615000\n",
       "Name: gc_frac, dtype: float64"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"gc_frac\"].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "b95ea43d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-01T22:20:27.577255Z",
     "start_time": "2023-06-01T22:20:22.316171Z"
    }
   },
   "outputs": [],
   "source": [
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "3435b718",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-01T22:20:38.441198Z",
     "start_time": "2023-06-01T22:20:38.424047Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'concat-null', 'endog', 'single-null'}"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df[\"exp.id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "6d3992be",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-01T22:20:39.424104Z",
     "start_time": "2023-06-01T22:20:38.880602Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: xlabel='gc_frac', ylabel='Count'>"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAj4AAAGsCAYAAADddK15AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy88F64QAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAntElEQVR4nO3df1BV953/8RcKEriAIiBGglZDsy0/tumWrJMYqmg3jTbR1TqjGyhRcJoxriO6Oq7titspKo1tF2113Vkxwaw2kNjsNkmxJis0XZOJcTfVFmMi0UR+iKIC9xoiXuV8/+jX21y5oOj9xf08HzNnJvfzuefc9/GTE185n/MjzLIsSwAAAAYYEugCAAAA/IXgAwAAjEHwAQAAxiD4AAAAYxB8AACAMQg+AADAGAQfAABgjPBAFxBMenp61NLSotjYWIWFhQW6HAAAcAssy5LD4dCYMWM0ZEj/53QIPp/T0tKi1NTUQJcBAABuQ2Njo+65555+v0Pw+ZzY2FhJf/qDi4uLC3A1AADgVtjtdqWmprr+Hu8Pwedzrk9vxcXFEXwAABhkbuUyFS5uBgAAxiD4AAAAYxB8AACAMQg+AADAGAQfAABgDIIPAAAwBsEHAAAYg+ADAACMQfABAADGIPgAAABjEHwAAIAxCD4AAMAYBB8AAGAMgg8AADAGwQcAABgjPNAFAIC/zCsoUktbu8e+MUnxqtpV4dPf8eZvALg9BB8Axmhpa1fK7FUe+2o3Fipn+pxe7bcTVvr6neaXNw1oOwC8j+ADAJKc1hDCCmAArvEBAADGIPgAAABjEHwAAIAxCD4AAMAYBB8AAGAMgg8AADAGwQcAABiD4AMAAIxB8AEAAMYg+AAAAGMQfAAAgDEIPgAAwBgEHwAAYAyCDwAAMAbBBwAAGIPgAwAAjEHwAQAAxiD4AAAAYxB8AACAMQg+AADAGAQfAABgDIIPAAAwBsEHAAAYg+ADAACMQfABAADGIPgAAABjEHwAAIAxwgNdAADAu+YVFKmlrd1j35ikeFXtqvBzRX/WV22BrgvmIPgAQIhpaWtXyuxVHvuaX97k52rc9VVboOuCOZjqAgAAxiD4AAAAYxB8AACAMQg+AADAGAQfAABgDIIPAAAwBsEHAAAYg+ADAACMQfABAADGIPgAAABjBDT4XLt2TVOmTNGCBQtcbe+8844mTpyomJgYjR8/XhUV7u9uqaysVFpammw2m7Kzs/X222+7bW/VqlVKTk5WbGysZs2apTNnzvhrdwAAQJALaPD5wQ9+oN/97neuz+3t7ZoxY4YKCgrU0dGhiooKLV++XIcOHZIk1dXVaenSpaqsrFRHR4fy8vI0c+ZMdXV1SZJKS0u1f/9+HT58WM3NzYqKitKiRYsCsm8AACD4BCz4HDhwQHv37tW3v/1tV9vevXuVkJCgJUuWKDw8XFOnTlVeXp62bt0qSdqxY4fmz5+vSZMmKSIiQsuXL1diYqKqqqpc/atXr1Zqaqri4uK0efNm1dTU6OTJkx5r6O7ult1ud1sAAEDoCkjwOXfunIqKirRnzx5FR0e72uvr65WVleX23fT0dB05cuSm/Z2dnWpqanLrT05OVnx8vI4ePeqxjo0bN2r48OGuJTU11Vu7CAAAglC4v3+wp6dH+fn5WrFihb7yla+49TkcDtlsNre26OhoXbp06ab9DodDkvpd/0Zr1qzRihUrXJ/tdjvhBwiQeQVFamlr79U+JileVbsqPKwBAAPn9+CzceNG3XXXXVq6dGmvPpvNpo6ODre2rq4uxcbGuvqvX8/z+f7ExERX4PHUf339G0VGRioyMvJ2dwWAF7W0tStl9qpe7c0vbwpANQBCld+nup5//nnV1dVpxIgRGjFihPbs2aM9e/ZoxIgRyszMVH19vdv3jx07pszMTEnqtz8+Pl4pKSlu/a2trbp48aJrfQAAYDa/B5/jx4/Lbrero6NDHR0deuKJJ/TEE0+oo6NDc+bMUWtrq8rLy+V0OlVbW6vdu3ersLBQklRYWKjdu3ertrZWTqdT5eXlOnv2rGbPni1JWrhwoUpLS3Xq1Ck5HA4VFxdr8uTJuvfee/29mwAAIAj5faqrPwkJCXr99de1bNkylZSUKCkpSVu2bFFubq4kadq0adq2bZsWL16spqYmZWRkqKamRiNHjpQklZSUyOl0KicnRw6HQ7m5uaqurg7kLgEAgCAS8ODz3HPPuX3Ozs7WwYMH+/x+fn6+8vPzPfZFRESorKxMZWVl3iwRAACECF5ZAQAAjEHwAQAAxiD4AAAAYxB8AACAMQg+AADAGAQfAABgDIIPAAAwBsEHAAAYg+ADAACMQfABAADGIPgAAABjBPxdXQAwWM0rKFJLW3uv9hMNDUoJQD34s77GZkxSvKp2VQSgIgQLgg8A3KaWtnalzF7Vq71+w6IAVIPP62tsml/eFIBqEEyY6gIAAMYg+AAAAGMw1QUAMMaHx99XzvQ5Hvu4/scMBB8AgDGc1hCP1/5IXP9jCqa6AACAMTjjAwBBgNuvAf8g+ABAEAj07dd9BS+J8IXQQvABAPQZvCSufUFo4RofAABgDM74APCb/qZTeM0DAH8g+ADwm/6mU3jNAwB/YKoLAAAYg+ADAACMQfABAADGIPgAAABjEHwAAIAxuKsLAIDbwNOuByeCDwAAt4GnXQ9OTHUBAABjEHwAAIAxCD4AAMAYBB8AAGAMLm4GgCD24fH3lTN9jsc+7hwCBo7gAwBBzGkN4c4hwIuY6gIAAMYg+AAAAGMQfAAAgDEIPgAAwBgEHwAAYAzu6gIAoB99vYz0REODUgJQD+4MwQcAgH709TLS+g2LAlAN7hRTXQAAwBgEHwAAYAyCDwAAMAbBBwAAGIPgAwAAjEHwAQAAxiD4AAAAYxB8AACAMQg+AADAGAQfAABgDIIPAAAwBsEHAAAYg+ADAACMQfABAADGIPgAAABjEHwAAIAxwgNdAADcjnkFRWppa/fYNyYpXlW7KvxcEYDBgOADYFBqaWtXyuxVHvuaX97k52oADBZMdQEAAGMQfAAAgDEIPgAAwBgEHwAAYIyABJ8DBw5o4sSJiouL0+jRo7V06VJ99tlnkqR33nlHEydOVExMjMaPH6+KCvc7MyorK5WWliabzabs7Gy9/fbbrr5r165p1apVSk5OVmxsrGbNmqUzZ874dd8AAEDw8nvwaWtr07e+9S0tXrxYHR0deu+991RXV6eysjK1t7drxowZKigoUEdHhyoqKrR8+XIdOnRIklRXV6elS5eqsrJSHR0dysvL08yZM9XV1SVJKi0t1f79+3X48GE1NzcrKipKixYt8vcuAgCAIOX34JOUlKRz585pwYIFCgsL04ULF3T58mUlJSVp7969SkhI0JIlSxQeHq6pU6cqLy9PW7dulSTt2LFD8+fP16RJkxQREaHly5crMTFRVVVVrv7Vq1crNTVVcXFx2rx5s2pqanTy5El/7yYAAAhCAZnqio2NlSSlpqYqKytLd999txYuXKj6+nplZWW5fTc9PV1HjhyRpH77Ozs71dTU5NafnJys+Ph4HT161GMd3d3dstvtbgsAAAhdAb24+cSJE2pubtbQoUM1d+5cORwO2Ww2t+9ER0fr0qVLktRvv8PhkKR+17/Rxo0bNXz4cNeSmprqrV0DAABBKKDBJyoqSmPGjNGPfvQj7du3TzabzXW9znVdXV2uM0T99V8PPP2tf6M1a9aos7PTtTQ2Nnpr1wAAQBDye/B566239KUvfUlXrlxxtXV3d2vYsGFKT09XfX292/ePHTumzMxMSVJmZmaf/fHx8UpJSXHrb21t1cWLF13r3ygyMlJxcXFuCwAACF1+Dz5/+Zd/qa6uLv3jP/6jrly5ok8++UQrV65UUVGR5s6dq9bWVpWXl8vpdKq2tla7d+9WYWGhJKmwsFC7d+9WbW2tnE6nysvLdfbsWc2ePVuStHDhQpWWlurUqVNyOBwqLi7W5MmTde+99/p7NwEAXjKvoEg50+f0WuYVFAW6NAxCfn9JaUxMjPbt26fi4mIlJydr+PDhys/P19q1axUZGanXX39dy5YtU0lJiZKSkrRlyxbl5uZKkqZNm6Zt27Zp8eLFampqUkZGhmpqajRy5EhJUklJiZxOp3JycuRwOJSbm6vq6mp/7yIAwIv6eiEtL6PF7QjI29nT09O1f/9+j33Z2dk6ePBgn+vm5+crPz/fY19ERITKyspUVlbmlToBAEBo4ZUVAADAGAQfAABgDIIPAAAwBsEHAAAYg+ADAACMQfABAADGCMjt7AACY15BkVra2nu1j0mKV9WuigBUBAD+RfABDMKD4ACYjqkuAABgDIIPAAAwBlNdAAAYxPRr/Qg+AAAYxPRr/Qg+AAAMUqafvbkdBB8AAAYp08/e3A6CDwAAQayvszqSdKKhQSl+rmewI/gAABDE+jqrI0n1Gxb5uZrBj9vZAQCAMQg+AADAGEx1Abgt/V13wB0lAIIVwQfAbenvugPuKAEQrAg+AACv4i4kBDOCDwDAq7gLCcHMaxc32+12b20KAADAJwYcfEaOHOmxfezYsXdcDAAAgC/d0lRXQ0ODnnrqKVmWJbvdrqlTp7r12+12jRgxwhf1AQAAeM0tBZ+0tDR9+9vfVltbmw4ePKjJkye79d911116/PHHfVIgAACAt9zyxc1PP/20JGn8+PEqKCjwWUEAAAC+MuC7ugoKCnTo0CF9+OGH6unp6dUHAAAQrAYcfL73ve/pRz/6ke6++25FRES42sPCwgg+AAAgqA04+Dz//PN69dVXNX36dF/UAwAA4DMDDj6XLl3So48+6otaAAAIabzjLvAGHHwee+wx7dmzR3l5eb6oBwCAkMU77gJvwMHn8uXLevLJJ7V+/XqNHj3are/AgQNeKwwAAMDbBhx8MjMzlZmZ6YtaAAAAfGrAwWfdunW+qAMAAMDnBhx8CgsL++zbuXPnHRUDAADgSwMOPpZluX0+f/68Dhw40G8gAjB49XUXyomGBqUEoB4AuBMDDj7PPvtsr7Y33nhD27Zt80pBAIJLX3eh1G9YFIBqAODODDj4ePKNb3xDc+fO9camAABB5sPj7ytn+pxe7Tx3BoPRHQefq1ev6he/+IWSkpK8UQ8AIMg4rSEez/rx3BkMRgMOPkOGDFFYWJj7RsLDtXnzZq8VBQAA4AsDDj61tbVun4cOHaq0tLReDzMEAAAINkMGusLkyZOVk5OjqKgonT17VpZladSoUb6oDQAAwKsGfMantbVVjz/+uH7/+98rISFB58+f13333af9+/frnnvu8UWNAAAAXjHgMz4rV67Ufffdp/b2drW2turChQu6//77tWLFCl/UBwAA4DUDPuNz4MABHT9+XDExMZKk4cOH61//9V81fvx4rxcHAADgTQMOPteuXdOQIe4nisLCwjRs2DCvFQXg5vp6ojLPVgGAvg04+OTm5mrx4sXavn27bDabLl26pKefflpTpkzxQXkA+tLXE5V5tgoA9G3AweeZZ57RN77xDcXHxysxMVHnz59Xenq6XnvtNV/UBwAA4DUDCj6WZcnpdOrYsWP63e9+p7Nnz+qTTz7RypUrNXToUF/VCAAA4BW3HHw+/fRTPfLII0pOTtYvf/lL5ebm6ty5cxo3bpxeffVV7du3TzabzZe1AgAA3JFbvp29tLRUw4YN0/bt211to0aN0ieffCKn06mNGzf6pEAAAABvueXg89JLL+nf//3fez2ledSoUdq+fbuqq6u9XhwAAIA33XLwOXfunNLS0jz23X///WptbfVaUQAAAL5wy8EnLi5OFy5c8Nh38eJFRUdHe60oAAAAX7jl4DNt2jRt3brVY9+2bdv04IMPeq0oAAAAX7jlu7q+973v6a/+6q/U1tam+fPna/To0Tpz5oyqqqq0c+dOvfnmm76sEwAA4I7dcvC5/gb2p556Slu3blVYWJgsy1JWVpZqamr0ta99zZd1Asbq69UUJxoalBKAegBgMBvQAwwfeugh/eEPf9DJkyfV1tamu+++W2PHjvVVbQDU96sp6jcsCkA1ADC4DfiVFZI0YcIETZgwwdu1AAAA+NQtX9wMAAAw2BF8AACAMQg+AADAGAQfAABgDIIPAAAwBsEHAAAYg+ADAACMEZDgc+TIEf3N3/yNRo4cqdGjR6ugoEDnz5+XJL3zzjuaOHGiYmJiNH78eFVUVLitW1lZqbS0NNlsNmVnZ+vtt9929V27dk2rVq1ScnKyYmNjNWvWLJ05c8av+wYAAIKX34PPZ599punTp+uhhx5Sa2ur6uvrdeHCBS1cuFDt7e2aMWOGCgoK1NHRoYqKCi1fvlyHDh2SJNXV1Wnp0qWqrKxUR0eH8vLyNHPmTHV1dUmSSktLtX//fh0+fFjNzc2KiorSokU83RYAAPyJ34PP6dOn9ZWvfEUlJSUaNmyYEhIS9NRTT+nNN9/U3r17lZCQoCVLlig8PFxTp05VXl6e663wO3bs0Pz58zVp0iRFRERo+fLlSkxMVFVVlat/9erVSk1NVVxcnDZv3qyamhqdPHnSYy3d3d2y2+1uCwAACF1+Dz5/8Rd/oZqaGg0dOtTV9tJLL+lrX/ua6uvrlZWV5fb99PR0HTlyRJL67e/s7FRTU5Nbf3JysuLj43X06FGPtWzcuFHDhw93Lampqd7aTQAAEIQCenGzZVn6p3/6J73yyivavHmzHA6HbDab23eio6N16dIlSeq33+FwSFK/699ozZo16uzsdC2NjY3e2jUAABCEbuslpd5gt9u1cOFC/e///q/efPNNZWVlyWazqaOjw+17XV1dio2NlfSnUHP9ep7P9ycmJroCj6f+6+vfKDIyUpGRkV7aIwAAEOwCcsbno48+0gMPPCC73a7Dhw+7pqcyMzNVX1/v9t1jx44pMzPzpv3x8fFKSUlx629tbdXFixdd6wMAALP5Pfi0t7dr6tSpeuihh/Sb3/xGiYmJrr45c+aotbVV5eXlcjqdqq2t1e7du1VYWChJKiws1O7du1VbWyun06ny8nKdPXtWs2fPliQtXLhQpaWlOnXqlBwOh4qLizV58mTde++9/t5NAAAQhPw+1fXss8/q9OnTqq6u1osvvujWd+nSJb3++utatmyZSkpKlJSUpC1btig3N1eSNG3aNG3btk2LFy9WU1OTMjIyVFNTo5EjR0qSSkpK5HQ6lZOTI4fDodzcXFVXV/t7F4E+zSsoUktbe6/2MUnxqtpV4WENAIA3+T34rFixQitWrOizPzs7WwcPHuyzPz8/X/n5+R77IiIiVFZWprKysjuuE/CFlrZ2pcxe1au9+eVNAagGAMzDKysAAIAxCD4AAMAYBB8AAGAMgg8AADAGwQcAABiD4AMAAIxB8AEAAMYg+AAAAGMQfAAAgDEIPgAAwBgEHwAAYAyCDwAAMAbBBwAAGIPgAwAAjEHwAQAAxiD4AAAAYxB8AACAMQg+AADAGAQfAABgDIIPAAAwRnigCwC8bV5BkVra2nu1j0mKV9WuigBUBAAIFgQfhJyWtnalzF7Vq7355U0BqAYAEEyY6gIAAMYg+AAAAGMQfAAAgDEIPgAAwBgEHwAAYAzu6gL60det8RK3xwPAYETwAfrR163xErfHA8BgxFQXAAAwBsEHAAAYg+ADAACMQfABAADGIPgAAABjcFcX4GX93QJ/oqFBKX6uBwDwZwQfwMv6uwW+fsMiP1cDAPg8proAAIAxCD4AAMAYBB8AAGAMgg8AADAGwQcAABiD4AMAAIxB8AEAAMYg+AAAAGMQfAAAgDEIPgAAwBgEHwAAYAyCDwAAMAbBBwAAGIPgAwAAjEHwAQAAxiD4AAAAYxB8AACAMcIDXQDQn3kFRWppa/fYNyYpXlW7KvxcEQBgMCP4IKi1tLUrZfYqj33NL2/yczUAgMGOqS4AAGAMgg8AADAGwQcAABiDa3wAAIA+PP6+cqbP6dUeajeSEHwAAICc1hCPN5OE2o0kTHUBAABjEHwAAIAxCD4AAMAYBB8AAGAMgg8AADAGwQcAABgjoMGnra1NaWlpqqurc7W98847mjhxomJiYjR+/HhVVLg/O6CyslJpaWmy2WzKzs7W22+/7eq7du2aVq1apeTkZMXGxmrWrFk6c+aMv3YHAAAEuYAFn4MHD+rBBx/URx995Gprb2/XjBkzVFBQoI6ODlVUVGj58uU6dOiQJKmurk5Lly5VZWWlOjo6lJeXp5kzZ6qrq0uSVFpaqv379+vw4cNqbm5WVFSUFi1aFJD9AwAAwScgwaeyslJPPPGE1q9f79a+d+9eJSQkaMmSJQoPD9fUqVOVl5enrVu3SpJ27Nih+fPna9KkSYqIiNDy5cuVmJioqqoqV//q1auVmpqquLg4bd68WTU1NTp58qTf9xEAAASfgASfb37zm/roo480b948t/b6+nplZWW5taWnp+vIkSM37e/s7FRTU5Nbf3JysuLj43X06FGPdXR3d8tut7stAAAgdAUk+IwePVrh4b3fluFwOGSz2dzaoqOjdenSpZv2OxwOSep3/Rtt3LhRw4cPdy2pqam3vU8AACD4BdW7umw2mzo6Otzaurq6FBsb6+q/fj3P5/sTExNdgcdT//X1b7RmzRqtWLHC9dlutxN+QlhfL+CTQu8lfAAAz4Iq+GRmZmr//v1ubceOHVNmZqarv76+vlf/jBkzFB8fr5SUFNXX17u+39raqosXL7o+3ygyMlKRkZE+2BOzzSsoUktbe6/2QIeLvl7AJ4XeS/gAAJ4F1XN85syZo9bWVpWXl8vpdKq2tla7d+9WYWGhJKmwsFC7d+9WbW2tnE6nysvLdfbsWc2ePVuStHDhQpWWlurUqVNyOBwqLi7W5MmTde+99wZyt4zT0taulNmrei2ewhAAAP4UVGd8EhIS9Prrr2vZsmUqKSlRUlKStmzZotzcXEnStGnTtG3bNi1evFhNTU3KyMhQTU2NRo4cKUkqKSmR0+lUTk6OHA6HcnNzVV1dHchdAgAAQSTgwceyLLfP2dnZOnjwYJ/fz8/PV35+vse+iIgIlZWVqayszKs1AgCA0BBUU10AAAC+RPABAADGIPgAAABjEHwAAIAxCD4AAMAYAb+rC8EtWB9GCADA7SD4oF/XH0Z4I550DAAYjJjqAgAAxuCMD/qczpKkEw0NSvFzPQAA+ArBB31OZ0lS/YZFfq4GAADfYaoLAAAYg+ADAACMQfABAADGIPgAAABjEHwAAIAxCD4AAMAYBB8AAGAMgg8AADAGwQcAABiDJzcjKPT12gxemQEA8CaCD4JCX6/N4JUZAABvYqoLAAAYg+ADAACMQfABAADGIPgAAABjEHwAAIAxCD4AAMAYBB8AAGAMgg8AADAGwQcAABiD4AMAAIxB8AEAAMYg+AAAAGMQfAAAgDEIPgAAwBjhgS4AAAAMTvMKitTS1t6rfUxSvKp2VQSgopsj+AAAgNvS0taulNmrerU3v7wpANXcGqa6AACAMQg+AADAGAQfAABgDIIPAAAwBhc3h6DBeJU9AAD+QPAJQYPxKnsAAPyBqS4AAGAMgg8AADAGwQcAABiD4AMAAIxB8AEAAMYg+AAAAGNwO/sg1dezeiTpREODUvxcDwAAgwHBZ5Dq61k9klS/YZGfqwEAYHBgqgsAABiD4AMAAIxB8AEAAMYg+AAAAGNwcTMAAPCqD4+/r5zpczz2jUmKV9WuCj9X9GcEHwAA4FVOa0ifdx43v7zJz9W4Y6oLAAAYg+ADAACMQfABAADGIPgAAABjEHwAAIAxCD4AAMAYBB8AAGAMgg8AADAGwQcAABiDJzcHgXkFRWppa/fYF+hHewMAEEpCLvicO3dO3/3ud1VXV6fw8HDl5+frxz/+scLDg3dXW9rag/bR3gAAhJKQm+qaN2+eYmJi1NLSokOHDumNN97Qv/zLvwS6LAAAEARCKvg0NDSorq5OzzzzjKKjozVhwgStXbtWP//5zwNdGgAACALBO/9zG+rr6zVy5EiNGTPG1Zaenq7Tp0+ro6NDI0aMcPt+d3e3uru7XZ87OzslSXa73Sf1LfjuEp250NGr/aOTJzXqs089rnP1qtNjPVevOuXsYx2r55rHPn9sq7/t3c46ff1+f9vzx7Zudx1v/nmG0tj0tz3Tx2Yw/nvLf1OCd2z6216g/729E9e3Z1nWzb9shZDnn3/eSk1NdWtraGiwJFmNjY29vr9u3TpLEgsLCwsLC0sILJ7+rr9RSJ3xsdls6urqcmu7/jk2NrbX99esWaMVK1a4Pvf09OjixYtKSEhQWFiYb4vtg91uV2pqqhobGxUXFxeQGuAZYxOcGJfgxdgEp1AcF8uy5HA43GZ8+hJSwSczM1MXLlzQ2bNnlZycLEk6duyY7rnnHg0fPrzX9yMjIxUZGenWduN0WKDExcWFzL+QoYaxCU6MS/BibIJTqI2Lp7/nPQmpi5u/+MUv6uGHH1ZxcbEcDodOnTqlH/7whyoqKgp0aQAAIAiEVPCRpJdeeklXr17V+PHjNXHiRD366KNau3ZtoMsCAABBIKSmuiQpOTlZL774YqDLuG2RkZFat25dryk4BB5jE5wYl+DF2AQn08clzLJu5d4vAACAwS/kproAAAD6QvABAADGIPgAAABjEHz84Ny5c/rbv/1bjRgxQomJiSouLtbVq1f7XeePf/yjoqOjVVdX59b+zDPP6J577pHNZtOUKVP0wQcf+LDy0Oetsenp6VFMTIxsNptiYmJcy6efen5kO/o3kHGZPn267rrrLrc/93379rn6OWa8y1tjwzHjXQMZl9/+9reaOHGiYmJilJqaqo0bN7r1h/wx452XRaA/U6ZMsfLy8qxPP/3U+uijj6yMjAzrmWee6fP7n376qZWZmWlJsmpra13tzz33nJWSkmL98Y9/tD777DNrxYoVVkZGhtXT0+OHvQhN3hqbP/zhD9awYcOs7u5uP1Qd+gYyLomJiVZdXZ3HPo4Z7/PW2HDMeNetjsv7779vRUdHW88995zV09NjHTlyxEpISLBefPFFy7LMOGYIPj524sQJS5LV3NzsanvhhRessWPH9rnOk08+aa1du7bXX66TJk2y1q9f7/p85coVKzY21jpw4IBPag913hybnTt3WtnZ2b4s1xgDGZeTJ09aQ4YMsex2u8dtccx4lzfHhmPGewYyLn//939vPfHEE25tH3zwgXXmzBnLssw4Zpjq8rGbvTH+Rrt27VJDQ4PWrVvncVtZWVmuzxEREfriF7+oI0eO+KT2UOfNsXn33Xf12Wef6YEHHlBSUpK+/vWv66233vJl+SFrIOPy7rvvKjY2VvPmzVNSUpIyMzO1c+dOt21xzHiPN8eGY8Z7BjIuhw4d0he+8AX93d/9nRITE/XlL39ZdXV1Gj16tGtboX7MEHx8zOFwyGazubVFR0dLki5duuTWfvz4cX3/+9/Xnj17NHTo0Fve1o3bwa3x5thERUVp4sSJ+s///E+dPn1aM2fO1De/+U2dOnXKdzsQogYyLt3d3XrwwQe1fv16tbS06Kc//amWLVvmeogpx4x3eXNsOGa8ZyDjcvHiRW3ZskX5+flqbW3Vv/3bv2nlypV66aWX+t1WKB0zBB8fu9U3xl++fFnz5s1TeXm5xo4dO6BteXrzPG7Om2Pzk5/8RBUVFUpJSVFUVJRWrlypsWPH6rXXXvPdDoSoWx0XSfrOd76jmpoaffWrX1VERIQeeeQRFRQUqKqqqt9tcczcHm+ODceM9wxkXCIjIzVr1ix961vfUnh4uL7+9a/rO9/5jqqrq/vdVigdMwQfH/v8G+Ov8/TG+HfffVcffvihioqKNGLECNdb4h977DE9/fTTrm3V19e71nE6nTpx4oQyMzP9szMhxptj8/3vf1/vvfee2/a7u7sVFRXl+x0JMbc6LpK0c+fOXq+o+fyfO8eMd3lzbDhmvGcg45Kenq7u7m63tmvXrsn6/y9xMOKYCfRFRiZ4+OGHrfnz51t2u906efKklZGRYa1bt+6m6+mGC2h37NhhpaSkWL///e9dV9unpaVZV65c8V3xIc5bYzNz5kwrJyfHOnPmjHX58mXrBz/4gZWUlGRduHDBd8WHsFsdl5/+9KfWqFGjrP/7v/+zrl27Zr366qtWVFSU9eabb1qWxTHjC94aG44Z77rVcfnv//5vKzw83Hr++eetnp4e67e//a0VExNj/dd//ZdlWWYcMwQfP2htbbXmzp1rJSQkWElJSdY//MM/WFevXrUsy7JsNpv1H//xHx7Xu/Ev156eHuvHP/6xNX78eCsmJsbKzc21PvjgA3/sQsjy1thcuHDBWrBggTVq1CjLZrNZubm51pEjR/yxCyHpVselp6fH+uEPf2iNGzfOioqKsjIyMly35V7v55jxLm+NDceMdw3kv2W//vWvrezsbCs2NtaaMGGCtX37dlefCccMLykFAADG4BofAABgDIIPAAAwBsEHAAAYg+ADAACMQfABAADGIPgAAABjEHwAAIAxCD4AAMAYBB8Axjhz5oz++q//WjabTfn5+YEuB0AAhAe6AADwlwMHDujjjz/WxYsXFRkZGehyAAQAZ3wABKX33ntPDz/8sGJjY3X//fdr/fr1+sIXviBJ2rNnjzIzMxUTE6Mvf/nLqq6uvun2tmzZoqKiIp0/f14JCQl64403NGXKFC1YsEDjxo3T2LFj5XA49Morr+ihhx7SqFGjFB0drcmTJ+vEiROu7dzObwMIIoF+WRgA3Kizs9MaNWqUtXbtWuvy5cvW0aNHrbFjx1rjxo2zamtrrcjISOu1116zrl27ZtXU1FgRERFWfX39Tbf77LPPWuPGjXN9njx5spWSkmI1NTVZ7e3tVmNjoxUZGWn96le/sizLss6fP2/l5ORY+fn5lmVZd/TbAIIDZ3wABJ1XXnlF4eHh+ud//mdFRkYqKytLq1evliRVVlZqzpw5mjFjhoYMGaJHH31UBw8eVEpKym391vTp05WSkqIRI0Zo1KhRqq+v1+OPPy6Hw6HGxkYlJiaqubnZJ78NwP+4xgdA0GlsbNTYsWM1ZMif/99swoQJkv50gfJXv/pVt+8/8MADt/1bY8aMcf1zRESEfvGLX2j79u0KCwtTVlaW7Ha7wsPDffLbAPyPMz4Ags64ceN0+vRpWZblavvkk08kSampqTp9+rTb93/yk5/orbfeuq3fCgsLc/1zdXW1fvazn6murk6NjY369a9/7RZ0vP3bAPyP4AMg6MycOVM9PT3asGGDrly5og8++ECbNm2SJC1YsEC//OUvtX//fvX09Og3v/mN1q1bp+HDh9/x73Z2dmro0KGKioqSZVnat2+fdu3apStXrvj8twH4B1NdAIKOzWbTr371Ky1ZskQbNmzQfffdp0ceeURvvPGGJk2apF27dmnlypX6+OOPNW7cOL3wwgvKyMi449998skn9T//8z/KyMhQeHi4vvSlL6m4uFg///nPdeXKFZ/+NgD/CLM+fy4ZAILAhQsXdPz4cU2aNMnV9rOf/UwvvPCCDh48GMDKAAx2THUBCDpXr17V1KlTVVNTI0n6+OOPtW3bNj322GMBrgzAYEfwARB0kpOTVV1drdWrVys2NlY5OTmaO3euVq1a1ec6xcXFiomJ6XPZsGGDH/cAQLBiqgsAABiDMz4AAMAYBB8AAGAMgg8AADAGwQcAABiD4AMAAIxB8AEAAMYg+AAAAGMQfAAAgDH+H/JtLV1banlfAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.histplot(df[\"gc_frac\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "af2fbb8d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-01T22:20:40.028026Z",
     "start_time": "2023-06-01T22:20:40.004933Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>null_id+scaffold_id</th>\n",
       "      <th>null.id</th>\n",
       "      <th>scaffold.id</th>\n",
       "      <th>hg38.scaffold.coor</th>\n",
       "      <th>hg38.str</th>\n",
       "      <th>scaffold.activity</th>\n",
       "      <th>seq_frag</th>\n",
       "      <th>200bp_insert</th>\n",
       "      <th>200bp_insert+adaptors</th>\n",
       "      <th>gc_frac</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exp.id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>single-null</th>\n",
       "      <td>1830</td>\n",
       "      <td>1830</td>\n",
       "      <td>1830</td>\n",
       "      <td>1830</td>\n",
       "      <td>1830</td>\n",
       "      <td>1830</td>\n",
       "      <td>1830</td>\n",
       "      <td>1830</td>\n",
       "      <td>1830</td>\n",
       "      <td>1830</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             null_id+scaffold_id  null.id  scaffold.id  hg38.scaffold.coor  \\\n",
       "exp.id                                                                       \n",
       "single-null                 1830     1830         1830                1830   \n",
       "\n",
       "             hg38.str  scaffold.activity  seq_frag  200bp_insert  \\\n",
       "exp.id                                                             \n",
       "single-null      1830               1830      1830          1830   \n",
       "\n",
       "             200bp_insert+adaptors  gc_frac  \n",
       "exp.id                                       \n",
       "single-null                   1830     1830  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_gc = df.loc[df[\"gc_frac\"]>0.60]\n",
    "hi_gc.groupby(\"exp.id\").count()"
   ]
  }
 ],
 "metadata": {
  "gist": {
   "data": {
    "description": "nullomers/bin-lock/design_scaffold/4_library-insert-n-make-controls.ipynb",
    "public": false
   },
   "id": ""
  },
  "kernelspec": {
   "display_name": "Python (mamba)\n",
   "language": "python",
   "name": "mamba"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
