{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a4bf9332",
   "metadata": {},
   "source": [
    "Transform Guillermo's csv into a .bed file.\n",
    "\n",
    "- wide form -> long form table w microexon, upstream intron, and downstream intron intervals annotated\n",
    "\n",
    "Note:\n",
    "    ME coordinates are zero indexed\n",
    "    upIntStart\tupIntEnd\tdnIntStart\tdnIntEnd are one-indexed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a6bd65f8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:12:17.733943Z",
     "start_time": "2023-01-19T23:12:14.678171Z"
    }
   },
   "outputs": [],
   "source": [
    "import os, sys\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70a9d421",
   "metadata": {},
   "source": [
    "# Load config, dataframe "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "819660de",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:12:18.479805Z",
     "start_time": "2023-01-19T23:12:18.474071Z"
    }
   },
   "outputs": [],
   "source": [
    "# path to data\n",
    "PATH =\"/wynton/home/ahituv/fongsl/microexons/data\"\n",
    "\n",
    "# read\n",
    "F = os.path.join(PATH, \"human.hg38.total_detected.csv\")\n",
    "\n",
    "#write\n",
    "OUTF = os.path.join(PATH, \"human.hg38.total_detected.bed\")\n",
    "OUTFT = os.path.join(PATH, \"human.hg38.total_detected_truncated.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "80c457e3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:13:21.097600Z",
     "start_time": "2023-01-19T23:13:21.047863Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1514, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ME</th>\n",
       "      <th>event</th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>upIntStart</th>\n",
       "      <th>upIntEnd</th>\n",
       "      <th>dnIntStart</th>\n",
       "      <th>dnIntEnd</th>\n",
       "      <th>lengthDiff</th>\n",
       "      <th>mes3</th>\n",
       "      <th>mes5</th>\n",
       "      <th>conservation</th>\n",
       "      <th>neural_high_pred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_+_65353105_65353109</td>\n",
       "      <td>putative_microexon_000715026</td>\n",
       "      <td>DNAJC6</td>\n",
       "      <td>ENSG00000116675</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>65254650</td>\n",
       "      <td>65353105</td>\n",
       "      <td>65353110</td>\n",
       "      <td>65364634</td>\n",
       "      <td>4</td>\n",
       "      <td>8.867524</td>\n",
       "      <td>5.803994</td>\n",
       "      <td>0.097132</td>\n",
       "      <td>0.000222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6_-_109450989_109451014</td>\n",
       "      <td>putative_microexon_000716496</td>\n",
       "      <td>MICAL1</td>\n",
       "      <td>ENSG00000135596</td>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>109451015</td>\n",
       "      <td>109451599</td>\n",
       "      <td>109450086</td>\n",
       "      <td>109450989</td>\n",
       "      <td>25</td>\n",
       "      <td>1.687872</td>\n",
       "      <td>5.262086</td>\n",
       "      <td>0.096876</td>\n",
       "      <td>0.018178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10_-_37853628_37853645</td>\n",
       "      <td>putative_microexon_000716992</td>\n",
       "      <td>ZNF248</td>\n",
       "      <td>ENSG00000198105</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>37853646</td>\n",
       "      <td>37856295</td>\n",
       "      <td>37838112</td>\n",
       "      <td>37853628</td>\n",
       "      <td>17</td>\n",
       "      <td>2.265708</td>\n",
       "      <td>7.074537</td>\n",
       "      <td>0.096784</td>\n",
       "      <td>0.001780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX_-_85923678_85923681</td>\n",
       "      <td>putative_microexon_000769555</td>\n",
       "      <td>CHM</td>\n",
       "      <td>ENSG00000188419</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>85923682</td>\n",
       "      <td>85956152</td>\n",
       "      <td>85911339</td>\n",
       "      <td>85923678</td>\n",
       "      <td>3</td>\n",
       "      <td>0.918879</td>\n",
       "      <td>9.664478</td>\n",
       "      <td>0.087598</td>\n",
       "      <td>0.010956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11_+_13358220_13358225</td>\n",
       "      <td>putative_microexon_000771565</td>\n",
       "      <td>ARNTL</td>\n",
       "      <td>ENSG00000133794</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>13357097</td>\n",
       "      <td>13358220</td>\n",
       "      <td>13358226</td>\n",
       "      <td>13358433</td>\n",
       "      <td>5</td>\n",
       "      <td>5.090162</td>\n",
       "      <td>7.195993</td>\n",
       "      <td>0.087253</td>\n",
       "      <td>0.003300</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           ME                         event geneName  \\\n",
       "0    chr1_+_65353105_65353109  putative_microexon_000715026   DNAJC6   \n",
       "1  chr6_-_109450989_109451014  putative_microexon_000716496   MICAL1   \n",
       "2   chr10_-_37853628_37853645  putative_microexon_000716992   ZNF248   \n",
       "3    chrX_-_85923678_85923681  putative_microexon_000769555      CHM   \n",
       "4   chr11_+_13358220_13358225  putative_microexon_000771565    ARNTL   \n",
       "\n",
       "            geneID  chrom strand  upIntStart   upIntEnd  dnIntStart  \\\n",
       "0  ENSG00000116675   chr1      +    65254650   65353105    65353110   \n",
       "1  ENSG00000135596   chr6      -   109451015  109451599   109450086   \n",
       "2  ENSG00000198105  chr10      -    37853646   37856295    37838112   \n",
       "3  ENSG00000188419   chrX      -    85923682   85956152    85911339   \n",
       "4  ENSG00000133794  chr11      +    13357097   13358220    13358226   \n",
       "\n",
       "    dnIntEnd  lengthDiff      mes3      mes5  conservation  neural_high_pred  \n",
       "0   65364634           4  8.867524  5.803994      0.097132          0.000222  \n",
       "1  109450989          25  1.687872  5.262086      0.096876          0.018178  \n",
       "2   37853628          17  2.265708  7.074537      0.096784          0.001780  \n",
       "3   85923678           3  0.918879  9.664478      0.087598          0.010956  \n",
       "4   13358433           5  5.090162  7.195993      0.087253          0.003300  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# open dataframe\n",
    "df = pd.read_csv(F)\n",
    "\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "27d64963",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:13:21.337663Z",
     "start_time": "2023-01-19T23:13:21.329542Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1250"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# N unique genes w/ microexons\n",
    "len(df[\"geneName\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "04aeb522",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:13:21.963785Z",
     "start_time": "2023-01-19T23:13:21.673293Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1514"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# N unique genes w/ microexons\n",
    "len(df[\"event\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4eb06544",
   "metadata": {},
   "source": [
    "## str split ME column into multiple columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fc1d1195",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:13:22.407419Z",
     "start_time": "2023-01-19T23:13:22.367008Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ME</th>\n",
       "      <th>event</th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>upIntStart</th>\n",
       "      <th>upIntEnd</th>\n",
       "      <th>dnIntStart</th>\n",
       "      <th>dnIntEnd</th>\n",
       "      <th>lengthDiff</th>\n",
       "      <th>mes3</th>\n",
       "      <th>mes5</th>\n",
       "      <th>conservation</th>\n",
       "      <th>neural_high_pred</th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_+_65353105_65353109</td>\n",
       "      <td>putative_microexon_000715026</td>\n",
       "      <td>DNAJC6</td>\n",
       "      <td>ENSG00000116675</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>65254650</td>\n",
       "      <td>65353105</td>\n",
       "      <td>65353110</td>\n",
       "      <td>65364634</td>\n",
       "      <td>4</td>\n",
       "      <td>8.867524</td>\n",
       "      <td>5.803994</td>\n",
       "      <td>0.097132</td>\n",
       "      <td>0.000222</td>\n",
       "      <td>chr1</td>\n",
       "      <td>65353105</td>\n",
       "      <td>65353109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6_-_109450989_109451014</td>\n",
       "      <td>putative_microexon_000716496</td>\n",
       "      <td>MICAL1</td>\n",
       "      <td>ENSG00000135596</td>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>109451015</td>\n",
       "      <td>109451599</td>\n",
       "      <td>109450086</td>\n",
       "      <td>109450989</td>\n",
       "      <td>25</td>\n",
       "      <td>1.687872</td>\n",
       "      <td>5.262086</td>\n",
       "      <td>0.096876</td>\n",
       "      <td>0.018178</td>\n",
       "      <td>chr6</td>\n",
       "      <td>109450989</td>\n",
       "      <td>109451014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10_-_37853628_37853645</td>\n",
       "      <td>putative_microexon_000716992</td>\n",
       "      <td>ZNF248</td>\n",
       "      <td>ENSG00000198105</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>37853646</td>\n",
       "      <td>37856295</td>\n",
       "      <td>37838112</td>\n",
       "      <td>37853628</td>\n",
       "      <td>17</td>\n",
       "      <td>2.265708</td>\n",
       "      <td>7.074537</td>\n",
       "      <td>0.096784</td>\n",
       "      <td>0.001780</td>\n",
       "      <td>chr10</td>\n",
       "      <td>37853628</td>\n",
       "      <td>37853645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX_-_85923678_85923681</td>\n",
       "      <td>putative_microexon_000769555</td>\n",
       "      <td>CHM</td>\n",
       "      <td>ENSG00000188419</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>85923682</td>\n",
       "      <td>85956152</td>\n",
       "      <td>85911339</td>\n",
       "      <td>85923678</td>\n",
       "      <td>3</td>\n",
       "      <td>0.918879</td>\n",
       "      <td>9.664478</td>\n",
       "      <td>0.087598</td>\n",
       "      <td>0.010956</td>\n",
       "      <td>chrX</td>\n",
       "      <td>85923678</td>\n",
       "      <td>85923681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11_+_13358220_13358225</td>\n",
       "      <td>putative_microexon_000771565</td>\n",
       "      <td>ARNTL</td>\n",
       "      <td>ENSG00000133794</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>13357097</td>\n",
       "      <td>13358220</td>\n",
       "      <td>13358226</td>\n",
       "      <td>13358433</td>\n",
       "      <td>5</td>\n",
       "      <td>5.090162</td>\n",
       "      <td>7.195993</td>\n",
       "      <td>0.087253</td>\n",
       "      <td>0.003300</td>\n",
       "      <td>chr11</td>\n",
       "      <td>13358220</td>\n",
       "      <td>13358225</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           ME                         event geneName  \\\n",
       "0    chr1_+_65353105_65353109  putative_microexon_000715026   DNAJC6   \n",
       "1  chr6_-_109450989_109451014  putative_microexon_000716496   MICAL1   \n",
       "2   chr10_-_37853628_37853645  putative_microexon_000716992   ZNF248   \n",
       "3    chrX_-_85923678_85923681  putative_microexon_000769555      CHM   \n",
       "4   chr11_+_13358220_13358225  putative_microexon_000771565    ARNTL   \n",
       "\n",
       "            geneID  chrom strand  upIntStart   upIntEnd  dnIntStart  \\\n",
       "0  ENSG00000116675   chr1      +    65254650   65353105    65353110   \n",
       "1  ENSG00000135596   chr6      -   109451015  109451599   109450086   \n",
       "2  ENSG00000198105  chr10      -    37853646   37856295    37838112   \n",
       "3  ENSG00000188419   chrX      -    85923682   85956152    85911339   \n",
       "4  ENSG00000133794  chr11      +    13357097   13358220    13358226   \n",
       "\n",
       "    dnIntEnd  lengthDiff      mes3      mes5  conservation  neural_high_pred  \\\n",
       "0   65364634           4  8.867524  5.803994      0.097132          0.000222   \n",
       "1  109450989          25  1.687872  5.262086      0.096876          0.018178   \n",
       "2   37853628          17  2.265708  7.074537      0.096784          0.001780   \n",
       "3   85923678           3  0.918879  9.664478      0.087598          0.010956   \n",
       "4   13358433           5  5.090162  7.195993      0.087253          0.003300   \n",
       "\n",
       "    #chr      start        end  \n",
       "0   chr1   65353105   65353109  \n",
       "1   chr6  109450989  109451014  \n",
       "2  chr10   37853628   37853645  \n",
       "3   chrX   85923678   85923681  \n",
       "4  chr11   13358220   13358225  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#str split ME col into .bed features\n",
    "df[\"#chr\"] = df[\"ME\"].apply(lambda x: x.split(\"_\")[0])\n",
    "df[\"strand\"] = df[\"ME\"].apply(lambda x: x.split(\"_\")[1])\n",
    "df[\"start\"] = df[\"ME\"].apply(lambda x: x.split(\"_\")[2])\n",
    "df[\"end\"] = df[\"ME\"].apply(lambda x: x.split(\"_\")[3])\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db45d4b8",
   "metadata": {},
   "source": [
    "## filter intron starts to nearest 300bp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d8deed54",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:15:58.693034Z",
     "start_time": "2023-01-19T23:15:58.685775Z"
    }
   },
   "outputs": [],
   "source": [
    "def intronBoundaries(df, flank_len):\n",
    "    \"\"\"\n",
    "    create intron boundaries in dataframe as flanking length from exon start, stop\n",
    "    \n",
    "    input\n",
    "        df (pd.DataFrame) - dataframe w/ columns \"start\", \"end\" representing exon boundaries. \n",
    "        flank_len (int) - length to flank exon start, end to create introns\n",
    "        \n",
    "    method\n",
    "        Upstream intron start, end = exon start - flank_len, exon start - 1\n",
    "        Downstream intron start, end = exon end (remember, half-open start, fully closed end in bed files), exon end + flank_len\n",
    "        \n",
    "    return \n",
    "        df (pd.DataFrame) - original df w/ upstream, downstream introns start/end coordinates\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    df[\"upIntStart\"],df[\"upIntEnd\"] = df[\"start\"].map(int)-flank_len, df[\"start\"].map(int)\n",
    "    df[\"dnIntStart\"],df[\"dnIntEnd\"] = df[\"end\"].map(int), df[\"end\"].map(int)+flank_len\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3993666e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:16:00.450541Z",
     "start_time": "2023-01-19T23:16:00.418351Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ME</th>\n",
       "      <th>event</th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>upIntStart</th>\n",
       "      <th>upIntEnd</th>\n",
       "      <th>dnIntStart</th>\n",
       "      <th>dnIntEnd</th>\n",
       "      <th>lengthDiff</th>\n",
       "      <th>mes3</th>\n",
       "      <th>mes5</th>\n",
       "      <th>conservation</th>\n",
       "      <th>neural_high_pred</th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_+_65353105_65353109</td>\n",
       "      <td>putative_microexon_000715026</td>\n",
       "      <td>DNAJC6</td>\n",
       "      <td>ENSG00000116675</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>65352805</td>\n",
       "      <td>65353105</td>\n",
       "      <td>65353109</td>\n",
       "      <td>65353409</td>\n",
       "      <td>4</td>\n",
       "      <td>8.867524</td>\n",
       "      <td>5.803994</td>\n",
       "      <td>0.097132</td>\n",
       "      <td>0.000222</td>\n",
       "      <td>chr1</td>\n",
       "      <td>65353105</td>\n",
       "      <td>65353109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6_-_109450989_109451014</td>\n",
       "      <td>putative_microexon_000716496</td>\n",
       "      <td>MICAL1</td>\n",
       "      <td>ENSG00000135596</td>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>109450689</td>\n",
       "      <td>109450989</td>\n",
       "      <td>109451014</td>\n",
       "      <td>109451314</td>\n",
       "      <td>25</td>\n",
       "      <td>1.687872</td>\n",
       "      <td>5.262086</td>\n",
       "      <td>0.096876</td>\n",
       "      <td>0.018178</td>\n",
       "      <td>chr6</td>\n",
       "      <td>109450989</td>\n",
       "      <td>109451014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10_-_37853628_37853645</td>\n",
       "      <td>putative_microexon_000716992</td>\n",
       "      <td>ZNF248</td>\n",
       "      <td>ENSG00000198105</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>37853328</td>\n",
       "      <td>37853628</td>\n",
       "      <td>37853645</td>\n",
       "      <td>37853945</td>\n",
       "      <td>17</td>\n",
       "      <td>2.265708</td>\n",
       "      <td>7.074537</td>\n",
       "      <td>0.096784</td>\n",
       "      <td>0.001780</td>\n",
       "      <td>chr10</td>\n",
       "      <td>37853628</td>\n",
       "      <td>37853645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX_-_85923678_85923681</td>\n",
       "      <td>putative_microexon_000769555</td>\n",
       "      <td>CHM</td>\n",
       "      <td>ENSG00000188419</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>85923378</td>\n",
       "      <td>85923678</td>\n",
       "      <td>85923681</td>\n",
       "      <td>85923981</td>\n",
       "      <td>3</td>\n",
       "      <td>0.918879</td>\n",
       "      <td>9.664478</td>\n",
       "      <td>0.087598</td>\n",
       "      <td>0.010956</td>\n",
       "      <td>chrX</td>\n",
       "      <td>85923678</td>\n",
       "      <td>85923681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11_+_13358220_13358225</td>\n",
       "      <td>putative_microexon_000771565</td>\n",
       "      <td>ARNTL</td>\n",
       "      <td>ENSG00000133794</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>13357920</td>\n",
       "      <td>13358220</td>\n",
       "      <td>13358225</td>\n",
       "      <td>13358525</td>\n",
       "      <td>5</td>\n",
       "      <td>5.090162</td>\n",
       "      <td>7.195993</td>\n",
       "      <td>0.087253</td>\n",
       "      <td>0.003300</td>\n",
       "      <td>chr11</td>\n",
       "      <td>13358220</td>\n",
       "      <td>13358225</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           ME                         event geneName  \\\n",
       "0    chr1_+_65353105_65353109  putative_microexon_000715026   DNAJC6   \n",
       "1  chr6_-_109450989_109451014  putative_microexon_000716496   MICAL1   \n",
       "2   chr10_-_37853628_37853645  putative_microexon_000716992   ZNF248   \n",
       "3    chrX_-_85923678_85923681  putative_microexon_000769555      CHM   \n",
       "4   chr11_+_13358220_13358225  putative_microexon_000771565    ARNTL   \n",
       "\n",
       "            geneID  chrom strand  upIntStart   upIntEnd  dnIntStart  \\\n",
       "0  ENSG00000116675   chr1      +    65352805   65353105    65353109   \n",
       "1  ENSG00000135596   chr6      -   109450689  109450989   109451014   \n",
       "2  ENSG00000198105  chr10      -    37853328   37853628    37853645   \n",
       "3  ENSG00000188419   chrX      -    85923378   85923678    85923681   \n",
       "4  ENSG00000133794  chr11      +    13357920   13358220    13358225   \n",
       "\n",
       "    dnIntEnd  lengthDiff      mes3      mes5  conservation  neural_high_pred  \\\n",
       "0   65353409           4  8.867524  5.803994      0.097132          0.000222   \n",
       "1  109451314          25  1.687872  5.262086      0.096876          0.018178   \n",
       "2   37853945          17  2.265708  7.074537      0.096784          0.001780   \n",
       "3   85923981           3  0.918879  9.664478      0.087598          0.010956   \n",
       "4   13358525           5  5.090162  7.195993      0.087253          0.003300   \n",
       "\n",
       "    #chr      start        end  \n",
       "0   chr1   65353105   65353109  \n",
       "1   chr6  109450989  109451014  \n",
       "2  chr10   37853628   37853645  \n",
       "3   chrX   85923678   85923681  \n",
       "4  chr11   13358220   13358225  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FLANKLEN = 300\n",
    "df = intronBoundaries(df, FLANKLEN)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a200e3eb",
   "metadata": {},
   "source": [
    "# Make long-form DFs for each data type \n",
    " - microexons (ME)\n",
    " - downstream intron (ds)\n",
    " - upstream intron (up)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e51d7624",
   "metadata": {},
   "source": [
    "## functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e65b0e32",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:16:05.703441Z",
     "start_time": "2023-01-19T23:16:05.692549Z"
    }
   },
   "outputs": [],
   "source": [
    "def header_names(prefix):\n",
    "    \"\"\"\n",
    "    get prefix-specific column header list\n",
    "    \n",
    "    input\n",
    "        prefix (str) - key for column header set to return\n",
    "        \n",
    "    method\n",
    "        1. make dictionary of prefix (str): column header (list)\n",
    "        2. select prefix-specific header\n",
    "        \n",
    "    return \n",
    "        header (list) - list of column header names corresponding to prefix\n",
    "    \"\"\"\n",
    "    \n",
    "    #1\n",
    "    headerdict = {\n",
    "        None:[\"#chr\", \"start\", \"end\", \n",
    "             \"event\", \"geneName\", \"geneID\", \n",
    "             \"strand\", \"lengthDiff\",\"mes3\", \"mes5\", \n",
    "             \"conservation\", \"neural_high_pred\"], \n",
    "        'dn': [\"#chr\", \"dnIntStart\", \"dnIntEnd\", \n",
    "             \"event\", \"geneName\", \"geneID\", \n",
    "             \"strand\", \"lengthDiff\",\"mes3\", \"mes5\", \n",
    "             \"conservation\", \"neural_high_pred\"],\n",
    "        'up':[\"#chr\", \"upIntStart\", \"upIntEnd\", \n",
    "             \"event\", \"geneName\", \"geneID\", \n",
    "             \"strand\", \"lengthDiff\",\"mes3\", \"mes5\", \n",
    "             \"conservation\", \"neural_high_pred\"]\n",
    "    }\n",
    "    \n",
    "    #2\n",
    "    header = headerdict[prefix]\n",
    "\n",
    "    return header\n",
    "\n",
    "\n",
    "def prefixSpecificDf(header_list, prefix, label, df):\n",
    "    \n",
    "    \"\"\"\n",
    "    subset dataframe by prefix header list, label\n",
    "    \n",
    "    input\n",
    "        header_list (list) - list of header columns to include in dataframe\n",
    "        label (str) - name for label columns\n",
    "        df (pd dataframe object) - full pandas dataframe\n",
    "        \n",
    "    method\n",
    "        1. subset dataframe by header_list, drop duplicates\n",
    "        2. add label column\n",
    "        3. if dealing w/ intronic regions\n",
    "            3.1 rename start and end columns\n",
    "            3.2 correct: 1-index -> 0-index\n",
    "                for more info on 0 v 1 indexing - https://genome-blog.soe.ucsc.edu/blog/2016/12/12/the-ucsc-genome-browser-coordinate-counting-systems/\n",
    "                \n",
    "    return \n",
    "        subset_df (pd dataframe object) - dataframe of subsetted data w/ label column, corrected 0-index (if intron)\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    #1 subset dataframe\n",
    "    subset_df = df[header_list].copy().drop_duplicates()\n",
    "\n",
    "    #2 add label\n",
    "    subset_df[\"label\"] = label\n",
    "\n",
    "    if prefix is not None:\n",
    "    \n",
    "        #3.1 rename columns\n",
    "        subset_df = subset_df.rename(columns = {\n",
    "                                                f'{prefix}IntStart': \"start\",\n",
    "                                                f\"{prefix}IntEnd\": \"end\"\n",
    "                                                })\n",
    "        #3.2 correct 1-index\n",
    "        #subset_df[\"start\"] = subset_df[\"start\"]-1 \n",
    "\n",
    "\n",
    "    return subset_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2bb3673",
   "metadata": {},
   "source": [
    "## prefixes for longform dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b0d45b85",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:16:06.524908Z",
     "start_time": "2023-01-19T23:16:06.521584Z"
    }
   },
   "outputs": [],
   "source": [
    "prefixes = [(None, \"microexon\"), \n",
    "            (\"dn\", \"dsINT\"),\n",
    "            (\"up\", \"upINT\")\n",
    "           ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94eccb5e",
   "metadata": {},
   "source": [
    "## subset dataframes to make long form"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "73ee4552",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:16:07.663321Z",
     "start_time": "2023-01-19T23:16:07.629938Z"
    }
   },
   "outputs": [],
   "source": [
    "collect_dfs = {} # to collect subset_dfs\n",
    "\n",
    "for prefix, label in prefixes:\n",
    "    \n",
    "    header_list = header_names(prefix) # get headers\n",
    "    \n",
    "    subset_df = prefixSpecificDf(header_list, prefix, label, df) # get dataframe subset\n",
    "\n",
    "    collect_dfs[label] = subset_df  # add subset df to dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "870d25e5",
   "metadata": {},
   "source": [
    "## write long-form, concatenated bed file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6bfd7e55",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:16:09.730822Z",
     "start_time": "2023-01-19T23:16:09.700340Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4542, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>event</th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneID</th>\n",
       "      <th>strand</th>\n",
       "      <th>lengthDiff</th>\n",
       "      <th>mes3</th>\n",
       "      <th>mes5</th>\n",
       "      <th>conservation</th>\n",
       "      <th>neural_high_pred</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65353105</td>\n",
       "      <td>65353109</td>\n",
       "      <td>putative_microexon_000715026</td>\n",
       "      <td>DNAJC6</td>\n",
       "      <td>ENSG00000116675</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>8.867524</td>\n",
       "      <td>5.803994</td>\n",
       "      <td>0.097132</td>\n",
       "      <td>0.000222</td>\n",
       "      <td>microexon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>109450989</td>\n",
       "      <td>109451014</td>\n",
       "      <td>putative_microexon_000716496</td>\n",
       "      <td>MICAL1</td>\n",
       "      <td>ENSG00000135596</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>1.687872</td>\n",
       "      <td>5.262086</td>\n",
       "      <td>0.096876</td>\n",
       "      <td>0.018178</td>\n",
       "      <td>microexon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>37853628</td>\n",
       "      <td>37853645</td>\n",
       "      <td>putative_microexon_000716992</td>\n",
       "      <td>ZNF248</td>\n",
       "      <td>ENSG00000198105</td>\n",
       "      <td>-</td>\n",
       "      <td>17</td>\n",
       "      <td>2.265708</td>\n",
       "      <td>7.074537</td>\n",
       "      <td>0.096784</td>\n",
       "      <td>0.001780</td>\n",
       "      <td>microexon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>85923678</td>\n",
       "      <td>85923681</td>\n",
       "      <td>putative_microexon_000769555</td>\n",
       "      <td>CHM</td>\n",
       "      <td>ENSG00000188419</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "      <td>0.918879</td>\n",
       "      <td>9.664478</td>\n",
       "      <td>0.087598</td>\n",
       "      <td>0.010956</td>\n",
       "      <td>microexon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>13358220</td>\n",
       "      <td>13358225</td>\n",
       "      <td>putative_microexon_000771565</td>\n",
       "      <td>ARNTL</td>\n",
       "      <td>ENSG00000133794</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>5.090162</td>\n",
       "      <td>7.195993</td>\n",
       "      <td>0.087253</td>\n",
       "      <td>0.003300</td>\n",
       "      <td>microexon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1509</th>\n",
       "      <td>chr2</td>\n",
       "      <td>62751326</td>\n",
       "      <td>62751626</td>\n",
       "      <td>putative_microexon_000645392</td>\n",
       "      <td>EHBP1</td>\n",
       "      <td>ENSG00000115504</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>1.562151</td>\n",
       "      <td>8.722870</td>\n",
       "      <td>0.110759</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>upINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1510</th>\n",
       "      <td>chr15</td>\n",
       "      <td>77361014</td>\n",
       "      <td>77361314</td>\n",
       "      <td>putative_microexon_000645748</td>\n",
       "      <td>PEAK1</td>\n",
       "      <td>ENSG00000173517</td>\n",
       "      <td>-</td>\n",
       "      <td>24</td>\n",
       "      <td>0.306288</td>\n",
       "      <td>9.801806</td>\n",
       "      <td>0.110682</td>\n",
       "      <td>0.000380</td>\n",
       "      <td>upINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1511</th>\n",
       "      <td>chr17</td>\n",
       "      <td>1825043</td>\n",
       "      <td>1825343</td>\n",
       "      <td>putative_microexon_000646316</td>\n",
       "      <td>SMYD4</td>\n",
       "      <td>ENSG00000186532</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "      <td>0.373406</td>\n",
       "      <td>5.958177</td>\n",
       "      <td>0.110557</td>\n",
       "      <td>0.305077</td>\n",
       "      <td>upINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1512</th>\n",
       "      <td>chr15</td>\n",
       "      <td>59351693</td>\n",
       "      <td>59351993</td>\n",
       "      <td>putative_microexon_000646965</td>\n",
       "      <td>MYO1E</td>\n",
       "      <td>ENSG00000157483</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "      <td>2.181511</td>\n",
       "      <td>6.646370</td>\n",
       "      <td>0.110431</td>\n",
       "      <td>0.000300</td>\n",
       "      <td>upINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1513</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188334881</td>\n",
       "      <td>188335181</td>\n",
       "      <td>putative_microexon_000621067</td>\n",
       "      <td>LPP</td>\n",
       "      <td>ENSG00000145012</td>\n",
       "      <td>+</td>\n",
       "      <td>21</td>\n",
       "      <td>2.491580</td>\n",
       "      <td>9.787718</td>\n",
       "      <td>0.115989</td>\n",
       "      <td>0.000378</td>\n",
       "      <td>upINT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4542 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #chr      start        end                         event geneName  \\\n",
       "0      chr1   65353105   65353109  putative_microexon_000715026   DNAJC6   \n",
       "1      chr6  109450989  109451014  putative_microexon_000716496   MICAL1   \n",
       "2     chr10   37853628   37853645  putative_microexon_000716992   ZNF248   \n",
       "3      chrX   85923678   85923681  putative_microexon_000769555      CHM   \n",
       "4     chr11   13358220   13358225  putative_microexon_000771565    ARNTL   \n",
       "...     ...        ...        ...                           ...      ...   \n",
       "1509   chr2   62751326   62751626  putative_microexon_000645392    EHBP1   \n",
       "1510  chr15   77361014   77361314  putative_microexon_000645748    PEAK1   \n",
       "1511  chr17    1825043    1825343  putative_microexon_000646316    SMYD4   \n",
       "1512  chr15   59351693   59351993  putative_microexon_000646965    MYO1E   \n",
       "1513   chr3  188334881  188335181  putative_microexon_000621067      LPP   \n",
       "\n",
       "               geneID strand  lengthDiff      mes3      mes5  conservation  \\\n",
       "0     ENSG00000116675      +           4  8.867524  5.803994      0.097132   \n",
       "1     ENSG00000135596      -          25  1.687872  5.262086      0.096876   \n",
       "2     ENSG00000198105      -          17  2.265708  7.074537      0.096784   \n",
       "3     ENSG00000188419      -           3  0.918879  9.664478      0.087598   \n",
       "4     ENSG00000133794      +           5  5.090162  7.195993      0.087253   \n",
       "...               ...    ...         ...       ...       ...           ...   \n",
       "1509  ENSG00000115504      +           4  1.562151  8.722870      0.110759   \n",
       "1510  ENSG00000173517      -          24  0.306288  9.801806      0.110682   \n",
       "1511  ENSG00000186532      -           4  0.373406  5.958177      0.110557   \n",
       "1512  ENSG00000157483      -           3  2.181511  6.646370      0.110431   \n",
       "1513  ENSG00000145012      +          21  2.491580  9.787718      0.115989   \n",
       "\n",
       "      neural_high_pred      label  \n",
       "0             0.000222  microexon  \n",
       "1             0.018178  microexon  \n",
       "2             0.001780  microexon  \n",
       "3             0.010956  microexon  \n",
       "4             0.003300  microexon  \n",
       "...                ...        ...  \n",
       "1509          0.000689      upINT  \n",
       "1510          0.000380      upINT  \n",
       "1511          0.305077      upINT  \n",
       "1512          0.000300      upINT  \n",
       "1513          0.000378      upINT  \n",
       "\n",
       "[4542 rows x 13 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# concatenate microexon, upintron, downint dataframes\n",
    "out = pd.concat(collect_dfs.values())\n",
    "print(out.shape)\n",
    "\n",
    "# show\n",
    "out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "c9ff2597",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:16:11.775167Z",
     "start_time": "2023-01-19T23:16:11.749096Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>event</th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneID</th>\n",
       "      <th>strand</th>\n",
       "      <th>lengthDiff</th>\n",
       "      <th>mes3</th>\n",
       "      <th>mes5</th>\n",
       "      <th>conservation</th>\n",
       "      <th>neural_high_pred</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1170</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188335181</td>\n",
       "      <td>188335197</td>\n",
       "      <td>putative_microexon_000631360</td>\n",
       "      <td>LPP</td>\n",
       "      <td>ENSG00000145012</td>\n",
       "      <td>+</td>\n",
       "      <td>16</td>\n",
       "      <td>2.49158</td>\n",
       "      <td>10.284364</td>\n",
       "      <td>0.113750</td>\n",
       "      <td>0.000544</td>\n",
       "      <td>microexon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1513</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188335181</td>\n",
       "      <td>188335202</td>\n",
       "      <td>putative_microexon_000621067</td>\n",
       "      <td>LPP</td>\n",
       "      <td>ENSG00000145012</td>\n",
       "      <td>+</td>\n",
       "      <td>21</td>\n",
       "      <td>2.49158</td>\n",
       "      <td>9.787718</td>\n",
       "      <td>0.115989</td>\n",
       "      <td>0.000378</td>\n",
       "      <td>microexon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1170</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188335197</td>\n",
       "      <td>188335497</td>\n",
       "      <td>putative_microexon_000631360</td>\n",
       "      <td>LPP</td>\n",
       "      <td>ENSG00000145012</td>\n",
       "      <td>+</td>\n",
       "      <td>16</td>\n",
       "      <td>2.49158</td>\n",
       "      <td>10.284364</td>\n",
       "      <td>0.113750</td>\n",
       "      <td>0.000544</td>\n",
       "      <td>dsINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1513</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188335202</td>\n",
       "      <td>188335502</td>\n",
       "      <td>putative_microexon_000621067</td>\n",
       "      <td>LPP</td>\n",
       "      <td>ENSG00000145012</td>\n",
       "      <td>+</td>\n",
       "      <td>21</td>\n",
       "      <td>2.49158</td>\n",
       "      <td>9.787718</td>\n",
       "      <td>0.115989</td>\n",
       "      <td>0.000378</td>\n",
       "      <td>dsINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1170</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188334881</td>\n",
       "      <td>188335181</td>\n",
       "      <td>putative_microexon_000631360</td>\n",
       "      <td>LPP</td>\n",
       "      <td>ENSG00000145012</td>\n",
       "      <td>+</td>\n",
       "      <td>16</td>\n",
       "      <td>2.49158</td>\n",
       "      <td>10.284364</td>\n",
       "      <td>0.113750</td>\n",
       "      <td>0.000544</td>\n",
       "      <td>upINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1513</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188334881</td>\n",
       "      <td>188335181</td>\n",
       "      <td>putative_microexon_000621067</td>\n",
       "      <td>LPP</td>\n",
       "      <td>ENSG00000145012</td>\n",
       "      <td>+</td>\n",
       "      <td>21</td>\n",
       "      <td>2.49158</td>\n",
       "      <td>9.787718</td>\n",
       "      <td>0.115989</td>\n",
       "      <td>0.000378</td>\n",
       "      <td>upINT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      #chr      start        end                         event geneName  \\\n",
       "1170  chr3  188335181  188335197  putative_microexon_000631360      LPP   \n",
       "1513  chr3  188335181  188335202  putative_microexon_000621067      LPP   \n",
       "1170  chr3  188335197  188335497  putative_microexon_000631360      LPP   \n",
       "1513  chr3  188335202  188335502  putative_microexon_000621067      LPP   \n",
       "1170  chr3  188334881  188335181  putative_microexon_000631360      LPP   \n",
       "1513  chr3  188334881  188335181  putative_microexon_000621067      LPP   \n",
       "\n",
       "               geneID strand  lengthDiff     mes3       mes5  conservation  \\\n",
       "1170  ENSG00000145012      +          16  2.49158  10.284364      0.113750   \n",
       "1513  ENSG00000145012      +          21  2.49158   9.787718      0.115989   \n",
       "1170  ENSG00000145012      +          16  2.49158  10.284364      0.113750   \n",
       "1513  ENSG00000145012      +          21  2.49158   9.787718      0.115989   \n",
       "1170  ENSG00000145012      +          16  2.49158  10.284364      0.113750   \n",
       "1513  ENSG00000145012      +          21  2.49158   9.787718      0.115989   \n",
       "\n",
       "      neural_high_pred      label  \n",
       "1170          0.000544  microexon  \n",
       "1513          0.000378  microexon  \n",
       "1170          0.000544      dsINT  \n",
       "1513          0.000378      dsINT  \n",
       "1170          0.000544      upINT  \n",
       "1513          0.000378      upINT  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out.loc[out[\"geneName\"]==\"LPP\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "8c1dc2ed",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:16:21.131008Z",
     "start_time": "2023-01-19T23:16:21.049339Z"
    }
   },
   "outputs": [],
   "source": [
    "# write out file\n",
    "out.to_csv(OUTF, sep ='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "339c219c",
   "metadata": {},
   "source": [
    "## write truncated long form bed file (just bed fields, gene name, and label) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "cf809e84",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-19T23:16:38.571977Z",
     "start_time": "2023-01-19T23:16:38.513612Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>geneName</th>\n",
       "      <th>label</th>\n",
       "      <th>event</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1170</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188335181</td>\n",
       "      <td>188335197</td>\n",
       "      <td>LPP</td>\n",
       "      <td>microexon</td>\n",
       "      <td>putative_microexon_000631360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1513</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188335181</td>\n",
       "      <td>188335202</td>\n",
       "      <td>LPP</td>\n",
       "      <td>microexon</td>\n",
       "      <td>putative_microexon_000621067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1170</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188335197</td>\n",
       "      <td>188335497</td>\n",
       "      <td>LPP</td>\n",
       "      <td>dsINT</td>\n",
       "      <td>putative_microexon_000631360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1513</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188335202</td>\n",
       "      <td>188335502</td>\n",
       "      <td>LPP</td>\n",
       "      <td>dsINT</td>\n",
       "      <td>putative_microexon_000621067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1170</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188334881</td>\n",
       "      <td>188335181</td>\n",
       "      <td>LPP</td>\n",
       "      <td>upINT</td>\n",
       "      <td>putative_microexon_000631360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1513</th>\n",
       "      <td>chr3</td>\n",
       "      <td>188334881</td>\n",
       "      <td>188335181</td>\n",
       "      <td>LPP</td>\n",
       "      <td>upINT</td>\n",
       "      <td>putative_microexon_000621067</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      #chr      start        end geneName      label  \\\n",
       "1170  chr3  188335181  188335197      LPP  microexon   \n",
       "1513  chr3  188335181  188335202      LPP  microexon   \n",
       "1170  chr3  188335197  188335497      LPP      dsINT   \n",
       "1513  chr3  188335202  188335502      LPP      dsINT   \n",
       "1170  chr3  188334881  188335181      LPP      upINT   \n",
       "1513  chr3  188334881  188335181      LPP      upINT   \n",
       "\n",
       "                             event  \n",
       "1170  putative_microexon_000631360  \n",
       "1513  putative_microexon_000621067  \n",
       "1170  putative_microexon_000631360  \n",
       "1513  putative_microexon_000621067  \n",
       "1170  putative_microexon_000631360  \n",
       "1513  putative_microexon_000621067  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out[[\"#chr\", \"start\", \"end\", \"geneName\", \"label\", \"event\"]].drop_duplicates().to_csv(OUTFT, sep ='\\t', index=False)\n",
    "\n",
    "# show\n",
    "out[[\"#chr\", \"start\", \"end\", \"geneName\", \"label\", \"event\"]].loc[out[\"geneName\"]==\"LPP\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5afbfbbd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (wynenv)",
   "language": "python",
   "name": "wynenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
