{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a38fcfaa-77c1-4bfb-abe0-05dbc4e4e87d",
   "metadata": {},
   "source": [
    "20240227\n",
    "\n",
    "SarahFong\n",
    "\n",
    "Take output from DiffBind summarized peak and count matrix and prepare data for \n",
    "1. Clean diffbind normalized count data (remove bad coordinates)\n",
    "2. perform PCA on ATAC read count data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "24791ba6-ba9e-47c8-99ff-824a21c53b9b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:16:17.634060Z",
     "start_time": "2024-03-10T03:16:12.058366Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('sans-serif', 'Arial', 18)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "import glob\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib_venn import venn2\n",
    "import numpy as np\n",
    "import os, sys\n",
    "import pandas as pd\n",
    "import pybedtools as pbt\n",
    "\n",
    "from scipy import stats\n",
    "from scipy.stats import zscore, boxcox\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import plot_params as pp\n",
    "pp.fonts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de4c9f36-8609-4502-8a4c-3279a57f5c77",
   "metadata": {},
   "source": [
    "# load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "15b0b616-7254-4b4e-be68-8d745f4a629f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:16:17.642584Z",
     "start_time": "2024-03-10T03:16:17.636417Z"
    }
   },
   "outputs": [],
   "source": [
    "PATH = \"/wynton/group/ahituv/data/US-MPRA/ATAC-seq/Diffbind_results\"\n",
    "os.chdir(PATH)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c13717b-4462-468b-aded-c85fa8632585",
   "metadata": {},
   "source": [
    "# functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b1a306a-6a6f-4c5f-8c0c-ed8d27dc1153",
   "metadata": {},
   "source": [
    "## clean bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f1e5e3b1-98ce-4ba9-b573-4213e3da74d0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:16:17.793658Z",
     "start_time": "2024-03-10T03:16:17.644278Z"
    }
   },
   "outputs": [],
   "source": [
    "def cleanBed(file):\n",
    "\n",
    "    # name of the outfile\n",
    "    outfile = \".\".join(file.split('.')[:-1]) + \".cleaned.bed\"\n",
    "\n",
    "    # python command for cleaning bed file\n",
    "    cmd = ' '.join([\n",
    "        \"python\", \n",
    "        \"/wynton/home/ahituv/fongsl/tools/genome/bed_clean.py\",  # custom script that removes bad coordinates (start>end, start<0)\n",
    "        file\n",
    "    ])\n",
    "\n",
    "    if os.path.exists(outfile) is False:  # if not already cleaned\n",
    "        os.system(cmd)  # execute command\n",
    "    else:\n",
    "        print(\"already made?\", outfile)\n",
    "\n",
    "    return outfile"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61b32da8-c08f-4840-ac7a-df3175679ae8",
   "metadata": {},
   "source": [
    "## compute replicate mean, peak centers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44dc60b8-00a8-4904-b9a9-30757fca4446",
   "metadata": {},
   "source": [
    "## count pbt obj lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cd9e3e4d-05c1-495e-86b9-2a20c7a32d31",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:16:17.929253Z",
     "start_time": "2024-03-10T03:16:17.796735Z"
    }
   },
   "outputs": [],
   "source": [
    "def pbtCounter(pbt_obj):\n",
    "    \"\"\"count lines in pybedtool object, return count\"\"\"\n",
    "    \n",
    "    v = 0\n",
    "    for i in pbt_obj:\n",
    "        v+=1\n",
    "    print(v)\n",
    "    \n",
    "    return v"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf8531bd-419d-4963-bfbe-fbc04d15a81a",
   "metadata": {},
   "source": [
    "## label elements w/ names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "70c5993e-b492-48ac-b27e-423df1af4283",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:16:18.062527Z",
     "start_time": "2024-03-10T03:16:17.931670Z"
    }
   },
   "outputs": [],
   "source": [
    "def labelDf(df, label_name, label_value):\n",
    "    \"\"\"add an indexed label to dataframe\"\"\"\n",
    "    df[\"label\"]= label_value\n",
    "    df[label_name]=label_value + \".\" + df.index.map(str)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fe8539a-2cd0-41d2-995f-71f99bb73ffb",
   "metadata": {},
   "source": [
    "## make fa into pd.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "af260271-5a56-4e0a-84fb-fe3b4de03521",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:16:18.143573Z",
     "start_time": "2024-03-10T03:16:18.064964Z"
    }
   },
   "outputs": [],
   "source": [
    "def fastaDf(file):\n",
    "    \"\"\" turn fasta file into pd dataframe\"\"\"\n",
    "\n",
    "    fasta_df = {}\n",
    "    print(file)\n",
    "    \n",
    "    # parse open fasta file\n",
    "    with open(file, \"r\") as reader:\n",
    "        for value in SimpleFastaParser(reader):\n",
    "            seqid, seq = value\n",
    "            fasta_df[seqid] = seq\n",
    "            \n",
    "    # make dataframe\n",
    "    df = pd.DataFrame(fasta_df.items())\n",
    "\n",
    "    # name columns\n",
    "    df.columns = [\"seq.id\", \"seq\"]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ba6c7d3-4107-4315-8748-e13b1108310b",
   "metadata": {},
   "source": [
    "## make chromosome list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "947fbb71-5880-4453-8414-51db88678771",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:16:18.211959Z",
     "start_time": "2024-03-10T03:16:18.146096Z"
    }
   },
   "outputs": [],
   "source": [
    "def chrList():\n",
    "    \"\"\"return  list of chromosomes\"\"\"\n",
    "    \n",
    "    chrs = []\n",
    "    \n",
    "    for n in np.arange(1,23):\n",
    "        chrs.append(f\"chr{n}\")\n",
    "    \n",
    "    # add sex chromosomes\n",
    "    chrs.append(\"chrX\")\n",
    "    chrs.append(\"chrY\")\n",
    "    \n",
    "    return chrs\n",
    "\n",
    "def filterConc(df, min_conc, col1, col2):\n",
    "    print(\"before filter shape\", df.shape)\n",
    "    filtered = df.loc[(df[col1]>min_conc)| \n",
    "                      (df[col2]>min_conc) \n",
    "                     ].drop_duplicates().copy()\n",
    "    \n",
    "    print(\"after filter shape\", filtered.shape)\n",
    "    return filtered\n",
    "\n",
    "def zscore(df, col):\n",
    "    \"\"\"compute z-score for a column\"\"\"\n",
    "    df[f'{col}.zscore'] = (df[col] - df[col].mean())/df[col].std()\n",
    "    \n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8317e94b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:16:18.311924Z",
     "start_time": "2024-03-10T03:16:18.214964Z"
    }
   },
   "outputs": [],
   "source": [
    "def plotHist(df, fdf, CL):\n",
    "    fig, ax= plt.subplots(figsize=(6,6))\n",
    "    sns.histplot(x=\"Conc_Ultrasound\", data=df, label=\"unfiltered\")\n",
    "    sns.histplot(x=\"Conc_Ultrasound\", data=fdf, label=\"filtered\")\n",
    "    ax.set(title=CL)\n",
    "    ax.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5b40e489",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:18:08.570906Z",
     "start_time": "2024-03-10T03:18:08.563734Z"
    }
   },
   "outputs": [],
   "source": [
    "def encodeIntersection(cl, df, dif_bed, fdf):\n",
    "    \n",
    "    ENCODE_CCREs = {'hepg2': \"/wynton/group/ahituv/data/encode/hepg2\",\n",
    "                    \"k562\": \"/wynton/group/ahituv/data/encode/k562\",\n",
    "                    \"hob\": \"/wynton/group/ahituv/data/encode/MG63\",  # imperfect match, human osteosarcoma - https://www.sigmaaldrich.com/US/en/product/sigma/cb_86051601?utm_source=google&utm_medium=cpc&utm_campaign=8906396346&utm_content=88453247983&gclid=Cj0KCQiArrCvBhCNARIsAOkAGcVymGqSXznvXF-1hmqzVx8mK0z7eopHlS356QdcjYOyInUuoAHrmHEaAvWHEALw_wcB\n",
    "                    \"bj\": \"/wynton/group/ahituv/data/encode/bj\"\n",
    "                    }\n",
    "\n",
    "    ENCODE_PATH = ENCODE_CCREs[cl]\n",
    "    annots = glob.glob(os.path.join(ENCODE_PATH, \"*.bed\"))\n",
    "\n",
    "    b = pbt.BedTool.from_dataframe(df)\n",
    "    d = pbt.BedTool(dif_bed)\n",
    "    f = pbt.BedTool.from_dataframe(fdf)\n",
    "    dxf = d.intersect(f, wa=True)  # evaluate the filtered diff accessible elements\n",
    "    \n",
    "    df_count, difacc_count, fdf_count, fdifacc_count = pbtCounter(\n",
    "        b), pbtCounter(d), pbtCounter(f), pbtCounter(dxf)\n",
    "\n",
    "    reg_annots = [\"CTCF-only,CTCF-bound\",\n",
    "                  \"dELS\",\n",
    "                  \"PLS,CTCF-bound\",\n",
    "                  \"pELS\",\n",
    "                  \"PLS\", \"pELS,CTCF-bound\",\n",
    "                  \"dELS\", \"CTCF-bound\"]\n",
    "    \n",
    "    results = {}\n",
    "    \n",
    "    results[\"totals\"] = [cl, \"total.atac\", None, df_count, fdf_count, difacc_count, fdifacc_count]\n",
    "    \n",
    "    for a in annots:\n",
    "        \n",
    "        annot = ((a.split(\"/\")[-1]).split(\".\")[0])\n",
    "        \n",
    "        if annot in reg_annots: # per annotation\n",
    "            print(\"\\n\\n\", cl,  annot)\n",
    "            encode_bed = pbt.BedTool(a)\n",
    "            encode_count = pbtCounter(encode_bed)\n",
    "            \n",
    "            result_count_vector= [cl, annot, encode_count]\n",
    "\n",
    "            for i in [b, f, d, dxf]:  # per dataset\n",
    "                ixe = i.intersect(encode_bed, wa=True)\n",
    "                result_count_vector.append(pbtCounter(ixe))\n",
    "\n",
    "            \n",
    "            results[annot] = result_count_vector\n",
    "            \n",
    "            \n",
    "    colnames=[\"cl\", \"annotation\", \"N_cCRE\", \"N_allAcc\", \"N_filterAcc\", \"N_diffAcc\", \"N_filterDiffAcc\"]\n",
    "    return pd.DataFrame(results.values(), columns = colnames)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6144f3a-643a-42f5-bacc-40d77de7f899",
   "metadata": {},
   "source": [
    "# Main"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e6e23d8-f383-4567-a7bb-ad49825c524b",
   "metadata": {},
   "source": [
    "## clean bed files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cdcf5c44",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:18:38.614031Z",
     "start_time": "2024-03-10T03:18:09.665863Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hepg2\n",
      "before filter shape (127857, 11)\n",
      "after filter shape (49249, 11)\n",
      "127857\n",
      "1916\n",
      "49249\n",
      "777\n",
      "\n",
      "\n",
      " hepg2 CTCF-only,CTCF-bound\n",
      "20944\n",
      "12332\n",
      "6072\n",
      "194\n",
      "140\n",
      "\n",
      "\n",
      " hepg2 dELS\n",
      "25435\n",
      "15164\n",
      "11296\n",
      "196\n",
      "149\n",
      "\n",
      "\n",
      " hepg2 PLS,CTCF-bound\n",
      "2377\n",
      "2017\n",
      "1972\n",
      "12\n",
      "12\n",
      "\n",
      "\n",
      " hepg2 pELS\n",
      "16656\n",
      "5931\n",
      "5331\n",
      "50\n",
      "42\n",
      "\n",
      "\n",
      " hepg2 PLS\n",
      "13597\n",
      "10324\n",
      "10085\n",
      "58\n",
      "56\n",
      "\n",
      "\n",
      " hepg2 pELS,CTCF-bound\n",
      "2613\n",
      "1226\n",
      "1159\n",
      "7\n",
      "6\n",
      "k562\n",
      "before filter shape (102300, 11)\n",
      "after filter shape (82505, 11)\n",
      "102300\n",
      "890\n",
      "82505\n",
      "655\n",
      "\n",
      "\n",
      " k562 pELS,CTCF-bound\n",
      "6580\n",
      "2207\n",
      "2206\n",
      "1\n",
      "0\n",
      "\n",
      "\n",
      " k562 CTCF-only,CTCF-bound\n",
      "27649\n",
      "19741\n",
      "18179\n",
      "130\n",
      "116\n",
      "\n",
      "\n",
      " k562 PLS,CTCF-bound\n",
      "6568\n",
      "5146\n",
      "5144\n",
      "10\n",
      "10\n",
      "\n",
      "\n",
      " k562 pELS\n",
      "33276\n",
      "6502\n",
      "6461\n",
      "31\n",
      "31\n",
      "\n",
      "\n",
      " k562 PLS\n",
      "12752\n",
      "7926\n",
      "7920\n",
      "20\n",
      "19\n",
      "\n",
      "\n",
      " k562 dELS\n",
      "30119\n",
      "16034\n",
      "15794\n",
      "154\n",
      "150\n",
      "hob\n",
      "before filter shape (159246, 11)\n",
      "after filter shape (124148, 11)\n",
      "159246\n",
      "2324\n",
      "124148\n",
      "2218\n",
      "\n",
      "\n",
      " hob PLS\n",
      "18798\n",
      "13636\n",
      "13560\n",
      "567\n",
      "567\n",
      "bj\n",
      "before filter shape (117245, 11)\n",
      "after filter shape (67339, 11)\n",
      "117245\n",
      "433\n",
      "67339\n",
      "242\n",
      "\n",
      "\n",
      " bj PLS,CTCF-bound\n",
      "3835\n",
      "3345\n",
      "3330\n",
      "12\n",
      "12\n",
      "\n",
      "\n",
      " bj PLS\n",
      "9445\n",
      "7770\n",
      "7680\n",
      "23\n",
      "23\n"
     ]
    }
   ],
   "source": [
    "# params\n",
    "MIN_CONC= 3\n",
    "CLS =['hepg2', 'k562', \"hob\", \"bj\"]\n",
    "\n",
    "cl_results = {}\n",
    "for CL in CLS:\n",
    "    print(CL)\n",
    "\n",
    "    ALL = f\"{CL}_deseq2-nondiff.csv\"\n",
    "    ALL_BED = ALL.strip(\".csv\") + \".bed\"\n",
    "\n",
    "    # DESEQ2 diff acc\n",
    "    DIFF = f\"{CL}_deseq2.csv\"\n",
    "    DIFF_BED = DIFF.strip(\".csv\") + \".bed\"\n",
    "\n",
    "    DIFF_IN_ALL = f\"{CL}_deseq2-nondiff.in.diffAcc.bed\"\n",
    "\n",
    "    df = pd.read_csv(ALL)\n",
    "    df = df.loc[df[\"seqnames\"].isin(chrList())]\n",
    "\n",
    "    fdf = filterConc(df, MIN_CONC, \"Conc_Ultrasound\", \"Conc_Control\")\n",
    "\n",
    "    #plotHist(df, fdf, CL)\n",
    "\n",
    "    # intersect w/ encode\n",
    "\n",
    "    cl_results[CL] = encodeIntersection(CL, df, DIFF_BED, fdf)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5ecd2693",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:19:56.821621Z",
     "start_time": "2024-03-10T03:19:56.801341Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cl</th>\n",
       "      <th>annotation</th>\n",
       "      <th>N_cCRE</th>\n",
       "      <th>N_allAcc</th>\n",
       "      <th>N_filterAcc</th>\n",
       "      <th>N_diffAcc</th>\n",
       "      <th>N_filterDiffAcc</th>\n",
       "      <th>frac_filtered</th>\n",
       "      <th>frac_filtered_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>total.atac</td>\n",
       "      <td>NaN</td>\n",
       "      <td>127857</td>\n",
       "      <td>49249</td>\n",
       "      <td>1916</td>\n",
       "      <td>777</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>CTCF-only,CTCF-bound</td>\n",
       "      <td>20944.0</td>\n",
       "      <td>12332</td>\n",
       "      <td>6072</td>\n",
       "      <td>194</td>\n",
       "      <td>140</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>dELS</td>\n",
       "      <td>25435.0</td>\n",
       "      <td>15164</td>\n",
       "      <td>11296</td>\n",
       "      <td>196</td>\n",
       "      <td>149</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>PLS,CTCF-bound</td>\n",
       "      <td>2377.0</td>\n",
       "      <td>2017</td>\n",
       "      <td>1972</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>0.98</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>pELS</td>\n",
       "      <td>16656.0</td>\n",
       "      <td>5931</td>\n",
       "      <td>5331</td>\n",
       "      <td>50</td>\n",
       "      <td>42</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>PLS</td>\n",
       "      <td>13597.0</td>\n",
       "      <td>10324</td>\n",
       "      <td>10085</td>\n",
       "      <td>58</td>\n",
       "      <td>56</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>pELS,CTCF-bound</td>\n",
       "      <td>2613.0</td>\n",
       "      <td>1226</td>\n",
       "      <td>1159</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k562</td>\n",
       "      <td>total.atac</td>\n",
       "      <td>NaN</td>\n",
       "      <td>102300</td>\n",
       "      <td>82505</td>\n",
       "      <td>890</td>\n",
       "      <td>655</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k562</td>\n",
       "      <td>pELS,CTCF-bound</td>\n",
       "      <td>6580.0</td>\n",
       "      <td>2207</td>\n",
       "      <td>2206</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k562</td>\n",
       "      <td>CTCF-only,CTCF-bound</td>\n",
       "      <td>27649.0</td>\n",
       "      <td>19741</td>\n",
       "      <td>18179</td>\n",
       "      <td>130</td>\n",
       "      <td>116</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k562</td>\n",
       "      <td>PLS,CTCF-bound</td>\n",
       "      <td>6568.0</td>\n",
       "      <td>5146</td>\n",
       "      <td>5144</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k562</td>\n",
       "      <td>pELS</td>\n",
       "      <td>33276.0</td>\n",
       "      <td>6502</td>\n",
       "      <td>6461</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>k562</td>\n",
       "      <td>PLS</td>\n",
       "      <td>12752.0</td>\n",
       "      <td>7926</td>\n",
       "      <td>7920</td>\n",
       "      <td>20</td>\n",
       "      <td>19</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>k562</td>\n",
       "      <td>dELS</td>\n",
       "      <td>30119.0</td>\n",
       "      <td>16034</td>\n",
       "      <td>15794</td>\n",
       "      <td>154</td>\n",
       "      <td>150</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hob</td>\n",
       "      <td>total.atac</td>\n",
       "      <td>NaN</td>\n",
       "      <td>159246</td>\n",
       "      <td>124148</td>\n",
       "      <td>2324</td>\n",
       "      <td>2218</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hob</td>\n",
       "      <td>PLS</td>\n",
       "      <td>18798.0</td>\n",
       "      <td>13636</td>\n",
       "      <td>13560</td>\n",
       "      <td>567</td>\n",
       "      <td>567</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bj</td>\n",
       "      <td>total.atac</td>\n",
       "      <td>NaN</td>\n",
       "      <td>117245</td>\n",
       "      <td>67339</td>\n",
       "      <td>433</td>\n",
       "      <td>242</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bj</td>\n",
       "      <td>PLS,CTCF-bound</td>\n",
       "      <td>3835.0</td>\n",
       "      <td>3345</td>\n",
       "      <td>3330</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bj</td>\n",
       "      <td>PLS</td>\n",
       "      <td>9445.0</td>\n",
       "      <td>7770</td>\n",
       "      <td>7680</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cl            annotation   N_cCRE  N_allAcc  N_filterAcc  N_diffAcc  \\\n",
       "0  hepg2            total.atac      NaN    127857        49249       1916   \n",
       "1  hepg2  CTCF-only,CTCF-bound  20944.0     12332         6072        194   \n",
       "2  hepg2                  dELS  25435.0     15164        11296        196   \n",
       "3  hepg2        PLS,CTCF-bound   2377.0      2017         1972         12   \n",
       "4  hepg2                  pELS  16656.0      5931         5331         50   \n",
       "5  hepg2                   PLS  13597.0     10324        10085         58   \n",
       "6  hepg2       pELS,CTCF-bound   2613.0      1226         1159          7   \n",
       "0   k562            total.atac      NaN    102300        82505        890   \n",
       "1   k562       pELS,CTCF-bound   6580.0      2207         2206          1   \n",
       "2   k562  CTCF-only,CTCF-bound  27649.0     19741        18179        130   \n",
       "3   k562        PLS,CTCF-bound   6568.0      5146         5144         10   \n",
       "4   k562                  pELS  33276.0      6502         6461         31   \n",
       "5   k562                   PLS  12752.0      7926         7920         20   \n",
       "6   k562                  dELS  30119.0     16034        15794        154   \n",
       "0    hob            total.atac      NaN    159246       124148       2324   \n",
       "1    hob                   PLS  18798.0     13636        13560        567   \n",
       "0     bj            total.atac      NaN    117245        67339        433   \n",
       "1     bj        PLS,CTCF-bound   3835.0      3345         3330         12   \n",
       "2     bj                   PLS   9445.0      7770         7680         23   \n",
       "\n",
       "   N_filterDiffAcc  frac_filtered  frac_filtered_diff  \n",
       "0              777           0.39                0.41  \n",
       "1              140           0.49                0.72  \n",
       "2              149           0.74                0.76  \n",
       "3               12           0.98                1.00  \n",
       "4               42           0.90                0.84  \n",
       "5               56           0.98                0.97  \n",
       "6                6           0.95                0.86  \n",
       "0              655           0.81                0.74  \n",
       "1                0           1.00                0.00  \n",
       "2              116           0.92                0.89  \n",
       "3               10           1.00                1.00  \n",
       "4               31           0.99                1.00  \n",
       "5               19           1.00                0.95  \n",
       "6              150           0.99                0.97  \n",
       "0             2218           0.78                0.95  \n",
       "1              567           0.99                1.00  \n",
       "0              242           0.57                0.56  \n",
       "1               12           1.00                1.00  \n",
       "2               23           0.99                1.00  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = pd.concat(cl_results.values())\n",
    "\n",
    "results[\"frac_filtered\"] = results['N_filterAcc']/results[\"N_allAcc\"]\n",
    "results[\"frac_filtered_diff\"] = results['N_filterDiffAcc']/results[\"N_diffAcc\"]\n",
    "results.round(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e4926a4",
   "metadata": {},
   "source": [
    "# filtering results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "877becfc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:20:03.058243Z",
     "start_time": "2024-03-10T03:20:03.046235Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cl</th>\n",
       "      <th>annotation</th>\n",
       "      <th>N_cCRE</th>\n",
       "      <th>N_allAcc</th>\n",
       "      <th>N_filterAcc</th>\n",
       "      <th>N_diffAcc</th>\n",
       "      <th>N_filterDiffAcc</th>\n",
       "      <th>frac_filtered</th>\n",
       "      <th>frac_filtered_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>total.atac</td>\n",
       "      <td>NaN</td>\n",
       "      <td>127857</td>\n",
       "      <td>49249</td>\n",
       "      <td>1916</td>\n",
       "      <td>777</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k562</td>\n",
       "      <td>total.atac</td>\n",
       "      <td>NaN</td>\n",
       "      <td>102300</td>\n",
       "      <td>82505</td>\n",
       "      <td>890</td>\n",
       "      <td>655</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hob</td>\n",
       "      <td>total.atac</td>\n",
       "      <td>NaN</td>\n",
       "      <td>159246</td>\n",
       "      <td>124148</td>\n",
       "      <td>2324</td>\n",
       "      <td>2218</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bj</td>\n",
       "      <td>total.atac</td>\n",
       "      <td>NaN</td>\n",
       "      <td>117245</td>\n",
       "      <td>67339</td>\n",
       "      <td>433</td>\n",
       "      <td>242</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.56</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cl  annotation  N_cCRE  N_allAcc  N_filterAcc  N_diffAcc  \\\n",
       "0  hepg2  total.atac     NaN    127857        49249       1916   \n",
       "0   k562  total.atac     NaN    102300        82505        890   \n",
       "0    hob  total.atac     NaN    159246       124148       2324   \n",
       "0     bj  total.atac     NaN    117245        67339        433   \n",
       "\n",
       "   N_filterDiffAcc  frac_filtered  frac_filtered_diff  \n",
       "0              777           0.39                0.41  \n",
       "0              655           0.81                0.74  \n",
       "0             2218           0.78                0.95  \n",
       "0              242           0.57                0.56  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.loc[results[\"annotation\"]==\"total.atac\"].round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d6d9bf7f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-10T03:20:03.856691Z",
     "start_time": "2024-03-10T03:20:03.842731Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cl</th>\n",
       "      <th>annotation</th>\n",
       "      <th>N_cCRE</th>\n",
       "      <th>N_allAcc</th>\n",
       "      <th>N_filterAcc</th>\n",
       "      <th>N_diffAcc</th>\n",
       "      <th>N_filterDiffAcc</th>\n",
       "      <th>frac_filtered</th>\n",
       "      <th>frac_filtered_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>PLS</td>\n",
       "      <td>13597.0</td>\n",
       "      <td>10324</td>\n",
       "      <td>10085</td>\n",
       "      <td>58</td>\n",
       "      <td>56</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>k562</td>\n",
       "      <td>PLS</td>\n",
       "      <td>12752.0</td>\n",
       "      <td>7926</td>\n",
       "      <td>7920</td>\n",
       "      <td>20</td>\n",
       "      <td>19</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hob</td>\n",
       "      <td>PLS</td>\n",
       "      <td>18798.0</td>\n",
       "      <td>13636</td>\n",
       "      <td>13560</td>\n",
       "      <td>567</td>\n",
       "      <td>567</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bj</td>\n",
       "      <td>PLS</td>\n",
       "      <td>9445.0</td>\n",
       "      <td>7770</td>\n",
       "      <td>7680</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>PLS,CTCF-bound</td>\n",
       "      <td>2377.0</td>\n",
       "      <td>2017</td>\n",
       "      <td>1972</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>0.98</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k562</td>\n",
       "      <td>PLS,CTCF-bound</td>\n",
       "      <td>6568.0</td>\n",
       "      <td>5146</td>\n",
       "      <td>5144</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bj</td>\n",
       "      <td>PLS,CTCF-bound</td>\n",
       "      <td>3835.0</td>\n",
       "      <td>3345</td>\n",
       "      <td>3330</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>dELS</td>\n",
       "      <td>25435.0</td>\n",
       "      <td>15164</td>\n",
       "      <td>11296</td>\n",
       "      <td>196</td>\n",
       "      <td>149</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>k562</td>\n",
       "      <td>dELS</td>\n",
       "      <td>30119.0</td>\n",
       "      <td>16034</td>\n",
       "      <td>15794</td>\n",
       "      <td>154</td>\n",
       "      <td>150</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>pELS</td>\n",
       "      <td>16656.0</td>\n",
       "      <td>5931</td>\n",
       "      <td>5331</td>\n",
       "      <td>50</td>\n",
       "      <td>42</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k562</td>\n",
       "      <td>pELS</td>\n",
       "      <td>33276.0</td>\n",
       "      <td>6502</td>\n",
       "      <td>6461</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hepg2</td>\n",
       "      <td>pELS,CTCF-bound</td>\n",
       "      <td>2613.0</td>\n",
       "      <td>1226</td>\n",
       "      <td>1159</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k562</td>\n",
       "      <td>pELS,CTCF-bound</td>\n",
       "      <td>6580.0</td>\n",
       "      <td>2207</td>\n",
       "      <td>2206</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cl       annotation   N_cCRE  N_allAcc  N_filterAcc  N_diffAcc  \\\n",
       "5  hepg2              PLS  13597.0     10324        10085         58   \n",
       "5   k562              PLS  12752.0      7926         7920         20   \n",
       "1    hob              PLS  18798.0     13636        13560        567   \n",
       "2     bj              PLS   9445.0      7770         7680         23   \n",
       "3  hepg2   PLS,CTCF-bound   2377.0      2017         1972         12   \n",
       "3   k562   PLS,CTCF-bound   6568.0      5146         5144         10   \n",
       "1     bj   PLS,CTCF-bound   3835.0      3345         3330         12   \n",
       "2  hepg2             dELS  25435.0     15164        11296        196   \n",
       "6   k562             dELS  30119.0     16034        15794        154   \n",
       "4  hepg2             pELS  16656.0      5931         5331         50   \n",
       "4   k562             pELS  33276.0      6502         6461         31   \n",
       "6  hepg2  pELS,CTCF-bound   2613.0      1226         1159          7   \n",
       "1   k562  pELS,CTCF-bound   6580.0      2207         2206          1   \n",
       "\n",
       "   N_filterDiffAcc  frac_filtered  frac_filtered_diff  \n",
       "5               56           0.98                0.97  \n",
       "5               19           1.00                0.95  \n",
       "1              567           0.99                1.00  \n",
       "2               23           0.99                1.00  \n",
       "3               12           0.98                1.00  \n",
       "3               10           1.00                1.00  \n",
       "1               12           1.00                1.00  \n",
       "2              149           0.74                0.76  \n",
       "6              150           0.99                0.97  \n",
       "4               42           0.90                0.84  \n",
       "4               31           0.99                1.00  \n",
       "6                6           0.95                0.86  \n",
       "1                0           1.00                0.00  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.loc[results[\"annotation\"].str.contains(\"LS\")].sort_values(by=\"annotation\").round(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73ac0f87",
   "metadata": {},
   "source": [
    "# PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dadcc7e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T21:25:29.369842Z",
     "start_time": "2024-02-27T21:25:15.796583Z"
    }
   },
   "outputs": [],
   "source": [
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de349e0a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T21:31:29.149452Z",
     "start_time": "2024-02-27T21:31:29.094648Z"
    }
   },
   "outputs": [],
   "source": [
    "values = pd.merge(ctrl[[\"coor\", \"rep1\", \"rep2\", ]], us[[\"rep1\", \"rep2\", \"coor\"]], left_on=\"coor\", right_on=\"coor\")\n",
    "values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fd8a0fa",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T21:32:32.665370Z",
     "start_time": "2024-02-27T21:32:32.661329Z"
    }
   },
   "outputs": [],
   "source": [
    "X = values[[\"rep1_x\", \"rep2_x\", \"rep1_y\", \"rep2_y\"]]#.pivot(columns = 'coor', values=['rep1', 'rep2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6300273",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T21:41:03.065111Z",
     "start_time": "2024-02-27T21:41:03.057067Z"
    }
   },
   "outputs": [],
   "source": [
    "pca = PCA(n_components=4)\n",
    "X_ = pca.fit_transform(X)\n",
    "\n",
    "print(pca.explained_variance_ratio_) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b8b9f0a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T21:41:05.510796Z",
     "start_time": "2024-02-27T21:41:05.505638Z"
    }
   },
   "outputs": [],
   "source": [
    "X_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bccce96",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (mamba)",
   "language": "python",
   "name": "mamba"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
