{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "f53b364e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-04-19T21:32:48.694446Z",
     "start_time": "2023-04-19T21:32:48.643642Z"
    },
    "code_folding": [
     115,
     175,
     296
    ]
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Take kmer spectra, nullomer spectra, mutagenize sequences to get kmer spectra of the mutants. \n",
    "\n",
    "input\n",
    "    mer (str) - identity of the kmer length\n",
    "    nmuts (str) - number of positions to mutagenize simultaneously. \n",
    "    cell line (str) - cell ine\n",
    "    build (str) - genome build\n",
    "    \n",
    "\"\"\"\n",
    "\n",
    "import argparse\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "\n",
    "from collections import Counter\n",
    "import glob\n",
    "import gzip\n",
    "from itertools import product, combinations\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import subprocess as sp\n",
    "\n",
    "sys.path.append(\"/wynton/home/ahituv/fongsl/tools/py_\")\n",
    "import config_readwrite as crw\n",
    "\"\"\"\n",
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument(\"cellline\", type=str, help = \"cellline\")\n",
    "parser.add_argument(\"kmer_length\", type=int, help = \"length of kmer\")\n",
    "parser.add_argument(\"n_mut\", type=int, help = \"number of mutations to make\")\n",
    "parser.add_argument(\"build\", type=str, help = \"genome build\")\n",
    "parser.add_argument(\"fo_only\", type=bool, default=True, help = \"keep first order nullomers only?\")\n",
    "parser.add_argument(\"config\", type=str, help = \"path to config file\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "CL, MER, NMUTS, BUILD, FO_ONLY, CONFIG = args.cellline, args.kmer_length, args.n_mut, args.build, args.fo_only, args.config\n",
    "\n",
    "\"\"\"\n",
    "# dev\n",
    "CL, MER, NMUTS, BUILD, FO_ONLY = \"common\", \"14\", \"2\", \"hs1\", True\n",
    "CONFIG = os.path.join(os.getcwd(), \"config.ini\")\n",
    "\n",
    "# functions\n",
    "\n",
    "\n",
    "def getFirstOrderNulls(kmer_len, build):\n",
    "    \"\"\"\n",
    "    return set of first order nullomers in genome background. \n",
    "\n",
    "    input\n",
    "        kmer_len (int) - length of kmer\n",
    "        build (str) - genome build of nullomer\n",
    "    method\n",
    "        1. make empty set\n",
    "        2. glob all first order nullomer files\n",
    "        3. for each file, read and add nullomer sequence to set\n",
    "\n",
    "    return\n",
    "        first_order (set) - set of first_order nullomers.\n",
    "    \"\"\"\n",
    "    first_order = set()\n",
    "    for f in glob.glob(f\"/wynton/home/ahituv/fongsl/dna/{build}/kmers/{kmer_len}mers/order*.tsv\"):\n",
    "        with open(f, \"r\") as reader:\n",
    "            for line in reader:\n",
    "                first_order.add(line.strip(\"\\n\").split(\"\\t\")\n",
    "                                [0])  # add seq to set\n",
    "\n",
    "    print(len(first_order))\n",
    "    return first_order\n",
    "\n",
    "\n",
    "def getSeqDict(file_list):\n",
    "    \"\"\"\n",
    "    get dictionary of kmer sequences and counts, guides\n",
    "\n",
    "    input\n",
    "        file_list (list) - list of files w/ full path\n",
    "\n",
    "    method \n",
    "        1. instantiate collection dict\n",
    "        2. read file, extract sequence and kmer counts, add to element specific dictionary\n",
    "    return\n",
    "        seq_dict (dict) - dictionary of kmers and counts\n",
    "\n",
    "    \"\"\"\n",
    " \n",
    "    # print(\"getting sequence file\", file)\n",
    "\n",
    "    # 1\n",
    "    seq_dict = {}\n",
    "\n",
    "    # 2\n",
    "    for file in file_list:\n",
    "        key = os.path.split(file)[1]\n",
    "\n",
    "        if \"fwd\" in file or \"rev\" in file:\n",
    "             key = key.split(\".csv.gz\")[0]\n",
    "\n",
    "        else:\n",
    "            key = (key).split(\".\")[0]  # get the key AAAA\n",
    "\n",
    "        if key not in seq_dict:\n",
    "            seq_dict[key] = {}\n",
    "\n",
    "        file_dict=seq_dict[key]\n",
    "\n",
    "        with gzip.open(file, \"rt\") as reader:\n",
    "\n",
    "            for line in reader.readlines():\n",
    "                seq, count = line.strip(\"\\n\").split(\",\")\n",
    "                file_dict[seq] = count\n",
    "\n",
    "    return seq_dict\n",
    "\n",
    "\n",
    "def genKmers(length):\n",
    "    \"\"\"\n",
    "    return all sequence permutations, including repeats (AAAAA, GGGGG, CCCCC etc.)\n",
    "\n",
    "    require\n",
    "        itertools.product\n",
    "\n",
    "    input\n",
    "        length (int)\n",
    "\n",
    "    return\n",
    "        mers_list (list) - list of all nucleotide permutations \n",
    "    \"\"\"\n",
    "    print(\"generating kmer space length\", length)\n",
    "\n",
    "    mers = product(\"ACTG\", repeat=length)\n",
    "\n",
    "    mers_list = list(\"\".join(i) for i in mers)\n",
    "\n",
    "    return mers_list\n",
    "\n",
    "\n",
    "def getPosLetterCombos(nmuts, sequence):\n",
    "    \"\"\"\n",
    "    return combinations of (1) indices (2) mutated bases for mutating a sequence\n",
    "\n",
    "    require \n",
    "        itertools\n",
    "\n",
    "    inputs \n",
    "        nmuts (int) - number mutations to make\n",
    "        sequence (str) - sequence to be mutated\n",
    "\n",
    "    method\n",
    "        1. get index combinations based on sequence length and number of mutations to make. \n",
    "            This makes a map of all possible combinations of sequences to mutate\n",
    "            1.1 combinations requires that each index is unique and non-redundant. \n",
    "                Order does not matter - \n",
    "                    e.g. (2,4,5) is the same as (5,2,4) because indexes 2, 4, and 5, will all be mutated.\n",
    "\n",
    "        2. get sequence product to mutate at indexes\n",
    "            2.1 - product allows for repeats of the same base in different positions\n",
    "\n",
    "    return\n",
    "        mut_pos (list) - list of positional combinations to mutate\n",
    "        mut_bases (list) - list of letter combinations to mutate\n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "    # print(\"making index combinations, nucletide permutations of length\", nmuts)\n",
    "\n",
    "    # 1 index combinations\n",
    "    mut_pos = list(combinations(np.arange(len(sequence)), nmuts))\n",
    "\n",
    "    # 2 nucleotide permutations per index combo.\n",
    "    mut_bases = list(product(\"ACGT\", repeat=nmuts))\n",
    "\n",
    "    return mut_pos, mut_bases\n",
    "\n",
    "\n",
    "def buildSeqMut(sequence, mut_pos, mut_bases):\n",
    "    \"\"\"\n",
    "    mutate sequence at position with letter\n",
    "    multiple positions and letters can be inserted into the sequence simultaneously. \n",
    "\n",
    "    input\n",
    "        sequence (str) - original sequence\n",
    "        mut_pos (set) - Sets of single, tuple, threeple positional index(es) to mutate\n",
    "        mut_bases (tuple) - Sets of single, tuple, threeple nucleotide combinations to mutate sequence to\n",
    "\n",
    "    method\n",
    "        1. instantiate seqs set to collect mutated sequences, add identity to seq set\n",
    "        2. per positions to mutate in set\n",
    "            2.1 per base combinations to mutate at these positions\n",
    "        3. zip and iterate through positions and bases, mutating input sequence\n",
    "        4. IF mut_seq != input sequence, then return. Else, skip\n",
    "\n",
    "\n",
    "    return \n",
    "        seqs (set) - set of sequences with mutations \n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # 1\n",
    "    seqs, mut_seq = set(), \"\"\n",
    "\n",
    "    # 2\n",
    "    for pos in mut_pos:\n",
    "\n",
    "        # 2.1\n",
    "        for letters in mut_bases:\n",
    "\n",
    "            # 3\n",
    "            for p, l in zip(pos, letters):\n",
    "\n",
    "                if mut_seq == \"\":\n",
    "                    mut_seq = sequence[:p] + l + sequence[p + 1:]\n",
    "\n",
    "                else:\n",
    "                    mut_seq = mut_seq[:p] + l + mut_seq[p + 1:]\n",
    "\n",
    "            # 4\n",
    "            if mut_seq != sequence:\n",
    "                seqs.add(mut_seq)\n",
    "                mut_seq = \"\"\n",
    "\n",
    "            else:\n",
    "                # print('no mut', sequence, mut_seq\n",
    "                mut_seq = \"\"\n",
    "                pass\n",
    "    # add the identity sequence back into the mix.\n",
    "    seqs.add(sequence)\n",
    "    return seqs\n",
    "\n",
    "\n",
    "def generateMismatchSpectra(sequence, nmuts, kmer_spectra):\n",
    "    \"\"\"\n",
    "    input \n",
    "        sequence (str) - sequence to mutate and match with kmer-spectra\n",
    "        nmuts (int) - max number of mismatches to mutate each kmer sequence by\n",
    "        kmer_spectra (dict) - dictionary of kmer keys and their frequency count (value)\n",
    "\n",
    "    require\n",
    "        getPosLetterCombos function\n",
    "        buildSeqMut function\n",
    "        prettifySeq function\n",
    "\n",
    "    method\n",
    "        1. instantiate kmer_match, null_match list.\n",
    "\n",
    "        2. get all combinations of indexes and mutated bases to try. \n",
    "            mutate bases will be a single, tuple, or threeple depending on the number of mutations desired.\n",
    "\n",
    "        3. mutate all positions of the nullomer with that base. \n",
    "\n",
    "            3.1 - get mutated sequences using buildSeqMut function\n",
    "                NOTE - mutBase function will remove identity sequence. See function method step 4.3.\n",
    "\n",
    "        4. per mutated sequence \n",
    "            4.1 look up kmer count of the mutated sequence. Test whether it is a prime. \n",
    "            4.2 add any resurfaced nullomer sequences to the resurface dict \n",
    "                - that is, resurfaced nullomers are sequences where mutations create nullomers\n",
    "                - return this dictionary\n",
    "            4.3 add kmer count to the dictionary as pretty key\n",
    "        7. Add seq kmer dictionary back into the collection dictionary. \n",
    "\n",
    "\n",
    "    return\n",
    "        kmer_match (list) - all mis-match sequences w/ n muts that are kmers \n",
    "        null_match (list) - all mis-match sequences w/ n muts that are NOT kmers (therefor nullomers) \n",
    "            every mismatch combination for each sequence \n",
    "            and per mismatch combo, the kmer counts for each mismatch\n",
    "\n",
    "    \"\"\" \n",
    "    ## print(\"making kmer mismatch spectra w/ N mutations =\", nmuts)\n",
    "    # 1\n",
    "    kmer_match, null_match = Counter(), Counter()\n",
    "\n",
    "    # 2\n",
    "    mut_pos, mut_bases = getPosLetterCombos(int(nmuts), sequence)  # (1,2) (A,C)\n",
    "\n",
    "    # 3\n",
    "    seqs = buildSeqMut(sequence, mut_pos, mut_bases)\n",
    "\n",
    "    # 4\n",
    "    for mut_seq in seqs:\n",
    "\n",
    "        # 4.1 - kmer frequency\n",
    "        if mut_seq in kmer_spectra[mut_seq[:4]].keys():\n",
    "            kmer_match[mut_seq] += 1\n",
    "\n",
    "        # 4.2 - nullomer after mutations, not in kmer-verse\n",
    "        else:\n",
    "            null_match[mut_seq] += 1  # append to list\n",
    "\n",
    "\n",
    "    #print(len(kmer_match), len(null_match), len(seqs))\n",
    "\n",
    "    return dict(kmer_match), dict(null_match)\n",
    "\n",
    "\n",
    "def prettifySeq(original, mut):\n",
    "    \"\"\"\n",
    "    prettify sequence. \n",
    "        All matching bases will be written as \".\"\n",
    "        All non matching bases will be written w mutated base identity. \n",
    "    \"\"\"\n",
    "    prettyseq = \"\"\n",
    "    for o, m in zip(original, mut):\n",
    "        if o != m:\n",
    "            prettyseq += m\n",
    "        else:\n",
    "            prettyseq += \".\"\n",
    "\n",
    "    return prettyseq\n",
    "\n",
    "\n",
    "def buildOtherSeqMuts(original_seq, pos1, nucs1, pos2, nucs2):\n",
    "    \"\"\"\n",
    "    mutate original sequence at positions 1 and 2 with two alternative bases, \n",
    "    which are not the original bases nor the target nullomer base. These\n",
    "    are controls for downstream analysis of nullomer mutations\n",
    "    \n",
    "    input\n",
    "        original_sequence (str) - original sequence\n",
    "        pos1 (int) - first position in sequence to mutate to other bases\n",
    "        nucs1 (tuple) - two other nucleotides (as tuple) to mutate position 1 to\n",
    "        pos2 (int) - second position in sequence to mutate to other bases\n",
    "        nucs2 (tuple) - two other nucleotides (as tuple) to mutate position 2 to \n",
    "        \n",
    "    method\n",
    "        1. make otherseqs, empty set to collect mutated sequence results. \n",
    "        2. iterate through possibilities (4) \n",
    "            2.1 set variables for each nucleotide identity at position one and two to be mutated to. \n",
    "            n1 - nucleotide identity at position one\n",
    "                e.g. pos1 = 5, nucs1 = (c,t)\n",
    "                When n1=0, insert \"c\" at position 1 (index 5). When n1=1, insert \"t\" at position 5\n",
    "            n2 - nucleotide identity at position two\n",
    "                e.g. pos2 = 9, nucs1 = (a,g)\n",
    "                When n1=0, insert \"a\" at position 2 (index 9). When n1=1, insert \"g\" at position 9\n",
    "        3. add the mutated sequence to the sequence set\n",
    "        \n",
    "    return\n",
    "        otherseqs (set) - set of other mutated sequences that are not original or nullomer mutations. \n",
    "        \n",
    "    \"\"\"\n",
    "    #1\n",
    "    otherseqs = set()\n",
    "    \n",
    "    #2\n",
    "    for i in np.arange(4):\n",
    "        #2.1\n",
    "        n1 = 0 if i <2 else 1  # toggle first position nucleotide\n",
    "        n2 = 0 if i%2 ==0 else 1  # toggel second position nucleotide\n",
    "\n",
    "        #3 build sequence with two mutations to other possible nucleotide combinations\n",
    "        otherseqs.add(original[:pos1] + nucs1[n1] + original[pos1 + 1:pos2] + nucs2[n2] + original[pos2 + 1:])\n",
    "\n",
    "    return otherseqs\n",
    "\n",
    "def makeNullrow(seq, seq_count, null_list, nmuts, seq_id, datatype):\n",
    "    \"\"\"\n",
    "    return array with information about the original kmer sequence, its frequency, and matching nullomers. \n",
    "\n",
    "    inputs\n",
    "        seq (str) - kmer sequence\n",
    "        seq_count (int) - count of kmer occurrence in dataset\n",
    "        null_list (list) -list  of nullomers that are n mutations away\n",
    "        nmuts (str) - number of mutations between kmer and nullomer\n",
    "        seq_id (str) - coordinates of sequence\n",
    "        datatype (str) -  nullomer, kmer, or kmer_control\n",
    "        \n",
    "        \n",
    "\n",
    "    method\n",
    "        1. prettify the nullomer mutations (makes it easy to read)\n",
    "        2. get gc fraction for kmer, nullomers\n",
    "        3. construct a list of row items \n",
    "        \n",
    "        rownames = [\"kmer\", 'nKmer', \"gc_frac\", \"relatedNulls\", \n",
    "                    \"nNulls\", gc_fracNulls\", prettyNulls\", \"nMuts\"\n",
    "                    ]\n",
    "        \n",
    "\n",
    "    return \n",
    "        (str) - tab-joined row list + new line\n",
    "    \"\"\"\n",
    "\n",
    "    # 1\n",
    "    rows = []\n",
    "    \n",
    "    gc_seq = round(gc_fraction(seq), 2)\n",
    "\n",
    "    for n, mut in enumerate(null_list):\n",
    "        \n",
    "        pretty_seq = str(prettifySeq(seq, mut))\n",
    "\n",
    "        gc_mut = str(round(gc_fraction(mut), 2)) \n",
    "\n",
    "        rowlist = [seq,  # original kmer\n",
    "               str(seq_count),  # number of times kmer occurs in set\n",
    "               str(gc_seq),  # kmer gc fraction\n",
    "               mut,  # related nullomer sequences\n",
    "               str(len(null_list)),  # number of related nullomers\n",
    "               gc_mut,  # gc fraction of related nullomers\n",
    "               pretty_seq,  # gc fraction of related nullomers\n",
    "               f\"{datatype}-{n}\", # to map to fa\n",
    "               ]\n",
    "        \n",
    "        rows.append(\"\\t\".join(rowlist) + \"\\n\")\n",
    "\n",
    "    return rows\n",
    "\n",
    "\n",
    "# WRITE FA files with this process \n",
    "\n",
    "def writeFa(seq, null_list, seq_id, datatype):\n",
    "    \n",
    "    rows = []\n",
    "    for n, null_seq in enumerate(null_list):\n",
    "        rows.append(f\">{seq}.{seq_id}.{datatype}-{n}\\n{null_seq}\\n\")\n",
    "    \n",
    "    return \"\".join(rows)\n",
    "\n",
    "def makeOutfiles(element, outdir, mer, nmuts, fo_only):\n",
    "    \n",
    "    if fo_only is True:\n",
    "        tsv = os.path.join(outdir, f\"{element}.{mer}mers.{nmuts}mut.nulls.fo.tsv\")\n",
    "        fa = os.path.join(outdir, f\"{element}.{mer}mers.{nmuts}mut.nulls.fo.fa\")\n",
    "        \n",
    "    else:\n",
    "        tsv = os.path.join(outdir, f\"{element}.{mer}mers.{nmuts}mut.nulls.tsv\")\n",
    "        fa = os.path.join(outdir, f\"{element}.{mer}mers.{nmuts}mut.nulls.fa\")\n",
    "\n",
    "    return tsv, fa\n",
    "\n",
    "def writer(row, outfile):\n",
    "    with open(outfile, \"a\") as writer:\n",
    "        writer.write(row)\n",
    "        writer.close()\n",
    "        \n",
    "def getCoor(key, mer):\n",
    "\n",
    "    #key, mer= \"AACTA\", '14'\n",
    "    COOR_REF = f\"/wynton/home/ahituv/fongsl/MPRA/agarwal_2023/kmers/{mer}mers/coor/{key}_coor.csv.gz\"\n",
    "    coor_dict = {}\n",
    "    with gzip.open(COOR_REF, 'rt') as reader:\n",
    "        for line in reader:\n",
    "            parts = line.strip('\\n').split(',')\n",
    "            value_seq, coors = parts[0], parts[1:]\n",
    "\n",
    "            stripped = set([i.split(\"_\")[0] for i in coors if i != \"\"]) # take set of coordinates, remove empty, remove +/-\n",
    "            coor_dict[key+value_seq] = \",\".join(stripped)\n",
    "    return coor_dict\n",
    "\n",
    "# main\n",
    "\n",
    "\n",
    "def main(argv):\n",
    "\n",
    "    \"\"\"\n",
    "    mutagenize kmers by n mutations and evaluate if they are still kmers or if they become nullomers\n",
    "\n",
    "    1. get kmer search space\n",
    "        1.1 change first order only flag to False if length of kmer is < 14. No first order nullomers below 14bp. \n",
    "        \n",
    "    2. get genome search space, outdir to write to \n",
    "        2.1 get first order nullomer set\n",
    "        \n",
    "    3. get kmer files, make them into dictionary. \n",
    "        add_key = True means file is separated by key sequence (in filename) \n",
    "        and value sequences (all sequences sharing that key). Full kmer sequences = key + value sequences. \n",
    "\n",
    "    4. get genome files, make those kmers into dictionary. \n",
    "        4.1 filter all the nullomer sequences out of the file list\n",
    "        4.2 add_key = False. Full kmer sequence (with key and value) are already written into genome file. \n",
    "        \n",
    "    5. retrieve outfiles to write (.tsv, .fa)\n",
    "        5.1 - write to config file. \n",
    "        \n",
    "    6. mutate kmers to identify related nullomers\n",
    "\n",
    "    7. write nullomers\n",
    "            7.1. get nullomers that match kmer\n",
    "            7.2. intersect first order w/ nullomers, check if any are first order. \n",
    "            7.3 if nulls (or first order), write .tsv, .fa\n",
    "    \"\"\"\n",
    "\n",
    "    #1\n",
    "    if CL ==\"common\":\n",
    "        KMER_REF = f\"/wynton/home/ahituv/fongsl/MPRA/agarwal_2023/kmers/{MER}mers/keys/*.csv.gz\"\n",
    "        COOR_REF = f\"/wynton/home/ahituv/fongsl/MPRA/agarwal_2023/kmers/{MER}mers/coor/*.csv.gz\"\n",
    "    else:\n",
    "        KMER_REF = f\"/wynton/home/ahituv/fongsl/nullomers/data/lock/{CL}/kmers/{MER}mers/*.csv.gz\"\n",
    "\n",
    "    #1.1\n",
    "    if int(MER)<14:\n",
    "        FO_ONLY=False\n",
    "\n",
    "    #2\n",
    "    GENOME_REF = f\"/wynton/home/ahituv/fongsl/dna/{BUILD}/kmers/{MER}mers/*.csv.gz\"\n",
    "    OUTDIR = f\"/wynton/home/ahituv/fongsl/nullomers/data/lock/{CL}/\"\n",
    "    \n",
    "    #2.1\n",
    "    FIRST_ORDER = getFirstOrderNulls(MER, BUILD)\n",
    "    \n",
    "    #3 open up and get kmer sequence dictionaries.\n",
    "    KMERS = glob.glob(KMER_REF)\n",
    "\n",
    "    kmer_seqs = getSeqDict(KMERS)\n",
    "\n",
    "    print(\"kmer seqs\", list(kmer_seqs.keys())[:3])\n",
    "\n",
    "    #4 get all kmer sequences in the genome.\n",
    "    genome_files = glob.glob(GENOME_REF)\n",
    "    GKMERS = []\n",
    "    \n",
    "    #4.1 filter out nullomer.csv files\n",
    "    [GKMERS.append(k) if \"nullomer\" not in k else next for k in genome_files]\n",
    "    \n",
    "    #4.2\n",
    "    genome_seqs = getSeqDict(GKMERS)\n",
    "\n",
    "    print(\"genome_seqs\", list(genome_seqs.keys())[:3])\n",
    "    \n",
    "\n",
    "    # 5.1 write to config\n",
    "    config, cfn = crw.read(CONFIG)\n",
    "    section = f\"{CL}.{MER}mer.{NMUTS}mut\"\n",
    "    crw.check(config, section)\n",
    "\n",
    "    config[section][\"KMER_REF\"] = KMER_REF\n",
    "    config[section][\"GENOME_REF\"] = GENOME_REF\n",
    "    config[section][\"PATH\"] = OUTDIR\n",
    "   \n",
    "    crw.write(config, cfn)\n",
    "    # 5 make the out file\n",
    "    tsv, fa = makeOutfiles(CL, OUTDIR, MER, NMUTS, True)  # make chr-based file\n",
    "\n",
    "    # 6 mutate kmers to find nullomer sequences.\n",
    "    for element, seq_dict in kmer_seqs.items():  # element = 'chr1.24114961-24115161.fwd'\n",
    "\n",
    "        seq_id = element.split(\".csv.gz\")[0]\n",
    "\n",
    "        for seq, seq_count in seq_dict.items():  #seq = \"ACCGGTTCCACGGT\"\n",
    "\n",
    "            sequence = seq_id + seq # seq_id = first letters of sequence, seq = last letters of sequence\n",
    "\n",
    "            # mutate one sequence\n",
    "            related_kmers, related_nullomers = generateMismatchSpectra(\n",
    "                sequence, int(NMUTS), genome_seqs)\n",
    "\n",
    "            # 7 if nullomers appear after mutagenesis\n",
    "\n",
    "            # 7.1\n",
    "            nulls = [s for s in related_nullomers.keys()]\n",
    "            control_kmers = [k for k in related_kmers.keys()]\n",
    "\n",
    "            # 7.2\n",
    "            first_order = FIRST_ORDER.intersection(set(nulls))\n",
    "\n",
    "            # First order filter - write ALL nullomers or ONLY first order (fo)?\n",
    "\n",
    "            null_list = list(first_order)\n",
    "\n",
    "            #null_list = nulls\n",
    "\n",
    "            # 7.3\n",
    "            if len(null_list) > 0:\n",
    "                print(\"write\", sequence)\n",
    "\n",
    "                infos = [\n",
    "                    ([sequence], \"kmer\"),\n",
    "                    (null_list, \"null\"),\n",
    "                    (list(np.random.choice(control_kmers, size = 30)), 'ctrl')\n",
    "                    ]\n",
    "\n",
    "                # get coordinate info\n",
    "                coor = getCoor(seq_id, \"14\")\n",
    "                seq_coor = coor[sequence]\n",
    "\n",
    "                for seq_list, datatype in infos:\n",
    "                    farow = writeFa(sequence, seq_list, seq_coor, datatype)\n",
    "                    rows = makeNullrow(sequence, seq_count, seq_list, NMUTS, seq_coor, datatype)\n",
    "\n",
    "                    # open writer files\n",
    "                    fwriter, twriter = open(fa, \"a\"), open(tsv, \"a\")\n",
    "\n",
    "                    # write fa\n",
    "                    fwriter.write(farow)\n",
    "\n",
    "                    # write tsv\n",
    "                    for row in rows:\n",
    "                        twriter.write(row)\n",
    "\n",
    "                    twriter.close(), fwriter.close()\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main(sys.argv[1:])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (wynenv)",
   "language": "python",
   "name": "wynenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
