{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "526ee122",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T20:38:31.523211Z",
     "start_time": "2023-03-08T20:38:31.516604Z"
    }
   },
   "outputs": [],
   "source": [
    "import argparse\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "from Bio.Seq import Seq\n",
    "import glob\n",
    "import gzip\n",
    "from itertools import product\n",
    "import numpy as np\n",
    "import os, sys\n",
    "\n",
    "from timeit import default_timer as timer\n",
    "from functools import partial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "74bc9a6d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T20:38:33.492157Z",
     "start_time": "2023-03-08T20:38:33.471470Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] directory length keysize\n",
      "ipykernel_launcher.py: error: the following arguments are required: length, keysize\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/wynton/home/ahituv/fongsl/.conda/envs/wynenv/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3441: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "arg_parser = argparse.ArgumentParser()\n",
    "\n",
    "arg_parser.add_argument(\"directory\", type=str, help='directory where chromosom.fa lives')\n",
    "arg_parser.add_argument(\"length\", type=int, help='kmer length')\n",
    "arg_parser.add_argument(\"keysize\", type=int, help='kmer key length for storing info')\n",
    "\n",
    "args = arg_parser.parse_args()\n",
    "\n",
    "PATH, WINDOW_SIZE, KEYSIZE = args.directory, args.length, args.keysize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3f56849b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T20:38:35.866764Z",
     "start_time": "2023-03-08T20:38:35.859399Z"
    }
   },
   "outputs": [],
   "source": [
    "PATH, WINDOW_SIZE, KEYSIZE  =  \"/wynton/home/ahituv/fongsl/dna/hg38/\", 11, 4\n",
    "path, window_size, keysize  = PATH, WINDOW_SIZE, KEYSIZE\n",
    "OUTDIR = os.path.join(PATH, \"kmers\", f\"{WINDOW_SIZE}mers\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b0266d2c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T20:40:25.516631Z",
     "start_time": "2023-03-08T20:40:25.506332Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/wynton/home/ahituv/fongsl/dna/hg38/kmers/11mers\n"
     ]
    }
   ],
   "source": [
    "OUTDIR = os.path.join(PATH, \"kmers\", f\"{WINDOW_SIZE}mers\")\n",
    "\n",
    "print(OUTDIR)\n",
    "# FUNCTIONS\n",
    "\n",
    "\n",
    "def makeChrList():\n",
    "    \"\"\"\n",
    "    return list of chromosome numbers\n",
    "    \"\"\"\n",
    "    chr_list = list(np.arange(1,23))\n",
    "    chr_list.extend([\"X\", \"Y\"])\n",
    "    \n",
    "    return chr_list\n",
    "\n",
    "\n",
    "def makeKeys(keysize):\n",
    "    \"\"\"\n",
    "    input\n",
    "        keysize (int) - length of kmer-keys to create\n",
    "    \n",
    "    require\n",
    "        itertools.product\n",
    "        python list comprehension\n",
    "\n",
    "    method\n",
    "        itertools.product kmers -> join kmers into str -> make list\n",
    "\n",
    "\n",
    "    return\n",
    "        key_set (set) - list of strings. \n",
    "    \"\"\" \n",
    "    key_set = set()\n",
    "    for item in product(\"ACGT\", repeat=keysize):\n",
    "        key_set.add(\"\".join(item))\n",
    "        \n",
    "    return key_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d43e5aea",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T20:42:21.674491Z",
     "start_time": "2023-03-08T20:42:21.656288Z"
    }
   },
   "outputs": [],
   "source": [
    "def countKeymerSpace(windowsize, path, key):\n",
    "    \"\"\"\n",
    "    input\n",
    "        universe_kmer - dictionary with the kmer universe    \n",
    "        windowsize(int) - size of the kmer universe\n",
    "        path(str) - path to dataframes\n",
    "        key (str) - sequence key that kmers are split on\n",
    "\n",
    "    method\n",
    "        1. per key, \n",
    "            1.1 calculate the length of the value sequences \n",
    "            1.2 get set of all possible value sequences \n",
    "            1.3 make a new dictionary of value sequences associated with that key\n",
    "\n",
    "        2. get all files with that key across chromosomes\n",
    "        3. per file, read sub-kmer (i.e. kmer = key + sub-kmer) counts (val) associated with that key\n",
    "\n",
    "        4. if key is in keymer universe, than sum add the new count to the sum. \n",
    "\n",
    "        6. write universe_key dictionary to file. \n",
    "        7. write the key to the chrlog\n",
    "        8. delete chr based files\n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    #1.1\n",
    "    value_seq_len = windowsize - len(key) # get length of value sequence (wo key)\n",
    "    \n",
    "    #1.2\n",
    "    value_seqs = makeKeys(value_seq_len)  # get set of all value sequence combos (len = windowsize-key) \n",
    "    \n",
    "    #1.3\n",
    "    key_universe = dict.fromkeys(value_seqs, 0)  # dict to sum sequence results across chromosomes\n",
    "\n",
    "    #2\n",
    "    query = os.path.join(path, f\"*.{windowsize}mers.{key}.csv\")\n",
    "    FS = glob.glob(query)\n",
    "    \n",
    "    #3\n",
    "    for F in FS:\n",
    "        start = timer()\n",
    "        \n",
    "        with open(F, \"r\") as reader:\n",
    "            for line in reader.readlines():\n",
    "                start = timer()\n",
    "                val_seq, count = line.strip(\"\\n\").split(\",\")\n",
    "                \n",
    "                #4\n",
    "                key_universe[val_seq] +=int(count)\n",
    "               \n",
    "    \n",
    "    #6\n",
    "    writeDictionary(f\"ALL\", key, path, windowsize, key_universe)\n",
    "    \n",
    "    #7\n",
    "    writeChrLog(key, path)\n",
    "    \n",
    "    #8 \n",
    "    #os.system(f\"rm {query}\")\n",
    "    \n",
    "    \n",
    "def writeDictionary(name, key, path, windowsize, result_dict):\n",
    "    \n",
    "    \"\"\"\n",
    "    write kmer universe dictionary to outfile, write nullomers\n",
    "    \n",
    "    input\n",
    "        name (str) - chr number\n",
    "        path (str) - path to write outfile\n",
    "        windowsize (int) - size of window used for kmers\n",
    "        results_dict (dictionary) - kmer universe w/ counts of kmer occurrences \n",
    "    \n",
    "    method\n",
    "        1. create the outfile\n",
    "        2. write key, value as comma-separated str to EITHER nullomer or kmer outfile\n",
    "        3. close the outfiles\n",
    "        4. delete empty files\n",
    "    \n",
    "    return \n",
    "        out_file (str) - written file name\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    #1\n",
    "    kmer_file = os.path.join(path, f\"{name}.{key}.{windowsize}mers.csv\")\n",
    "    null_file = os.path.join(path, f\"{name}.{key}.{windowsize}mers-nullomers.csv\")\n",
    "    \n",
    "    \n",
    "    #2\n",
    "    nullomer = open(null_file, \"w\")\n",
    "    with open(kmer_file, \"w\") as kmer:\n",
    "        \n",
    "        for value_seq, count in result_dict.items():\n",
    "            seq = key+value_seq # put the full sequence back together\n",
    "            \n",
    "            if int(count) > 0: # if nullomer\n",
    "                kmer.write(f\"{seq},{count}\\n\")\n",
    "            else:\n",
    "                nullomer.write(f\"{seq},{count}\\n\")\n",
    "    #3\n",
    "    kmer.close(), nullomer.close()\n",
    "    \n",
    "    #4\n",
    "\n",
    "    #print(\"\\n\\nwrote\", kmer_file)\n",
    "    \n",
    "    return kmer_file, null_file\n",
    "\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c88fd9a9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T20:42:17.123104Z",
     "start_time": "2023-03-08T20:42:17.110432Z"
    },
    "code_folding": [
     39
    ]
   },
   "outputs": [],
   "source": [
    "def readChrLog(path):\n",
    "    \"\"\"\n",
    "    read chr.log and determine whether chromosome has already been summed into file dataset\n",
    "    \n",
    "   input\n",
    "        chr_num (str) - chromosome number\n",
    "        path (str) - path to directory to write log\n",
    "        \n",
    "    method\n",
    "        1. make the log file\n",
    "            1.1 if log file does not exist, return False (none of the chromosomes have been run) \n",
    "        2. open existing log file. \n",
    "        3. append chr_num to list\n",
    "        4. check whether input chr_num is in list:\n",
    "            4.1 if yes, return True (chromosome has been run)\n",
    "            4.2 if no, return False (chromos dome has not been run)\n",
    "    \n",
    "    return\n",
    "        empty set\n",
    "        chr_runlist (set) - if key has been added to final nullomer count\n",
    "    \"\"\"\n",
    "    \n",
    "    #1\n",
    "    out = os.path.join(path, \"chr.log\")\n",
    "\n",
    "    #2\n",
    "    runlist=set()\n",
    "    if os.path.exists(out) is True:\n",
    "    \n",
    "        with open(out, \"r\") as chrlog:\n",
    "        #3\n",
    "            for line in chrlog.readlines():\n",
    "                runlist.add(str(line.split(\"\\n\")[0]))\n",
    "        \n",
    "        \n",
    "    return runlist\n",
    "       \n",
    "            \n",
    "            \n",
    "def writeChrLog(chr_num, path):\n",
    "    \"\"\"\n",
    "    write log of chrs written to the all mer\n",
    "    \n",
    "    input\n",
    "        chr_num (str) - chromosome number\n",
    "        path (str) - path to directory to write log\n",
    "        \n",
    "    method\n",
    "        1. make the log file\n",
    "        2. open the log file\n",
    "        3. append the chr_num to the log file\n",
    "        4. close the file\n",
    "    \"\"\"\n",
    "    #1\n",
    "    out = os.path.join(path, \"chr.log\")\n",
    "    \n",
    "    #2\n",
    "    with open(out, \"a\") as chrlog:\n",
    "        \n",
    "        #3\n",
    "        chrlog.write(f\"{chr_num}\\n\")\n",
    "    \n",
    "    #4\n",
    "    chrlog.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e3e1c88",
   "metadata": {},
   "outputs": [],
   "source": [
    "    \n",
    "# MAIN\n",
    "\n",
    "def main(argv):\n",
    "    \"\"\"\n",
    "    count number of kmer occurrences per chromosome, write output to file\n",
    "    \n",
    "    method\n",
    " \n",
    "        1. read log, check if you've run any keys already\n",
    "        2. get set of keys to read. \n",
    "        3. if key has been run (set), take difference of key_set from run_already set. \n",
    "            Else, take difference from empty set (i.e. difference is nothing and full kmer_key set gets run) \n",
    "        4. for each kmer key, \n",
    "            4.1 Use partial function to run kmer keys separately in parallel as a pool\n",
    "            4.2 Update kmer-key universe dict values w/ sum the number of instances across chromosomes\n",
    "            4.3 Write the universe dictionary for every kmer key. \n",
    "            4.4 Write kmer key to log file (this line is inside the countKeymerSpace function). \n",
    "                Thus, only when kmer key is complete will universe be written and chrlog be updated. \n",
    "        5. concatenate nullomer, kmer files\n",
    "    \"\"\"\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6807d4f9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T20:42:25.305790Z",
     "start_time": "2023-03-08T20:42:25.298887Z"
    }
   },
   "outputs": [],
   "source": [
    "    #1\n",
    "    run_already = readChrLog(OUTDIR) # check whether key has been run already\n",
    "\n",
    "    #2 get set of kmer keyes based on keysize \n",
    "    key_set = makeKeys(KEYSIZE)\n",
    "\n",
    "    #3\n",
    "    key_set = key_set.difference(run_already)  # remove run already set. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "705274ec",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T20:42:28.182847Z",
     "start_time": "2023-03-08T20:42:28.164736Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AAAA',\n",
       " 'AAAC',\n",
       " 'AAAG',\n",
       " 'AAAT',\n",
       " 'AACA',\n",
       " 'AACC',\n",
       " 'AACG',\n",
       " 'AACT',\n",
       " 'AAGA',\n",
       " 'AAGC',\n",
       " 'AAGG',\n",
       " 'AAGT',\n",
       " 'AATA',\n",
       " 'AATC',\n",
       " 'AATG',\n",
       " 'AATT',\n",
       " 'ACAA',\n",
       " 'ACAC',\n",
       " 'ACAG',\n",
       " 'ACAT',\n",
       " 'ACCA',\n",
       " 'ACCC',\n",
       " 'ACCG',\n",
       " 'ACCT',\n",
       " 'ACGA',\n",
       " 'ACGC',\n",
       " 'ACGG',\n",
       " 'ACGT',\n",
       " 'ACTA',\n",
       " 'ACTC',\n",
       " 'ACTG',\n",
       " 'ACTT',\n",
       " 'AGAA',\n",
       " 'AGAC',\n",
       " 'AGAG',\n",
       " 'AGAT',\n",
       " 'AGCA',\n",
       " 'AGCC',\n",
       " 'AGCG',\n",
       " 'AGCT',\n",
       " 'AGGA',\n",
       " 'AGGC',\n",
       " 'AGGG',\n",
       " 'AGGT',\n",
       " 'AGTA',\n",
       " 'AGTC',\n",
       " 'AGTG',\n",
       " 'AGTT',\n",
       " 'ATAA',\n",
       " 'ATAC',\n",
       " 'ATAG',\n",
       " 'ATAT',\n",
       " 'ATCA',\n",
       " 'ATCC',\n",
       " 'ATCG',\n",
       " 'ATCT',\n",
       " 'ATGA',\n",
       " 'ATGC',\n",
       " 'ATGG',\n",
       " 'ATGT',\n",
       " 'ATTA',\n",
       " 'ATTC',\n",
       " 'ATTG',\n",
       " 'ATTT',\n",
       " 'CAAA',\n",
       " 'CAAC',\n",
       " 'CAAG',\n",
       " 'CAAT',\n",
       " 'CACA',\n",
       " 'CACC',\n",
       " 'CACG',\n",
       " 'CACT',\n",
       " 'CAGA',\n",
       " 'CAGC',\n",
       " 'CAGG',\n",
       " 'CAGT',\n",
       " 'CATA',\n",
       " 'CATC',\n",
       " 'CATG',\n",
       " 'CATT',\n",
       " 'CCAA',\n",
       " 'CCAC',\n",
       " 'CCAG',\n",
       " 'CCAT',\n",
       " 'CCCA',\n",
       " 'CCCC',\n",
       " 'CCCG',\n",
       " 'CCCT',\n",
       " 'CCGA',\n",
       " 'CCGC',\n",
       " 'CCGG',\n",
       " 'CCGT',\n",
       " 'CCTA',\n",
       " 'CCTC',\n",
       " 'CCTG',\n",
       " 'CCTT',\n",
       " 'CGAA',\n",
       " 'CGAC',\n",
       " 'CGAG',\n",
       " 'CGAT',\n",
       " 'CGCA',\n",
       " 'CGCC',\n",
       " 'CGCG',\n",
       " 'CGCT',\n",
       " 'CGGA',\n",
       " 'CGGC',\n",
       " 'CGGG',\n",
       " 'CGGT',\n",
       " 'CGTA',\n",
       " 'CGTC',\n",
       " 'CGTG',\n",
       " 'CGTT',\n",
       " 'CTAA',\n",
       " 'CTAC',\n",
       " 'CTAG',\n",
       " 'CTAT',\n",
       " 'CTCA',\n",
       " 'CTCC',\n",
       " 'CTCG',\n",
       " 'CTCT',\n",
       " 'CTGA',\n",
       " 'CTGC',\n",
       " 'CTGG',\n",
       " 'CTGT',\n",
       " 'CTTA',\n",
       " 'CTTC',\n",
       " 'CTTG',\n",
       " 'CTTT',\n",
       " 'GAAA',\n",
       " 'GAAC',\n",
       " 'GAAG',\n",
       " 'GAAT',\n",
       " 'GACA',\n",
       " 'GACC',\n",
       " 'GACG',\n",
       " 'GACT',\n",
       " 'GAGA',\n",
       " 'GAGC',\n",
       " 'GAGG',\n",
       " 'GAGT',\n",
       " 'GATA',\n",
       " 'GATC',\n",
       " 'GATG',\n",
       " 'GATT',\n",
       " 'GCAA',\n",
       " 'GCAC',\n",
       " 'GCAG',\n",
       " 'GCAT',\n",
       " 'GCCA',\n",
       " 'GCCC',\n",
       " 'GCCG',\n",
       " 'GCCT',\n",
       " 'GCGA',\n",
       " 'GCGC',\n",
       " 'GCGG',\n",
       " 'GCGT',\n",
       " 'GCTA',\n",
       " 'GCTC',\n",
       " 'GCTG',\n",
       " 'GCTT',\n",
       " 'GGAA',\n",
       " 'GGAC',\n",
       " 'GGAG',\n",
       " 'GGAT',\n",
       " 'GGCA',\n",
       " 'GGCC',\n",
       " 'GGCG',\n",
       " 'GGCT',\n",
       " 'GGGA',\n",
       " 'GGGC',\n",
       " 'GGGG',\n",
       " 'GGGT',\n",
       " 'GGTA',\n",
       " 'GGTC',\n",
       " 'GGTG',\n",
       " 'GGTT',\n",
       " 'GTAA',\n",
       " 'GTAC',\n",
       " 'GTAG',\n",
       " 'GTAT',\n",
       " 'GTCA',\n",
       " 'GTCC',\n",
       " 'GTCG',\n",
       " 'GTCT',\n",
       " 'GTGA',\n",
       " 'GTGC',\n",
       " 'GTGG',\n",
       " 'GTGT',\n",
       " 'GTTA',\n",
       " 'GTTC',\n",
       " 'GTTG',\n",
       " 'GTTT',\n",
       " 'TAAA',\n",
       " 'TAAC',\n",
       " 'TAAG',\n",
       " 'TAAT',\n",
       " 'TACA',\n",
       " 'TACC',\n",
       " 'TACG',\n",
       " 'TACT',\n",
       " 'TAGA',\n",
       " 'TAGC',\n",
       " 'TAGG',\n",
       " 'TAGT',\n",
       " 'TATA',\n",
       " 'TATC',\n",
       " 'TATG',\n",
       " 'TATT',\n",
       " 'TCAA',\n",
       " 'TCAC',\n",
       " 'TCAG',\n",
       " 'TCAT',\n",
       " 'TCCA',\n",
       " 'TCCC',\n",
       " 'TCCG',\n",
       " 'TCCT',\n",
       " 'TCGA',\n",
       " 'TCGC',\n",
       " 'TCGG',\n",
       " 'TCGT',\n",
       " 'TCTA',\n",
       " 'TCTC',\n",
       " 'TCTG',\n",
       " 'TCTT',\n",
       " 'TGAA',\n",
       " 'TGAC',\n",
       " 'TGAG',\n",
       " 'TGAT',\n",
       " 'TGCA',\n",
       " 'TGCC',\n",
       " 'TGCG',\n",
       " 'TGCT',\n",
       " 'TGGA',\n",
       " 'TGGC',\n",
       " 'TGGG',\n",
       " 'TGGT',\n",
       " 'TGTA',\n",
       " 'TGTC',\n",
       " 'TGTG',\n",
       " 'TGTT',\n",
       " 'TTAA',\n",
       " 'TTAC',\n",
       " 'TTAG',\n",
       " 'TTAT',\n",
       " 'TTCA',\n",
       " 'TTCC',\n",
       " 'TTCG',\n",
       " 'TTCT',\n",
       " 'TTGA',\n",
       " 'TTGC',\n",
       " 'TTGG',\n",
       " 'TTGT',\n",
       " 'TTTA',\n",
       " 'TTTC',\n",
       " 'TTTG',\n",
       " 'TTTT'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "key_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b66097ae",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T20:46:05.320675Z",
     "start_time": "2023-03-08T20:42:37.624108Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "    partial_keymerSpace = partial(countKeymerSpace, WINDOW_SIZE, OUTDIR)\n",
    "\n",
    "    #4 run partial to write kmer dictionaries summed across sequences. \n",
    "    [partial_keymerSpace(key) for key in key_set]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2bd79bb1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T20:50:14.982007Z",
     "start_time": "2023-03-08T20:50:00.560487Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "    os.chdir(OUTDIR)\n",
    "\n",
    "    # concat KMERS\n",
    "    os.system(f\"cat ALL.*.{WINDOW_SIZE}mers.csv > ALL.{WINDOW_SIZE}mers-kmers.csv && gzip ALL.{WINDOW_SIZE}mers-kmers.csv\")\n",
    "\n",
    "    # concat Nullomers\n",
    "    os.system(f\"cat ALL.*.{WINDOW_SIZE}mers-nullomers.csv > ALL.{WINDOW_SIZE}mers-nullomers.csv && gzip ALL.{WINDOW_SIZE}mers-nullomers.csv\")\n",
    "    \n",
    "    # delete supporting files. \n",
    "    #os.system(f\"gzip ALL.*.{WINDOW_SIZE}mers.csv\")\n",
    "    #os.system(f\"gzip ALL.*.{WINDOW_SIZE}mers-nullomers.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "df643bd2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T20:49:56.887190Z",
     "start_time": "2023-03-08T20:49:56.880438Z"
    }
   },
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "expected an indented block after 'if' statement on line 15 (2814680114.py, line 16)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn [12], line 16\u001b[0;36m\u001b[0m\n\u001b[0;31m    main(sys.argv[1:])\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m expected an indented block after 'if' statement on line 15\n"
     ]
    }
   ],
   "source": [
    "if __name__ == '__main__':\n",
    "    main(sys.argv[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "185699b4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (wynenv)",
   "language": "python",
   "name": "wynenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
