{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3763188d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T21:51:00.525144Z",
     "start_time": "2023-03-08T21:51:00.516926Z"
    }
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "# coding: utf-8\n",
    "\"\"\"\n",
    "sarahfong\n",
    "\n",
    "Summary\n",
    "    - count kmers per chromosome\n",
    "    - split kmer into multiple files where the first N letters of the kmer is the \"key\"\n",
    "    - one file per key will be written w/ counts of sequences associated w/ that key + value\n",
    "\n",
    "Example\n",
    "    Say we find an 11mer, GCGTACGTACG, that appears 15025 times on chrN. The key length is 4. \n",
    "        The key is the first 4 letters \"GCGT\". \n",
    "        The value, or other 7 letters, \"ACGTACG\" will be written to chrN.GCGT.csv\n",
    "        The 11mer counts for GCGTACGTACG, will be written to chrN.GCGT.csv\n",
    "\n",
    "        The resulting file will look like this:\n",
    "\n",
    "        ./chrN.GCGT.csv\n",
    "            ACGTACG, 15025\n",
    "Limitations\n",
    "    - Skips sequences with \"Ns\"\n",
    "            \n",
    "\"\"\"\n",
    "\n",
    "import argparse\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "from Bio.Seq import Seq\n",
    "import glob\n",
    "import gzip\n",
    "from itertools import product\n",
    "import numpy as np\n",
    "import os, sys\n",
    "\n",
    "from multiprocessing import Pool, cpu_count\n",
    "\n",
    "from timeit import default_timer as timer\n",
    "from functools import partial\n",
    "# # parse arguments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "644c9a43",
   "metadata": {},
   "outputs": [],
   "source": [
    "arg_parser = argparse.ArgumentParser()\n",
    "\n",
    "arg_parser.add_argument(\"chromosome\", type=str, help='chromosome number')\n",
    "arg_parser.add_argument(\"directory\", type=str, help='directory where chromosom.fa lives')\n",
    "arg_parser.add_argument(\"length\", type=int, help='kmer length')\n",
    "arg_parser.add_argument(\"keysize\", type=int, help='kmer key length for storing info')\n",
    "\n",
    "args = arg_parser.parse_args()\n",
    "\n",
    "CHR_NUM, PATH, WINDOW_SIZE, KEYSIZE = args.chromosome, args.directory, args.length, args.keysize\n",
    "\n",
    "# make outdirectory for\n",
    "OUTDIR = os.path.join(PATH, \"kmers\", f\"{WINDOW_SIZE}mers\")\n",
    "\n",
    "print(\"OUTDIR\", OUTDIR) \n",
    "if os.path.exists(OUTDIR) is False:\n",
    "    os.mkdir(OUTDIR) \n",
    "\n",
    "# parse if chromosome number should be x, y from array input. \n",
    "if CHR_NUM == \"23\":\n",
    "    CHR_NUM = \"X\"\n",
    "elif CHR_NUM == \"24\":\n",
    "    CHR_NUM = \"Y\"\n",
    "\n",
    "\n",
    "# # functions\n",
    "# ## extract fa sequence\n",
    "\n",
    "# In[ ]:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebcbf546",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T21:51:02.834840Z",
     "start_time": "2023-03-08T21:51:02.828147Z"
    }
   },
   "outputs": [],
   "source": [
    "CHR_NUM, PATH, WINDOW_SIZE, KEYSIZE  = \"hg38\", \"/wynton/home/ahituv/fongsl/dna/hg38/\", 12, 4\n",
    "chr_num, path, window_size, keysize  = CHR_NUM, PATH, WINDOW_SIZE, KEYSIZE\n",
    "OUTDIR = os.path.join(PATH, \"kmers\", f\"{WINDOW_SIZE}mers\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eec35b4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T21:55:01.328963Z",
     "start_time": "2023-03-08T21:55:01.318448Z"
    }
   },
   "outputs": [],
   "source": [
    "def extractFaSeq(chr_num, path):\n",
    "    \"\"\"\n",
    "    get chr-specific sequence, reverse complement\n",
    "\n",
    "    require\n",
    "        Bio.SeqIO.FastaIO - SimpleFastaParser\n",
    "\n",
    "    input\n",
    "        chr_num (str) - chromosome number e.g. 1,2,3,X,Y\n",
    "        path (str) - path to .fa files\n",
    "\n",
    "    method\n",
    "        1. get fa.gz name\n",
    "        2. open fa file\n",
    "        3. get sequence w/ simpleFastaParser\n",
    "        4. get reverse complement by turning seq into Seq object\n",
    "\n",
    "    return \n",
    "        seq (seq record) - str of sequence (should be one per chromosome)\n",
    "        rev (seq record) - str of reverse complement sequence (should be one per chromosome)\n",
    "    \"\"\"\n",
    "\n",
    "    print(\"reading fa\", str(chr_num) +\".fa.gz\")\n",
    "\n",
    "    #1\n",
    "    if \"h\" in chr_num:\n",
    "        fa = os.path.join(path, f\"{chr_num}.fa.gz\")\n",
    "    else:\n",
    "        fa = os.path.join(path,\"chromosomes\", f\"chr{chr_num}.fa.gz\")\n",
    "\n",
    "    #2 open\n",
    "    with gzip.open(fa, \"rt\") as handle:\n",
    "        seq, rev = \"\", \"\"\n",
    "        #3  read using simpleFastaParser\n",
    "        for val in SimpleFastaParser(handle):\n",
    "            seq_id, seq_ = val\n",
    "            seq +=seq_\n",
    "\n",
    "            #4\n",
    "            rev += str(Seq(seq).reverse_complement())\n",
    "            print(seq_id, len(seq), len(rev))\n",
    "    return seq, rev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8343162a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T21:52:40.529918Z",
     "start_time": "2023-03-08T21:52:40.522387Z"
    }
   },
   "outputs": [],
   "source": [
    "# make key-mer universe\n",
    "\n",
    "\n",
    "def makeKeys(keysize):\n",
    "    \"\"\"\n",
    "    input\n",
    "        keysize (int) - length of kmer-keys to create\n",
    "\n",
    "    require\n",
    "        itertools.product\n",
    "        python list comprehension\n",
    "\n",
    "    method\n",
    "        itertools.product kmers -> join kmers into str -> make list\n",
    "\n",
    "\n",
    "    return\n",
    "        key_set (set) - list of strings. \n",
    "    \"\"\" \n",
    "    key_set = set()\n",
    "    for item in product(\"ACGT\", repeat=keysize):\n",
    "        key_set.add(\"\".join(item))\n",
    "\n",
    "\n",
    "    return key_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acdba85c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T21:52:50.995614Z",
     "start_time": "2023-03-08T21:52:50.990017Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# ## retreive one kmer\n",
    "# \n",
    "# ### !Major question! What to do about N's\n",
    "\n",
    "\n",
    "\n",
    "def getOneKmer(sequence, windowsize, start):\n",
    "\n",
    "    \"\"\"\n",
    "    get 1 kmer; break up sequence into equally sized kmers w sliding window of with windowsize, stepsize\n",
    "    \n",
    "    required packages\n",
    "        - numpy\n",
    "        \n",
    "    inputs\n",
    "        sequence (str) - sequence to break into kmers\n",
    "        windowsize (int) - windowsize to make for sequence\n",
    "        start (int) - index to get sequence\n",
    "\n",
    "        \n",
    "    method\n",
    "        1. get kmer\n",
    "        2. if kmer does not have N in sequence\n",
    "    \n",
    "    return \n",
    "        kmer (str) - windowsize from input sequence\n",
    "        \n",
    "    \"\"\"\n",
    "    #1\n",
    "    kmer = sequence[start:start+windowsize].upper()\n",
    "\n",
    "    #2\n",
    "    if \"N\" not in kmer.upper(): # append sequence windows within range of possible windows\n",
    "\n",
    "        return kmer\n",
    "    \n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb4beafe",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T21:52:58.030204Z",
     "start_time": "2023-03-08T21:52:58.021676Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# ## count kmers in universe\n",
    "\n",
    "def countKmersUniverse(kmer_list, keysize):\n",
    "    \n",
    "    \"\"\"\n",
    "    add kmer counts to chromosome universe dictionary:\n",
    "\n",
    "    input\n",
    "        kmer_list (list) - list of kmers in fasta sequence\n",
    "\n",
    "    method\n",
    "        1. build kmer-universe dict while parsing kmers\n",
    "        2. add kmer counts to dictionary, removing NoneType instances. \n",
    "            2.1 get key-specific dictionary from universe dictionary\n",
    "            2.2 if value sequence is not in key dictionary, add. \n",
    "            2.3 if value sequence has been seen, increase dictionary value count. \n",
    "        3. print timer\n",
    "\n",
    "    return\n",
    "        universe_dict (dictionary) - chromosome universe w/ kmer counts. \n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    universe_dict = {}\n",
    "\n",
    "\n",
    "    start = timer()\n",
    "    \n",
    "    \n",
    "    #2 count items in dictionary\n",
    "    for i in kmer_list:\n",
    "\n",
    "        if i is not None:\n",
    "\n",
    "            key, value = i[:keysize], i[keysize:] # split string into key, value\n",
    "\n",
    "            if key not in universe_dict: # if key is not in universe dictionary\n",
    "                universe_dict[key]={}\n",
    "\n",
    "            #2.1\n",
    "            keydict = universe_dict[key]\n",
    "\n",
    "            #2.2 handle cases when value is not in dictionary\n",
    "            if value not in keydict:\n",
    "                keydict[value] = 1\n",
    "                universe_dict[key] = keydict\n",
    "\n",
    "            else:\n",
    "                keydict[value]+=1\n",
    "                universe_dict[key] = keydict\n",
    "               \n",
    "        else:\n",
    "            continue\n",
    "            \n",
    "    end = timer()\n",
    "\n",
    "    print(\"Counting kmer instances:\", end-start, \"\\n Writing dict\")\n",
    "    \n",
    "    return universe_dict\n",
    "\n",
    "\n",
    "# ## write dictionary\n",
    "\n",
    "def writeDictionary(chr_num, outpath, windowsize, result_dict):\n",
    "    \"\"\"\n",
    "    write kmer universe dictionary to outfile\n",
    "    \n",
    "    input\n",
    "        chr_num (str) - chr number or all\n",
    "        path (str) - path to write outfile\n",
    "        windowsize (int) - size of window used for kmers\n",
    "        results_dict (dictionary) - kmer universe w/ counts of kmer occurrences \n",
    "    \n",
    "    method\n",
    "        1. create the outfile as key of keys\n",
    "        2. write key, value as comma-separated str to outfile\n",
    "        3. close the outfile\n",
    "    \n",
    "    return \n",
    "        out_file (str) - written file name w asterisks for the key\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    #1\n",
    "    for key, value in result_dict.items():\n",
    "        out_file = os.path.join(outpath, f\"{chr_num}.{windowsize}mers.{key}.csv\")\n",
    "        with open(out_file, \"w\") as writer:\n",
    "            for secondkey, count in value.items():\n",
    "        \n",
    "                writer.write(f\"{secondkey},{count}\\n\")\n",
    "        #3\n",
    "        writer.close()\n",
    "    \n",
    "        #print(\"\\n\\nwrote\", out_file)\n",
    "\n",
    "    out_file = os.path.join(outpath, f\"{chr_num}.{windowsize}mers.*.csv\")\n",
    "    return out_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf8cc502",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-08T22:13:10.227767Z",
     "start_time": "2023-03-08T21:55:05.252097Z"
    }
   },
   "outputs": [],
   "source": [
    "# # main\n",
    "\n",
    "\n",
    "def main(argv):\n",
    "    \n",
    "    \"\"\"\n",
    "    per chromosome, write kmer count frequency dictionary as csv\n",
    "    \n",
    "    input\n",
    "        chr_num (str) - chromozome number to process kmers for\n",
    "        path (str) - path to fa files, and to write\n",
    "        window_size (int) - size of window to consider for kmer\n",
    "    \n",
    "    method\n",
    "        1. get sequence +/- from fa\n",
    "        2. generate set of keys to write kmer-space to. Keys will be used to quantify the known-kmer space.\n",
    "        3. start the timer. \n",
    "        4. set up to retrieve partial commands\n",
    "        5. pool forward, reverse kmers from sequence\n",
    "        6. extend and count kmers in genome\n",
    "        7. write dictionary to .csv\n",
    "\n",
    "    return\n",
    "        outcsv (str) - csvs of kmer universe, separated by keys\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    #1 get chr sequence, reverse complement \n",
    "    SEQ, REV = extractFaSeq(CHR_NUM, PATH)\n",
    "\n",
    "\n",
    "    #2 get kmer key set, universe dict\n",
    "    KEY_SET = makeKeys(KEYSIZE)\n",
    "    #universe_kmers = makeKmerUniverse(WINDOW_SIZE, KEY_SET, KEYSIZE)\n",
    "\n",
    "\n",
    "    #3 calculate the number of starting positions \n",
    "    STARTS = list(range(len(SEQ) - WINDOW_SIZE + 1))\n",
    "\n",
    "    #4 setup partial command to retrieve kmers\n",
    "    partial_seq = partial(getOneKmer, SEQ, WINDOW_SIZE)\n",
    "    partial_rev = partial(getOneKmer, REV, WINDOW_SIZE)\n",
    "\n",
    "    #5 get all kmers from chromosome\n",
    "    result_multi = [partial_seq(start) for start in STARTS]\n",
    "    result_multi_r = [partial_rev(start) for start in STARTS]\n",
    "\n",
    "    #6 count knownmers in kmer universe\n",
    "    result_multi.extend(result_multi_r) # make 2 lists into 1 big list\n",
    "\n",
    "    # count items, add to dictionary\n",
    "    genome_kmers = countKmersUniverse(result_multi, KEYSIZE)\n",
    "\n",
    "    #7 write dictionary```\n",
    "    outcsv = writeDictionary(CHR_NUM, OUTDIR, WINDOW_SIZE, genome_kmers)\n",
    "    print(outcsv)\n",
    "    \n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main(sys.argv[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4452b845",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (wynenv)",
   "language": "python",
   "name": "wynenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
