{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a38fcfaa-77c1-4bfb-abe0-05dbc4e4e87d",
   "metadata": {},
   "source": [
    "20240306\n",
    "\n",
    "SarahFong\n",
    "\n",
    "Take output from DiffBind deseq2 formatting\n",
    "\n",
    "          $HOME/EMF/US/bin/ATAC/1_prep_deseq2_ATAC.ipynb \n",
    "          \n",
    "Prepare data: \n",
    "1. Clean diffbind deseq2 conc data (remove bad coordinates)\n",
    "4. Format dataframe for LegNet, multitask, deepstarr multitask\n",
    "5. Split into training and test set - randomly hold out 2 chromosomes. \n",
    "6. Add fold num to training set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "24791ba6-ba9e-47c8-99ff-824a21c53b9b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:13:31.140947Z",
     "start_time": "2024-03-09T00:13:22.469505Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('sans-serif', 'Arial', 18)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "import config_readwrite as crw\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os, sys\n",
    "import pandas as pd\n",
    "\n",
    "from scipy import stats\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import plot_params as pp\n",
    "pp.fonts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4b15ddcf",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:13:31.151608Z",
     "start_time": "2024-03-09T00:13:31.142833Z"
    }
   },
   "outputs": [],
   "source": [
    "config_name =os.path.join(os.getcwd(), \"config.multi.ini\")\n",
    "config, cfn = crw.read(config_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c13717b-4462-468b-aded-c85fa8632585",
   "metadata": {},
   "source": [
    "# functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9c78663a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:13:31.249569Z",
     "start_time": "2024-03-09T00:13:31.154047Z"
    }
   },
   "outputs": [],
   "source": [
    "def setPrefixRules(prefix):\n",
    "    \"\"\" rules for processing datasets\"\"\"\n",
    "    if \"class.all\" in prefix:\n",
    "        QUANTILE_FILTER=False\n",
    "        MIN_CONC_FILTER=False\n",
    "        JOINT_ACCESSIBLE=True\n",
    "        CLASS_LABEL = True\n",
    "\n",
    "    elif \"class.nojoint\" in prefix:\n",
    "        QUANTILE_FILTER=False\n",
    "        MIN_CONC_FILTER=True\n",
    "        JOINT_ACCESSIBLE=False\n",
    "        CLASS_LABEL = True\n",
    "\n",
    "    elif \"reg.all\" in PREFIX:\n",
    "        QUANTILE_FILTER=False\n",
    "        MIN_CONC_FILTER=True\n",
    "        JOINT_ACCESSIBLE=True\n",
    "        CLASS_LABEL = False\n",
    "\n",
    "    elif \"reg.nojoint\" in prefix:\n",
    "        QUANTILE_FILTER=False\n",
    "        MIN_CONC_FILTER=True\n",
    "        JOINT_ACCESSIBLE=False\n",
    "        CLASS_LABEL = False\n",
    "    else:\n",
    "        print('need to add rules for', prefix)\n",
    "        \n",
    "    return QUANTILE_FILTER, MIN_CONC_FILTER, JOINT_ACCESSIBLE, CLASS_LABEL\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ba6c7d3-4107-4315-8748-e13b1108310b",
   "metadata": {},
   "source": [
    "## make chromosome list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "947fbb71-5880-4453-8414-51db88678771",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:13:31.355052Z",
     "start_time": "2024-03-09T00:13:31.251568Z"
    }
   },
   "outputs": [],
   "source": [
    "def chrList():\n",
    "    \"\"\"return  list of chromosomes\"\"\"\n",
    "    \n",
    "    chrs = []\n",
    "    \n",
    "    for n in np.arange(1,23):\n",
    "        chrs.append(f\"chr{n}\")\n",
    "    \n",
    "    # add sex chromosomes\n",
    "    chrs.append(\"chrX\")\n",
    "    chrs.append(\"chrY\")\n",
    "    \n",
    "    return chrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91c587c3",
   "metadata": {},
   "source": [
    "## write fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "84051dad",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:13:32.779771Z",
     "start_time": "2024-03-09T00:13:31.357058Z"
    }
   },
   "outputs": [],
   "source": [
    "def writeFa(heldout_df, heldout_fa):\n",
    "    with open(heldout_fa, \"w\") as writer:\n",
    "        for row in heldout_df.iterrows():\n",
    "            seqid, seq=row[1][:2]\n",
    "            writer.write(f\">{seqid}\\n{seq}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e894b82",
   "metadata": {},
   "source": [
    "## train test split on chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "01b01445",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:13:33.200269Z",
     "start_time": "2024-03-09T00:13:32.781656Z"
    }
   },
   "outputs": [],
   "source": [
    "def splitTrainTestVal(df, val_chr_list=None, test_chr_list=None):\n",
    "    \"\"\"randomly sample and hold out 2 chromosomes for testing, validation\"\"\"\n",
    "    \n",
    "    cols =['coor.type',\"seq\"]\n",
    "    \n",
    "    # randomly sample test chromosomes (n=2)\n",
    "    if test_chr_list is None:\n",
    "        print('randomly sampling chromosomes for test')\n",
    "\n",
    "\n",
    "        # randomly sample\n",
    "        test_chr_list = list(np.random.choice(chrs, 2))\n",
    "   \n",
    "    # randomly sample validation chromosome (n=1)\n",
    "    if val_chr_list is None:\n",
    "        print('randomly sampling chromosomes for val')\n",
    "\n",
    "        # remove test chromosomes\n",
    "        for chr_ in test_chr_list:\n",
    "            chrs.remove(chr_)\n",
    "\n",
    "        # randomly sample\n",
    "        val_chr_list = list(np.random.choice(chrs, 1))\n",
    "\n",
    "    # separate held out chromosomes from  training chromosomes\n",
    "    test = df.loc[df[\"#chr\"].isin(test_chr_list)].copy()\n",
    "    val = df.loc[df[\"#chr\"].isin(val_chr_list)].copy()\n",
    "    train = df.loc[(~df[\"#chr\"].isin(test_chr_list))&\n",
    "                  (~df[\"#chr\"].isin(val_chr_list))].copy()\n",
    "\n",
    "    return train, val, test\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f4dc9fb",
   "metadata": {},
   "source": [
    "## filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f8f99643",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:13:33.364593Z",
     "start_time": "2024-03-09T00:13:33.202333Z"
    }
   },
   "outputs": [],
   "source": [
    "def minReadDepthFilter(table, MIN_CONC, col1, col2):\n",
    "\n",
    "    print(\"before read depth filter:\", table.shape)\n",
    "    table = table.loc[(table[col1] > MIN_CONC) |\n",
    "                      (table[col2] > MIN_CONC)].copy()\n",
    "    print(\"after:\", table.shape)\n",
    "\n",
    "    return table\n",
    "\n",
    "\n",
    "def quantileFilter(table, quantile=0.99):\n",
    "    \"\"\"upper quantile filter for read count values\"\"\"\n",
    "\n",
    "    print(\"before quantile filter:\", table.shape)\n",
    "\n",
    "    # quantiles\n",
    "    ctrl_thresh, us_thresh = table[[\"ctrl\", \"US\"]].quantile(quantile)\n",
    "\n",
    "    # filter table\n",
    "    table = table.loc[(table[\"ctrl\"] < ctrl_thresh) &\n",
    "                      (table[\"US\"] < us_thresh)]\n",
    "\n",
    "    print(\"after:\", table.shape)\n",
    "\n",
    "    return table\n",
    "\n",
    "\n",
    "def jointAccessibleFilter(table):\n",
    "    \"\"\"remove joint accessible regions, scramble new dataframe\"\"\"\n",
    "\n",
    "    print(\"before joint_accessible filter:\", table.shape)\n",
    "\n",
    "    # filter table\n",
    "    ctrl_only = table.loc[(table[\"ctrl\"] > 0) &\n",
    "                          (table[\"US\"] == 0)].copy()\n",
    "\n",
    "    US_only = table.loc[(table[\"ctrl\"] == 0) &\n",
    "                        (table[\"US\"] > 0)].copy()\n",
    "\n",
    "    # combine us and ctrl\n",
    "    # shuffle dataframe\n",
    "    table = pd.concat([ctrl_only, US_only]).sample(\n",
    "        frac=1).reset_index(drop=True)\n",
    "\n",
    "    print(\"after:\", table.shape)\n",
    "\n",
    "    return table\n",
    "\n",
    "# binarize\n",
    "\n",
    "\n",
    "def classLabel(table):\n",
    "    \"\"\"binarize read count column\"\"\"\n",
    "    cols_to_label = ['ctrl', 'US']\n",
    "    for col in cols_to_label:\n",
    "        table[col] = table[col].apply(lambda x: 0 if x == 0 else 1)\n",
    "\n",
    "    return table\n",
    "\n",
    "# scramble df\n",
    "\n",
    "\n",
    "def dfShuffle(df):\n",
    "    \"\"\"scramble dataframe\"\"\"\n",
    "    return df.sample(frac=1).reset_index(drop=True)\n",
    "\n",
    "def directionFilter(table):\n",
    "    \"\"\"filter for positive directional changes in activity\"\"\"\n",
    "\n",
    "    table = table.loc[table[\"US_DIF\"]<0] #where US_DIF = ctrl - US, and US_dif <0 means US has more activity than control\n",
    "    \n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e9ba04a8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:13:33.447041Z",
     "start_time": "2024-03-09T00:13:33.366602Z"
    }
   },
   "outputs": [],
   "source": [
    "def directionFilter(table, col):\n",
    "    \"\"\"filter for positive directional changes in activity\"\"\"\n",
    "\n",
    "    table = table.loc[table[col]<0] #where US_DIF = ctrl - US, and US_dif <0 means US has more activity than control\n",
    "    \n",
    "    return table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6144f3a-643a-42f5-bacc-40d77de7f899",
   "metadata": {},
   "source": [
    "# Main"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b85ac20f",
   "metadata": {},
   "source": [
    "## load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4dc1b99b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:13:33.629054Z",
     "start_time": "2024-03-09T00:13:33.450742Z"
    }
   },
   "outputs": [],
   "source": [
    "collection_dict = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "93862edd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:14:40.197636Z",
     "start_time": "2024-03-09T00:14:25.882326Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(80892, 10)\n",
      "before read depth filter: (80892, 10)\n",
      "after: (80892, 10)\n",
      "(2823, 10) (52606, 10) (2472, 10)\n",
      "(27488, 10)\n",
      "before read depth filter: (27488, 10)\n",
      "after: (27488, 10)\n",
      "(551, 10) (9920, 10) (372, 10)\n",
      "(55241, 10)\n",
      "before read depth filter: (55241, 10)\n",
      "after: (55241, 10)\n",
      "(1449, 10) (27766, 10) (1119, 10)\n",
      "(41641, 10)\n",
      "before read depth filter: (41641, 10)\n",
      "after: (41641, 10)\n",
      "(751, 10) (13320, 10) (639, 10)\n"
     ]
    }
   ],
   "source": [
    "CLS = [\"hob\", \"hepg2\", \"k562\", \"bj\"]\n",
    "for CL in CLS:\n",
    "\n",
    "    PREFIX = \"reg.all.8tasks\"\n",
    "\n",
    "    PATH = \"/wynton/group/ahituv/data/US-MPRA/ATAC-seq/Diffbind_results\"\n",
    "    DATA_PATH = f\"/wynton/home/ahituv/fongsl/EMF/US/ml_emf/data/deepstarr/deseq2/{CL}/{PREFIX}\"\n",
    "\n",
    "    TEST_CHR = [\"chr8\"]\n",
    "    VAL_CHR = \"chr12\"\n",
    "\n",
    "    # peak information\n",
    "    PEAK_SIZE = 270\n",
    "    MIN_CONC = 2\n",
    "\n",
    "    # genome information\n",
    "    HG38 = \"/wynton/group/ahituv/data/dna/hg38/hg38.chrom.sizes\"\n",
    "    FA_HG38 = \"/wynton/group/ahituv/data/dna/hg38/hg38.fa\"\n",
    "\n",
    "    cols = ['coor.type', \"seq\", \"ctrl\", \"US\", \"US_DIF\", \"fold_num\"]\n",
    "    fa_cols = ['coor.type', \"seq\"]\n",
    "\n",
    "    if os.path.exists(DATA_PATH) is False:\n",
    "        if os.path.dirname(DATA_PATH) is False:\n",
    "            os.mkdir(os.path.dirname(DATA_PATH))\n",
    "        os.mkdir(DATA_PATH)\n",
    "\n",
    "    # base files\n",
    "\n",
    "    # files\n",
    "\n",
    "    # DESEQ2 information\n",
    "    DIFF = f'./diffbind_results/{CL}_deseq2.csv'\n",
    "    NODIFF = f'./diffbind_results/{CL}_deseq2-nodiff.csv'\n",
    "    DIFF_BED = \"./\" + DIFF.strip(\".csv\") + \".bed\"\n",
    "\n",
    "    # write training, test files\n",
    "    FULL = f\"{CL}_deseq2-nondiff.trimmed.full.csv\"\n",
    "\n",
    "    # base config\n",
    "\n",
    "    section = f\"{CL}-ATAC-DESEQ2\"\n",
    "    crw.check(config, section)\n",
    "\n",
    "    config[section][\"path\"] = PATH\n",
    "\n",
    "    config[section][\"nondiff_bind_results\"] = \"%(path)s/\" + NODIFF\n",
    "    config[section][\"diff_bind_results\"] = \"%(path)s/\" + DIFF\n",
    "    config[section][\"diff_bind_results_bed\"] = \"%(path)s/\" + DIFF_BED\n",
    "\n",
    "    # deepstarr files\n",
    "\n",
    "    # file inputs for training, testing\n",
    "    # x\n",
    "    TRAIN_FA = f\"{PREFIX}.Sequences_Train.fa\"\n",
    "    VAL_FA = f\"{PREFIX}.Sequences_Val.fa\"\n",
    "    TEST_FA = f\"{PREFIX}.Sequences_Test.fa\"\n",
    "\n",
    "    # y\n",
    "    TRAIN_TARGET = f\"{PREFIX}.Sequences_activity_Train.txt\"\n",
    "    VAL_TARGET = f\"{PREFIX}.Sequences_activity_Val.txt\"\n",
    "    TEST_TARGET = f\"{PREFIX}.Sequences_activity_Test.txt\"\n",
    "\n",
    "    # rules\n",
    "\n",
    "    # deepstarr config\n",
    "\n",
    "    section = f\"{CL}.atac.deseq2.deepstarr\"\n",
    "    crw.check(config, section)\n",
    "\n",
    "    # deepstar config\n",
    "    config[section][\"data_path\"] = DATA_PATH\n",
    "    config[section][\"held_out_chr\"] = \",\".join(TEST_CHR)\n",
    "    config[section][\"val_chr\"] = VAL_CHR\n",
    "\n",
    "    # deepstarr+prefix config\n",
    "\n",
    "    # rules for data\n",
    "    QUANTILE_FILTER, MIN_CONC_FILTER, JOINT_ACCESSIBLE, CLASS_LABEL = setPrefixRules(\n",
    "        PREFIX)\n",
    "\n",
    "    DIRECTION_FILTER = True\n",
    "\n",
    "    # deepstarr prefix config\n",
    "    section = f\"Hepg2.atac.deseq2.deepstarr.{PREFIX}\"\n",
    "    crw.check(config, section)\n",
    "\n",
    "    config[section][\"train_fa\"] = TRAIN_FA\n",
    "    config[section][\"val_fa\"] = VAL_FA\n",
    "    config[section][\"test_fa\"] = TEST_FA\n",
    "\n",
    "    config[section][\"train_target\"] = TRAIN_TARGET\n",
    "    config[section][\"tval_target\"] = VAL_TARGET\n",
    "    config[section][\"test_target\"] = TEST_TARGET\n",
    "\n",
    "    config[section][\"filter_MIN_CONC\"] = str(MIN_CONC_FILTER)\n",
    "    if MIN_CONC_FILTER is True:\n",
    "        config[section][\"MIN_CONC\"] = str(MIN_CONC)\n",
    "\n",
    "    config[section][\"filter_quantile\"] = str(QUANTILE_FILTER)\n",
    "    config[section][\"filter_jointaccessible\"] = str(JOINT_ACCESSIBLE)\n",
    "    config[section][\"classlabel\"] = str(CLASS_LABEL)\n",
    "\n",
    "    crw.write(config, cfn)\n",
    "\n",
    "    # deal with data tables\n",
    "    os.chdir(PATH)\n",
    "    table = pd.read_csv(FULL, sep='\\t').drop_duplicates()\n",
    "\n",
    "    table[\"coor.type\"] = table[\"type\"] + \"|\" + table[\"seq.id\"]\n",
    "\n",
    "    if \"Conc_Ultrasound_dif\" not in list(table):\n",
    "        table[\"Conc_Ultrasound_dif\"] = table[\"Conc_Control\"] - \\\n",
    "            table[\"Conc_Ultrasound\"]\n",
    "\n",
    "    # rename columns w cl annotated\n",
    "    us_col, ctrl_col, dif_col = f\"US.{CL}\", f\"ctrl.{CL}\", f\"US_DIF.{CL}\"\n",
    "    \n",
    "    table.rename(columns={\"Conc_Ultrasound\": us_col,\n",
    "                          \"Conc_Control\": ctrl_col,\n",
    "                          \"Conc_Ultrasound_dif\": dif_col}, inplace=True)\n",
    "\n",
    "    print(table.shape)\n",
    "\n",
    "    # min read depth filter\n",
    "\n",
    "    # mean read depth filter\n",
    "    if MIN_CONC_FILTER is True:\n",
    "        table = minReadDepthFilter(table, MIN_CONC, us_col, ctrl_col)\n",
    "\n",
    "    # quantile filter\n",
    "    if QUANTILE_FILTER is True:\n",
    "\n",
    "        table = quantileFilter(table, quantile=0.99)\n",
    "\n",
    "    # joint accessible filter\n",
    "    if JOINT_ACCESSIBLE is False:\n",
    "        table = jointAccessibleFilter(table)\n",
    "\n",
    "    # apply class label\n",
    "    if CLASS_LABEL is True:\n",
    "        table = classLabel(table)\n",
    "        print(table.groupby([us_col, ctrl_col, dif_col])['#chr'].count())\n",
    "\n",
    "    if DIRECTION_FILTER is True:\n",
    "        table = directionFilter(table, dif_col)\n",
    "\n",
    "    # train on all atac peaks\n",
    "\n",
    "    os.chdir(DATA_PATH)\n",
    "\n",
    "    table = dfShuffle(table)  # shuffle the table before splitting\n",
    "\n",
    "    train, val, test = splitTrainTestVal(\n",
    "        table, val_chr_list=[VAL_CHR], test_chr_list=TEST_CHR)\n",
    "\n",
    "    writeFa(test[fa_cols], TEST_FA)\n",
    "    writeFa(train[fa_cols], TRAIN_FA)\n",
    "    writeFa(val[fa_cols], VAL_FA)\n",
    "\n",
    "    cols = [\"coor.type\",\n",
    "            # us_col,\n",
    "            ctrl_col, \n",
    "            dif_col \n",
    "           ]\n",
    "    test[cols].to_csv(TEST_TARGET, sep='\\t', index=False)\n",
    "    train[cols].to_csv(TRAIN_TARGET, sep='\\t', index=False)\n",
    "    val[cols].to_csv(VAL_TARGET, sep='\\t', index=False)\n",
    "\n",
    "    print(val.shape, train.shape, test.shape)\n",
    "\n",
    "    collection_dict[CL] = table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c3a6537",
   "metadata": {},
   "source": [
    "# combine all cell line ATAC data together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f8521dea",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:49:54.349544Z",
     "start_time": "2024-03-09T00:49:54.164153Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>start_trim</th>\n",
       "      <th>end_trim</th>\n",
       "      <th>type</th>\n",
       "      <th>seq.id</th>\n",
       "      <th>US.hob</th>\n",
       "      <th>ctrl.hob</th>\n",
       "      <th>US_DIF.hob</th>\n",
       "      <th>seq</th>\n",
       "      <th>coor.type</th>\n",
       "      <th>US.hepg2</th>\n",
       "      <th>ctrl.hepg2</th>\n",
       "      <th>US_DIF.hepg2</th>\n",
       "      <th>US.k562</th>\n",
       "      <th>ctrl.k562</th>\n",
       "      <th>US_DIF.k562</th>\n",
       "      <th>US.bj</th>\n",
       "      <th>ctrl.bj</th>\n",
       "      <th>US_DIF.bj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48492419</td>\n",
       "      <td>48492690</td>\n",
       "      <td>hob.146430</td>\n",
       "      <td>chrX:48492419-48492690</td>\n",
       "      <td>4.985740</td>\n",
       "      <td>4.964249</td>\n",
       "      <td>-0.021490</td>\n",
       "      <td>CCAGAGACAATGTGGCCAGGCTCCGGAGGGCTGGGAAGATGAGCAA...</td>\n",
       "      <td>hob.146430|chrX:48492419-48492690</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr20</td>\n",
       "      <td>4076708</td>\n",
       "      <td>4076979</td>\n",
       "      <td>hob.47070</td>\n",
       "      <td>chr20:4076708-4076979</td>\n",
       "      <td>4.637884</td>\n",
       "      <td>4.347412</td>\n",
       "      <td>-0.290472</td>\n",
       "      <td>ttttgatgaattccaatttattgatttcgatgttgtggggacaaga...</td>\n",
       "      <td>hob.47070|chr20:4076708-4076979</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>17591665</td>\n",
       "      <td>17591936</td>\n",
       "      <td>hob.142196</td>\n",
       "      <td>chr17:17591665-17591936</td>\n",
       "      <td>8.482293</td>\n",
       "      <td>8.470101</td>\n",
       "      <td>-0.012192</td>\n",
       "      <td>CAGCCACGCGCCCCCGGAACCGGACCTATAGAGCCGGGTAAGTGCC...</td>\n",
       "      <td>hob.142196|chr17:17591665-17591936</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6</td>\n",
       "      <td>106746723</td>\n",
       "      <td>106746994</td>\n",
       "      <td>hob.128551</td>\n",
       "      <td>chr6:106746723-106746994</td>\n",
       "      <td>4.958441</td>\n",
       "      <td>4.903755</td>\n",
       "      <td>-0.054687</td>\n",
       "      <td>AGATGACTGCTTACAATACACAGCTTTCATATAGGGAGTGAGTGTC...</td>\n",
       "      <td>hob.128551|chr6:106746723-106746994</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr5</td>\n",
       "      <td>177476824</td>\n",
       "      <td>177477095</td>\n",
       "      <td>hob.11791</td>\n",
       "      <td>chr5:177476824-177477095</td>\n",
       "      <td>5.505728</td>\n",
       "      <td>5.103355</td>\n",
       "      <td>-0.402373</td>\n",
       "      <td>TTCTGCATGCAGGGGTGAGGTGGGCTGGAGTCTGATCAGAAGTTGC...</td>\n",
       "      <td>hob.11791|chr5:177476824-177477095</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14705</th>\n",
       "      <td>chr14</td>\n",
       "      <td>58864931</td>\n",
       "      <td>58865202</td>\n",
       "      <td>bj.47457</td>\n",
       "      <td>chr14:58864931-58865202</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>TTTCCTGGAAGTTGAGAAGCGTCTTTAGTCTACTGTGTTGACAGCA...</td>\n",
       "      <td>bj.47457|chr14:58864931-58865202</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.281995</td>\n",
       "      <td>3.928428</td>\n",
       "      <td>-0.353567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14706</th>\n",
       "      <td>chr5</td>\n",
       "      <td>1523913</td>\n",
       "      <td>1524184</td>\n",
       "      <td>bj.29040</td>\n",
       "      <td>chr5:1523913-1524184</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cgcgggtctcggggcgcgggccgaggatgcgcggcggctggagcgg...</td>\n",
       "      <td>bj.29040|chr5:1523913-1524184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.836365</td>\n",
       "      <td>4.626367</td>\n",
       "      <td>-0.209998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14707</th>\n",
       "      <td>chr9</td>\n",
       "      <td>91887496</td>\n",
       "      <td>91887767</td>\n",
       "      <td>bj.51645</td>\n",
       "      <td>chr9:91887496-91887767</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ccagtcaataaatatatcacaagttttttcccactgttctgtgggt...</td>\n",
       "      <td>bj.51645|chr9:91887496-91887767</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.329817</td>\n",
       "      <td>4.137622</td>\n",
       "      <td>-0.192196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14708</th>\n",
       "      <td>chr11</td>\n",
       "      <td>46347175</td>\n",
       "      <td>46347446</td>\n",
       "      <td>bj.66233</td>\n",
       "      <td>chr11:46347175-46347446</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GGGATCGGGGGAGGAAAGATGCGCGTCTGGATGCGCGCAGTGCGAG...</td>\n",
       "      <td>bj.66233|chr11:46347175-46347446</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.883983</td>\n",
       "      <td>5.794700</td>\n",
       "      <td>-0.089284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14709</th>\n",
       "      <td>chr12</td>\n",
       "      <td>93785300</td>\n",
       "      <td>93785571</td>\n",
       "      <td>bj.114910</td>\n",
       "      <td>chr12:93785300-93785571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>AATTGAAAATGATTCAAACTGATTTCCTCTGACGAACATCCGAAAT...</td>\n",
       "      <td>bj.114910|chr12:93785300-93785571</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.436081</td>\n",
       "      <td>5.392410</td>\n",
       "      <td>-0.043671</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>113788 rows Ã— 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        #chr  start_trim   end_trim        type                    seq.id  \\\n",
       "0       chrX    48492419   48492690  hob.146430    chrX:48492419-48492690   \n",
       "1      chr20     4076708    4076979   hob.47070     chr20:4076708-4076979   \n",
       "2      chr17    17591665   17591936  hob.142196   chr17:17591665-17591936   \n",
       "3       chr6   106746723  106746994  hob.128551  chr6:106746723-106746994   \n",
       "4       chr5   177476824  177477095   hob.11791  chr5:177476824-177477095   \n",
       "...      ...         ...        ...         ...                       ...   \n",
       "14705  chr14    58864931   58865202    bj.47457   chr14:58864931-58865202   \n",
       "14706   chr5     1523913    1524184    bj.29040      chr5:1523913-1524184   \n",
       "14707   chr9    91887496   91887767    bj.51645    chr9:91887496-91887767   \n",
       "14708  chr11    46347175   46347446    bj.66233   chr11:46347175-46347446   \n",
       "14709  chr12    93785300   93785571   bj.114910   chr12:93785300-93785571   \n",
       "\n",
       "         US.hob  ctrl.hob  US_DIF.hob  \\\n",
       "0      4.985740  4.964249   -0.021490   \n",
       "1      4.637884  4.347412   -0.290472   \n",
       "2      8.482293  8.470101   -0.012192   \n",
       "3      4.958441  4.903755   -0.054687   \n",
       "4      5.505728  5.103355   -0.402373   \n",
       "...         ...       ...         ...   \n",
       "14705  0.000000  0.000000    0.000000   \n",
       "14706  0.000000  0.000000    0.000000   \n",
       "14707  0.000000  0.000000    0.000000   \n",
       "14708  0.000000  0.000000    0.000000   \n",
       "14709  0.000000  0.000000    0.000000   \n",
       "\n",
       "                                                     seq  \\\n",
       "0      CCAGAGACAATGTGGCCAGGCTCCGGAGGGCTGGGAAGATGAGCAA...   \n",
       "1      ttttgatgaattccaatttattgatttcgatgttgtggggacaaga...   \n",
       "2      CAGCCACGCGCCCCCGGAACCGGACCTATAGAGCCGGGTAAGTGCC...   \n",
       "3      AGATGACTGCTTACAATACACAGCTTTCATATAGGGAGTGAGTGTC...   \n",
       "4      TTCTGCATGCAGGGGTGAGGTGGGCTGGAGTCTGATCAGAAGTTGC...   \n",
       "...                                                  ...   \n",
       "14705  TTTCCTGGAAGTTGAGAAGCGTCTTTAGTCTACTGTGTTGACAGCA...   \n",
       "14706  cgcgggtctcggggcgcgggccgaggatgcgcggcggctggagcgg...   \n",
       "14707  ccagtcaataaatatatcacaagttttttcccactgttctgtgggt...   \n",
       "14708  GGGATCGGGGGAGGAAAGATGCGCGTCTGGATGCGCGCAGTGCGAG...   \n",
       "14709  AATTGAAAATGATTCAAACTGATTTCCTCTGACGAACATCCGAAAT...   \n",
       "\n",
       "                                 coor.type  US.hepg2  ctrl.hepg2  \\\n",
       "0        hob.146430|chrX:48492419-48492690       0.0         0.0   \n",
       "1          hob.47070|chr20:4076708-4076979       0.0         0.0   \n",
       "2       hob.142196|chr17:17591665-17591936       0.0         0.0   \n",
       "3      hob.128551|chr6:106746723-106746994       0.0         0.0   \n",
       "4       hob.11791|chr5:177476824-177477095       0.0         0.0   \n",
       "...                                    ...       ...         ...   \n",
       "14705     bj.47457|chr14:58864931-58865202       0.0         0.0   \n",
       "14706        bj.29040|chr5:1523913-1524184       0.0         0.0   \n",
       "14707      bj.51645|chr9:91887496-91887767       0.0         0.0   \n",
       "14708     bj.66233|chr11:46347175-46347446       0.0         0.0   \n",
       "14709    bj.114910|chr12:93785300-93785571       0.0         0.0   \n",
       "\n",
       "       US_DIF.hepg2  US.k562  ctrl.k562  US_DIF.k562     US.bj   ctrl.bj  \\\n",
       "0               0.0      0.0        0.0          0.0  0.000000  0.000000   \n",
       "1               0.0      0.0        0.0          0.0  0.000000  0.000000   \n",
       "2               0.0      0.0        0.0          0.0  0.000000  0.000000   \n",
       "3               0.0      0.0        0.0          0.0  0.000000  0.000000   \n",
       "4               0.0      0.0        0.0          0.0  0.000000  0.000000   \n",
       "...             ...      ...        ...          ...       ...       ...   \n",
       "14705           0.0      0.0        0.0          0.0  4.281995  3.928428   \n",
       "14706           0.0      0.0        0.0          0.0  4.836365  4.626367   \n",
       "14707           0.0      0.0        0.0          0.0  4.329817  4.137622   \n",
       "14708           0.0      0.0        0.0          0.0  5.883983  5.794700   \n",
       "14709           0.0      0.0        0.0          0.0  5.436081  5.392410   \n",
       "\n",
       "       US_DIF.bj  \n",
       "0       0.000000  \n",
       "1       0.000000  \n",
       "2       0.000000  \n",
       "3       0.000000  \n",
       "4       0.000000  \n",
       "...          ...  \n",
       "14705  -0.353567  \n",
       "14706  -0.209998  \n",
       "14707  -0.192196  \n",
       "14708  -0.089284  \n",
       "14709  -0.043671  \n",
       "\n",
       "[113788 rows x 19 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table = pd.concat(collection_dict.values()).fillna(0)\n",
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9d82b78b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:50:10.678896Z",
     "start_time": "2024-03-09T00:50:10.437377Z"
    }
   },
   "outputs": [],
   "source": [
    "PREFIX = \"reg.all.8task\"\n",
    "DATA_PATH = f\"/wynton/home/ahituv/fongsl/EMF/US/ml_emf/data/deepstarr/deseq2/all/{PREFIX}\"\n",
    "\n",
    "if os.path.exists(DATA_PATH) is False:\n",
    "    os.mkdir(DATA_PATH)\n",
    "    \n",
    "os.chdir(DATA_PATH)\n",
    "\n",
    "table = dfShuffle(table)  # shuffle the table before splitting\n",
    "\n",
    "train, val, test = splitTrainTestVal(\n",
    "    table, val_chr_list=[VAL_CHR], test_chr_list=TEST_CHR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ede51d7d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-09T00:50:24.186572Z",
     "start_time": "2024-03-09T00:50:11.885814Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5574, 19) (103612, 19) (4602, 19)\n"
     ]
    }
   ],
   "source": [
    "# FIles for all. \n",
    "\n",
    "TRAIN_FA = f\"{PREFIX}.Sequences_Train.fa\"\n",
    "VAL_FA = f\"{PREFIX}.Sequences_Val.fa\"\n",
    "TEST_FA = f\"{PREFIX}.Sequences_Test.fa\"\n",
    "\n",
    "# y\n",
    "TRAIN_TARGET = f\"{PREFIX}.Sequences_activity_Train.txt\"\n",
    "VAL_TARGET = f\"{PREFIX}.Sequences_activity_Val.txt\"\n",
    "TEST_TARGET = f\"{PREFIX}.Sequences_activity_Test.txt\"\n",
    "\n",
    "writeFa(test[fa_cols], TEST_FA)\n",
    "writeFa(train[fa_cols], TRAIN_FA)\n",
    "writeFa(val[fa_cols], VAL_FA)\n",
    "\n",
    "cols = [\"coor.type\",\n",
    "        'ctrl.hob',\n",
    "        'US_DIF.hob',\n",
    "        'ctrl.hepg2',\n",
    "        'US_DIF.hepg2',\n",
    "        'ctrl.k562',\n",
    "        'US_DIF.k562',\n",
    "        'ctrl.bj',\n",
    "        'US_DIF.bj'\n",
    "        ]\n",
    "test[cols].to_csv(TEST_TARGET, sep='\\t', index=False)\n",
    "train[cols].to_csv(TRAIN_TARGET, sep='\\t', index=False)\n",
    "val[cols].to_csv(VAL_TARGET, sep='\\t', index=False)\n",
    "\n",
    "print(val.shape, train.shape, test.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7830e136",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (mamba)",
   "language": "python",
   "name": "mamba"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "302.4px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
