{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4143f6c1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-04-05T17:54:29.564992Z",
     "start_time": "2023-04-05T17:54:29.520300Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "63b9922c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-04-05T21:55:33.484418Z",
     "start_time": "2023-04-05T21:55:33.439295Z"
    },
    "code_folding": [
     163
    ]
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "# coding: utf-8\n",
    "\n",
    "# 1]:\n",
    "\n",
    "import argparse\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "from Bio.Seq import Seq\n",
    "from collections import Counter\n",
    "from functools import partial\n",
    "import glob\n",
    "import gzip\n",
    "from itertools import product\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "\n",
    "from timeit import default_timer as timer\n",
    "\n",
    "\"\"\"\n",
    "arg_parser = argparse.ArgumentParser()\n",
    "arg_parser.add_argument(\"job_num\", type=str, help='job number to look up in array.tsv')\n",
    "arg_parser.add_argument(\"array\", type=str, help='ARRAY file of job numbers, chromosome ids')\n",
    "\n",
    "args = arg_parser.parse_args()\n",
    "JOBNUM, ARRAY, =  args.job_num, args.array\n",
    "\n",
    "\"\"\"\n",
    "JOBNUM, ARRAY = 3, \"/wynton/home/ahituv/fongsl/nullomers/bin-generate/arrays/array-sacCer3.11mer.tsv\"\n",
    "# PATH, WINDOW_SIZE, KEYSIZE, WRITENULL = \"/wynton/home/ahituv/fongsl/dna/sacCer3\", 11, 4, True\n",
    "\n",
    "\n",
    "# FUNCTIONS\n",
    "\n",
    "def array_reader(array, job_number):\n",
    "    \"\"\"\n",
    "    read array, return values matching job number\n",
    "    \"\"\"\n",
    "\n",
    "    with open(array, \"r\") as reader:\n",
    "        for line in reader:\n",
    "            num, key_pool, path, build, windowsize, keysize, writenull = line.strip(\n",
    "                \"\\n\").split(\"\\t\")\n",
    "            if int(num) == int(job_number):\n",
    "\n",
    "                # change datatypes\n",
    "                windowsize, keysize, writenull = int(\n",
    "                    windowsize), int(keysize), bool(writenull)\n",
    "                key_pool = key_pool.split(\",\")  # make list\n",
    "\n",
    "                return key_pool, path, build, windowsize, keysize, writenull\n",
    "\n",
    "# make outdirectory for results\n",
    "\n",
    "\n",
    "def makeOutDir(path, window_size):\n",
    "    \"\"\"\n",
    "    make outdirectory\n",
    "    \"\"\"\n",
    "\n",
    "    OUTDIR = os.path.join(path, \"kmers\", f\"{window_size}mers\")\n",
    "\n",
    "    print(\"OUTDIR\", OUTDIR)\n",
    "    if os.path.exists(OUTDIR) is False:\n",
    "        if os.path.exists(os.path.dirname(OUTDIR)) is False:\n",
    "            os.mkdir(os.path.dirname(OUTDIR))\n",
    "        os.mkdir(OUTDIR)\n",
    "    return OUTDIR\n",
    "\n",
    "\n",
    "def makeKeys(keysize):\n",
    "    \"\"\"\n",
    "    input\n",
    "        keysize (int) - length of kmer-keys to create\n",
    "\n",
    "    require\n",
    "        itertools.product\n",
    "        python list comprehension\n",
    "\n",
    "    method\n",
    "        itertools.product kmers -> join kmers into str -> make list\n",
    "\n",
    "\n",
    "    return\n",
    "        key_set (set) - list of strings. \n",
    "    \"\"\"\n",
    "    key_set = set()\n",
    "    for item in product(\"ACGT\", repeat=keysize):\n",
    "        key_set.add(\"\".join(item))\n",
    "\n",
    "    return key_set\n",
    "\n",
    "\n",
    "# 4]:\n",
    "\n",
    "\n",
    "# ## extract fa sequence\n",
    "def extractFaSeq(path, key_pool, keysize, window_size, build):\n",
    "    \"\"\"\n",
    "    get chr-specific sequence, reverse complement\n",
    "\n",
    "    require\n",
    "        Bio.SeqIO.FastaIO - SimpleFastaParser\n",
    "        gzip\n",
    "\n",
    "    input\n",
    "        chr_num (str) - chromosome number e.g. 1,2,3,X,Y\n",
    "        path (str) - path to .fa files\n",
    "\n",
    "    method\n",
    "        1. get fa.gz name\n",
    "        2. open fa file\n",
    "        3. get sequence w/ simpleFastaParser\n",
    "        4. get reverse complement by turning seq into Seq object\n",
    "\n",
    "    return \n",
    "        seq (seq record) - str of sequence (should be one per chromosome)\n",
    "        rev (seq record) - str of reverse complement sequence (should be one per chromosome)\n",
    "    \"\"\"\n",
    "\n",
    "    print(\"reading fa\", build)\n",
    "\n",
    "    fa_dict = {\n",
    "        \"hs1\": \"/wynton/home/ahituv/fongsl/dna/hs1/hs1.fa.gz\",\n",
    "        \"hg38\": \"/wynton/group/databases/goldenPath/hg38/bigZips/hg38.fa.gz\",\n",
    "        \"saccer3\": \"/wynton/home/ahituv/fongsl/dna/sacCer3/sacCer3.fa.gz\"\n",
    "    }\n",
    "\n",
    "    FA = fa_dict[build.lower()]\n",
    "\n",
    "    forward, reverse = {}, {}\n",
    "\n",
    "    # 2 open\n",
    "    with gzip.open(FA, \"rt\") as handle:\n",
    "\n",
    "        ### COULD PARALLEL HERE###\n",
    "        # 3  read using simpleFastaParser\n",
    "        for val in SimpleFastaParser(handle):\n",
    "            start = timer()\n",
    "            seq_id, seq = val\n",
    "            # print(seq_id)\n",
    "\n",
    "            f_dict = getKmerCount(key_pool, keysize, window_size, seq)\n",
    "            forward[seq_id] = f_dict\n",
    "\n",
    "            # 4\n",
    "            seq_rev = str(Seq(seq).reverse_complement())\n",
    "            r_dict = getKmerCount(key_pool, keysize, window_size, seq_rev)\n",
    "            reverse[seq_id] = r_dict\n",
    "\n",
    "            print(\"done w seq\", seq_id, \"in\", timer()-start)\n",
    "\n",
    "    return forward, reverse\n",
    "\n",
    "\n",
    "def getKmerCount(key_pool, keysize, windowsize, sequence):\n",
    "    \"\"\"\n",
    "    get 1 kmer; break up sequence into equally sized kmers w sliding window of with windowsize, stepsize\n",
    "\n",
    "    required packages\n",
    "        - numpy\n",
    "\n",
    "    inputs\n",
    "        key_pool (list) - list of keys key to filter sequences by \n",
    "        keysize (int) - size of key\n",
    "        windowsize (int) - windowsize to make for sequence\n",
    "        sequence (str) - sequence to break into kmers\n",
    "\n",
    "\n",
    "    method\n",
    "        1. establish all the positions for the sequence\n",
    "        2. iterate through start positions, get kmer, kmer key\n",
    "        3. if kmer does not have N in sequence and key in keypool, add to dictionary of dictionaries\n",
    "\n",
    "    return \n",
    "        key_dict (dict) - dict of keys + dict of kmers and counts that match key_pool in input sequence\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # 1\n",
    "    n_kmers, start = (len(sequence) - windowsize + 1), 0\n",
    "\n",
    "    key_dict = {}  # make dictionary of dictionaries w/ counters\n",
    "    for k in key_pool:\n",
    "        key_dict[k] = Counter()\n",
    "\n",
    "    # 2\n",
    "    for start in np.arange(n_kmers):\n",
    "\n",
    "        # get kmer from sequence\n",
    "        kmer = sequence[start:start+windowsize].upper()\n",
    "        kmer_key = kmer[:keysize]  # get sequence key\n",
    "\n",
    "        # 3\n",
    "        if kmer_key in key_pool:  # append sequence windows within range of possible windows\n",
    "            if kmer_key not in key_dict:\n",
    "                key_dict[kmer_key] = Counter()\n",
    "                key_counter = key_dict[kmer_key]\n",
    "\n",
    "            else:\n",
    "                key_counter = key_dict[kmer_key]\n",
    "\n",
    "            if \"N\" not in kmer.upper():\n",
    "                key_counter[kmer] += 1\n",
    "\n",
    "    return key_dict\n",
    "\n",
    "\n",
    "def countKmersUniverse(forward_dict, reverse_dict, keysize):\n",
    "    \"\"\"\n",
    "    add kmer counts across chromosomes per key\n",
    "\n",
    "    input\n",
    "        forward_dict (dict) - dict of forward kmers + counts in fasta sequence\n",
    "        reverse_dict (dict) - dict of reverse kmers + counts in fasta sequence\n",
    "        keysize (int) - size of key identifier sequence\n",
    "\n",
    "    method\n",
    "        1. make key_dict to build kmer-universe dict while parsing kmers. Reduces memory burden. \n",
    "            1.1 run forward and reverse dictionaries sequentially\n",
    "        2. per forward/reverse dictionary (keys = chr numbers, values = dictionary of dictionaries). Need only values here. \n",
    "            2.1 unpack dictionary of key, count_dict for each chromosome\n",
    "            2.2 if key sequence is not in key dictionary, add as Counter\n",
    "            2.3 iterate though key-related kmers, keep only kmer-snippets linked to key, kmer, and their counts,\n",
    "                increasing the kmer count for a kmer while iterating across forward and reverse, across chromosomes.  \n",
    "\n",
    "\n",
    "\n",
    "    return\n",
    "        key_dict (dictionary) - genome universe w/ kmer counts.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    key_dict = {}  # make dictionary of dictionaries w/ counters\n",
    "\n",
    "    start = timer()\n",
    "\n",
    "    # 2 count forward/reverse items in input dicts, where keys are chromosomes, values are dictionaries of dictionaries of keys, full kmers, and kmer counts\n",
    "    seqs = [forward_dict, reverse_dict]\n",
    "\n",
    "    for seq_dict in seqs:\n",
    "        for chr_dict in seq_dict.values():\n",
    "\n",
    "            # 2.1\n",
    "            for key, count_dict in chr_dict.items():\n",
    "\n",
    "                # 2.2 check that key is in dictionary\n",
    "                if key not in key_dict:\n",
    "                    key_dict[key] = Counter()\n",
    "\n",
    "                # get key-specific kmer counter.\n",
    "                key_counter = key_dict[key]\n",
    "\n",
    "                # 2.3\n",
    "                for seq, count in count_dict.items():\n",
    "                    # cut sequence into kmer snippet\n",
    "                    kmer_snippet = seq[keysize:]\n",
    "                    key_counter[kmer_snippet] += count\n",
    "    end = timer()\n",
    "\n",
    "    print(\"Counting kmer instances:\", end-start)\n",
    "\n",
    "    return dict(key_dict)\n",
    "\n",
    "\n",
    "def writeDictionary(path, windowsize, result_dict, writenull):\n",
    "    \"\"\"\n",
    "    write kmer universe dictionary of dictionaries to outfiles, write nullomers\n",
    "\n",
    "    input\n",
    "\n",
    "        path (str) - path to write outfile\n",
    "        windowsize (int) - size of window used for kmers\n",
    "        results_dict (dictionary) - dictionary of dictionaries, kmer universe w/ keys, kmer sequences, and counts of kmer occurrences \n",
    "        writenull (bool) - write nullomer file or don't\n",
    "\n",
    "    method\n",
    "        1. create the outfile, sequence universe\n",
    "        2. write key, value as comma-separated str to kmer outfile\n",
    "        3. write left-over sequences to nullomer file\n",
    "\n",
    "    return\n",
    "        written_keys(list) - list of keys\n",
    "    \"\"\"\n",
    "    written_keys = []\n",
    "    for key, kmer_dict in result_dict.items():\n",
    "        print(\"\\n Writing dict\", key)\n",
    "        # 1\n",
    "        kmer_file = os.path.join(path, f\"{key}.csv\")\n",
    "        null_file = os.path.join(path, f\"{key}.nullomers.csv\")\n",
    "\n",
    "        ### FUTURE SARAH - YOU WILL HAVE TO FIND A BETTER WAY TO MAKE KEYS FOR LARGER SEQUENCES ###\n",
    "\n",
    "        if writenull is True:\n",
    "            # sometimes the universe is too big to make (e.g. 4**18 = 6.9e10 sequence)\n",
    "            # set writenull to False\n",
    "            universe = makeKeys(windowsize-len(key))\n",
    "\n",
    "        # 2\n",
    "        with open(kmer_file, \"w\") as kmer:\n",
    "\n",
    "            # write kmers first\n",
    "            for value_seq, count in kmer_dict.items():\n",
    "\n",
    "                # full sequence back together\n",
    "                seq = key+value_seq\n",
    "\n",
    "                kmer.write(f\"{seq},{count}\\n\")\n",
    "\n",
    "                if writenull is True:\n",
    "                    # remove key from kmer universe\n",
    "                    universe.remove(value_seq)\n",
    "\n",
    "            kmer.close()\n",
    "            written_keys.append(key)\n",
    "\n",
    "        # 3\n",
    "        if writenull is True:\n",
    "            with open(null_file, \"w\") as nullomer:\n",
    "                for seq in universe:\n",
    "                    nullomer.write(f\"{key+seq}\\n\")\n",
    "                nullomer.close()\n",
    "\n",
    "    return written_keys\n",
    "\n",
    "\n",
    "def readKeyLog(path):\n",
    "    \"\"\"\n",
    "    read chr.log and determine whether chromosome has already been summed into file dataset\n",
    "\n",
    "   input\n",
    "        chr_num (str) - chromosome number\n",
    "        path (str) - path to directory to write log\n",
    "\n",
    "    method\n",
    "        1. make the log file\n",
    "            1.1 if log file does not exist, return False (none of the chromosomes have been run) \n",
    "        2. open existing log file. \n",
    "        3. append chr_num to list\n",
    "        4. check whether input chr_num is in list:\n",
    "            4.1 if yes, return True (chromosome has been run)\n",
    "            4.2 if no, return False (chromos dome has not been run)\n",
    "\n",
    "    return\n",
    "        empty set\n",
    "        chr_runlist (set) - if key has been added to final nullomer count\n",
    "    \"\"\"\n",
    "\n",
    "    # 1\n",
    "    out = os.path.join(path, \"chr.log\")\n",
    "\n",
    "    # 2\n",
    "    runlist = set()\n",
    "    if os.path.exists(out) is True:\n",
    "\n",
    "        with open(out, \"r\") as chrlog:\n",
    "            # 3\n",
    "            for line in chrlog.readlines():\n",
    "                runlist.add(str(line.split(\"\\n\")[0]))\n",
    "\n",
    "    return runlist\n",
    "\n",
    "\n",
    "def writeKeyLog(key_pool, path):\n",
    "    \"\"\"\n",
    "    write log of chrs written to the all mer\n",
    "\n",
    "    input\n",
    "        chr_num (str) - chromosome number\n",
    "        path (str) - path to directory to write log\n",
    "\n",
    "    method\n",
    "        1. make the log file\n",
    "        2. open the log file\n",
    "        3. append the chr_num to the log file\n",
    "        4. close the file\n",
    "    \"\"\"\n",
    "    # 1\n",
    "    out = os.path.join(path, \"chr.log\")\n",
    "\n",
    "    # 2\n",
    "    with open(out, \"a\") as chrlog:\n",
    "        for key in key_pool:\n",
    "\n",
    "            # 3\n",
    "            chrlog.write(f\"{key}\\n\")\n",
    "\n",
    "    # 4\n",
    "    chrlog.close()\n",
    "\n",
    "# MAIN\n",
    "\n",
    "def main(argv):\n",
    "    \"\"\"\n",
    "    count number of kmer occurrences per chromosome, genome, write output to file\n",
    "\n",
    "    method\n",
    "        0. read array, get key_pool from array, make outdirectory\n",
    "\n",
    "        1. read log, check if you've run any keys already\n",
    "\n",
    "        2. if key has not been run (set), run it\n",
    "\n",
    "        3. get genome sequence, forward and reverse as dictionaries \n",
    "            for each chromosome, each key, kmers, and counts. \n",
    "            (Should be 25 items corresponding to 25 chromosomes)\n",
    "            3.1 scan genome for kmer windows, keep only windows with keys in the 5'\n",
    "            \n",
    "        4. combine key, kmer counts across forward and reverse and across chromosomes. \n",
    "\n",
    "        5. write kmer keys\n",
    "    \"\"\"\n",
    "\n",
    "    # 0\n",
    "    KEY_POOL, PATH, BUILD, WINDOW_SIZE, KEYSIZE, WRITENULL = array_reader(ARRAY, JOBNUM)  #KEY_POOL = [\"AAAA\", \"TTTT\"]\n",
    "\n",
    "\n",
    "    OUTDIR = makeOutDir(PATH, WINDOW_SIZE)\n",
    "    \n",
    "    # 1\n",
    "    run_already = readKeyLog(OUTDIR)  # check whether key has been run already\n",
    "    \n",
    "    KEY_POOL_ = set(KEY_POOL).difference(run_already)\n",
    "    KEY_POOL=list(KEY_POOL_)\n",
    "    # 2\n",
    "    if len(KEY_POOL)>0:  # remove run already set.\n",
    "        print(\"get nullomers space from kmers\", KEY_POOL)\n",
    "\n",
    "        # 3 get forward, reverse complement genome sequences, filtering on key\n",
    "        FWD, REV = extractFaSeq(PATH, KEY_POOL, KEYSIZE, WINDOW_SIZE, BUILD)\n",
    "\n",
    "    # 4 count knownmers in kmer universe\n",
    "    key_universe = countKmersUniverse(FWD, REV, KEYSIZE)\n",
    "\n",
    "    #5 write kmer dictionaries summed across sequences per key identity. \n",
    "    written_keys = writeDictionary(OUTDIR, WINDOW_SIZE, key_universe, WRITENULL)\n",
    "    \n",
    "    # write the key to the out dir\n",
    "    writeKeyLog(written_keys, OUTDIR)\n",
    "\n",
    "    \n",
    "if __name__ == '__main__':\n",
    "    main(sys.argv[1:])    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30e1a84b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-04-05T16:24:00.515073Z",
     "start_time": "2023-04-05T16:24:00.494935Z"
    }
   },
   "outputs": [],
   "source": [
    "BUILD = \"saccer3\"\n",
    "fa_dict={\n",
    "\"hs1\":\"/wynton/home/ahituv/fongsl/dna/hs1/hs1.fa.gz\",\n",
    "\"hg38\":\"/wynton/group/databases/goldenPath/hg38/bigZips/hg38.fa.gz\",\n",
    "\"saccer3\":\"/wynton/home/ahituv/fongsl/dna/sacCer3/sacCer3.fa.gz\"\n",
    "}\n",
    "\n",
    "\n",
    "FA = fa_dict[BUILD]\n",
    "\n",
    "forward, reverse = {}, {}\n",
    "#2 open\n",
    "with gzip.open(FA, \"rt\") as handle:\n",
    "\n",
    "    #3  read using simpleFastaParser\n",
    "    for val in SimpleFastaParser(handle):\n",
    "        start = timer()\n",
    "        seq_id, seq = val\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41cda44d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-04-05T16:33:27.047811Z",
     "start_time": "2023-04-05T16:33:27.040807Z"
    }
   },
   "outputs": [],
   "source": [
    "# set up\n",
    "sequence = seq\n",
    "windowsize = WINDOW_SIZE\n",
    "keysize = KEYSIZE\n",
    "key_pool = [\"CCAC\", \"CACA\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a594e4c1",
   "metadata": {},
   "source": [
    "# kmerlist count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e8a8c13",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-04-05T16:33:27.047811Z",
     "start_time": "2023-04-05T16:33:27.040807Z"
    }
   },
   "outputs": [],
   "source": [
    "#1\n",
    "n_kmers, start = (len(sequence) - windowsize + 1),0\n",
    "\n",
    "key_dict={}  # make dictionary of dictionaries w/ counters\n",
    "for k in key_pool:\n",
    "    key_dict[k]= Counter()\n",
    "\n",
    "#2\n",
    "for start in np.arange(n_kmers):\n",
    "\n",
    "    kmer = sequence[start:start+windowsize].upper()\n",
    "    kmer_key = kmer[:keysize]\n",
    "\n",
    "    #3\n",
    "    if kmer_key in key_pool: # append sequence windows within range of possible windows\n",
    "        key_counter = key_dict[kmer_key]\n",
    "        if \"N\" not in kmer.upper():\n",
    "            key_counter[kmer] +=1\n",
    "    del kmer, kmer_key"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4d4e6f1",
   "metadata": {},
   "source": [
    "## count universe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e221759",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-04-05T17:39:13.759715Z",
     "start_time": "2023-04-05T17:39:13.754482Z"
    }
   },
   "outputs": [],
   "source": [
    "forward_dict = FWD \n",
    "reverse_dict = REV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90b51f03",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-04-05T17:52:47.393630Z",
     "start_time": "2023-04-05T17:52:47.022715Z"
    }
   },
   "outputs": [],
   "source": [
    "key_dict = {}  # make dictionary of dictionaries w/ counters\n",
    "\n",
    "start = timer()\n",
    "\n",
    "# 2 count forward items in dictionary, keys are chromosomes, values are dictionaries of dictionaries\n",
    "seqs=[forward_dict, reverse_dict]\n",
    "for seq_dict in seqs:\n",
    "    for chr_dict in seq_dict.values(): \n",
    "        for key, count_dict in chr_dict.items():\n",
    "\n",
    "            if key not in key_dict:\n",
    "                key_dict[key] = Counter()\n",
    "            key_counter = key_dict[key]\n",
    "\n",
    "            for seq, count in count_dict.items():\n",
    "                # split string into key, value\n",
    "                key, value = seq[:keysize], seq[keysize:]\n",
    "                key_counter[value] += count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5282270c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-04-05T17:52:00.853296Z",
     "start_time": "2023-04-05T17:52:00.507586Z"
    }
   },
   "outputs": [],
   "source": [
    "for i in [forward_dict.items(), reverse_dict.items()]:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cacf187",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-04-05T17:47:28.242123Z",
     "start_time": "2023-04-05T17:47:27.883515Z"
    }
   },
   "outputs": [],
   "source": [
    "# get matching key data from  reverse dictionary            \n",
    "for chr_dict in reverse_dict.values():\n",
    "    for key, count_dict in chr_dict.items():\n",
    "        \n",
    "        if key not in key_dict:\n",
    "            key_dict[key] = Counter()\n",
    "        key_counter = key_dict[key]\n",
    "        \n",
    "        for seq, count in count_dict.items():\n",
    "            \n",
    "            # split string into key, value\n",
    "            key, value = seq[:keysize], seq[keysize:]\n",
    "            key_counter[value] += count\n",
    "\n",
    "end = timer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89fdc151",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-04-05T17:52:56.163772Z",
     "start_time": "2023-04-05T17:52:56.117694Z"
    }
   },
   "outputs": [],
   "source": [
    "key_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "804b7bdf",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-04-05T17:50:22.605475Z",
     "start_time": "2023-04-05T17:50:22.530926Z"
    }
   },
   "outputs": [],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6796d4c3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (wynenv)",
   "language": "python",
   "name": "wynenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
