{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-18T18:29:56.391609Z",
     "start_time": "2023-12-18T18:29:51.492382Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('sans-serif', 'Arial', 18)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "import numpy as np\n",
    "import os, sys\n",
    "import pandas as pd\n",
    "import pybedtools as pb\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "LOCAL = True\n",
    "if LOCAL is True:\n",
    "    sys.path.append(\"/Users/sarahfong/tools/py_\")\n",
    "    PATH = \"/Users/sarahfong/Desktop/local_data/Biomarin_Jun_2023/\"\n",
    "    RE = os.path.join(PATH, \"results\")\n",
    "    CONFIG_NAME=\"config.local.neuron.mpra.ini\"\n",
    "    RE_MAPS = os.path.join(RE, \"enh_maps\")\n",
    "else:\n",
    "    PATH=\"/wynton/group/ahituv/biomarin/data\"\n",
    "    RE = \"/wynton/group/ahituv/biomarin/results\"\n",
    "    RE_MAPS = os.path.join(RE, \"enh_maps\")\n",
    "    CONFIG_NAME=\"config.neuron.mpra.orig.ini\"\n",
    "\n",
    "# nomination dataframe\n",
    "NOMS = os.path.join(PATH, \"GABA_GLUT_DF.for.noms.tsv\")\n",
    "\n",
    "# summarized across significant loci with support neighbor significance in 3/5 neighbors\n",
    "CANDIDATE_REGIONS = os.path.join(PATH, \"CANDIDATE.SIG.REGIONS.GABA.GLUT.tsv\")\n",
    "SUMMED_CANDIDATE_REGIONS = os.path.join(PATH, \"SUMMARIZED.CANDIDATE.SIG.REGIONS.GABA.GLUT.tsv\")\n",
    "\n",
    "import config_readwrite as crw\n",
    "import plot_params as pp\n",
    "pp.fonts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-18T18:29:56.403055Z",
     "start_time": "2023-12-18T18:29:56.396095Z"
    }
   },
   "outputs": [],
   "source": [
    "config, cfn = crw.read(CONFIG_NAME)\n",
    "\n",
    "section = 'lib2'\n",
    "crw.check(config, section)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-18T18:29:56.515835Z",
     "start_time": "2023-12-18T18:29:56.408023Z"
    }
   },
   "outputs": [],
   "source": [
    "if \"fasta\" not in list(config[section]):  # write files to config\n",
    "\n",
    "    MPRA = os.path.join(PATH, \"delta_rank.csv\")\n",
    "\n",
    "    FASTA = os.path.join(PATH, \"library_2\", \"Design\",\n",
    "                         \"biomarin-lib2-hg38-final.fasta\")\n",
    "    META_DATA = os.path.join(PATH, \"lib2.meta_data.tsv\")\n",
    "    TILE_BED = os.path.join(PATH, \"lib2.processed.bed\")\n",
    "    ENH_BED = os.path.join(PATH, 'lib2.processed.merged.bed')\n",
    "    GREAT_GENE_TSV = os.path.join(PATH, \"GREAT\", f\"region2gene_lib2hg38.tsv\")\n",
    "    DELTA_ACTIVITY = os.path.join(PATH, \"deltaMPRA.gaba.minus.glut.tsv\")\n",
    "    DELTA_ACTIVITY_Z = os.path.join(\n",
    "        PATH, \"deltaMPRA.gaba.minus.glut.zscore.tsv\")\n",
    "    SIG_ACTIVITY_Z = os.path.join(PATH, \"WilcoxonSig.gaba.glut.zscore.tsv\")\n",
    "    VISTA = os.path.join(PATH, \"vista.lib2.tiles.bed\")\n",
    "\n",
    "    # delta MPRA activity\n",
    "    config[section][\"delta.mpra_richa\"] = MPRA\n",
    "    config[section][\"delta.mpra_centered\"] = DELTA_ACTIVITY\n",
    "    config[section][\"delta.mpra_centeredz\"] = DELTA_ACTIVITY_Z\n",
    "\n",
    "    config[section][\"fasta\"] = FASTA\n",
    "\n",
    "    # genome coordinates\n",
    "    config[section][\"tile_bed\"] = TILE_BED\n",
    "    config[section][\"enh_bed\"] = ENH_BED\n",
    "\n",
    "    # GREAT\n",
    "    config[section][\"great_nearest\"] = GREAT_GENE_TSV\n",
    "\n",
    "    # meta_data\n",
    "    config[section][\"metadata\"] = META_DATA\n",
    "\n",
    "    # significance GABA v. GLUT\n",
    "    config[section][\"sig-wilcoxon\"] = SIG_ACTIVITY_Z\n",
    "\n",
    "    # vista\n",
    "    config[section][\"vista\"] = VISTA\n",
    "    crw.write(config, cfn)\n",
    "\n",
    "else:\n",
    "    MPRA = config[section][\"delta.mpra_richa\"]\n",
    "    DELTA_ACTIVITY = config[section][\"delta.mpra_centered\"]\n",
    "    DELTA_ACTIVITY_Z = config[section][\"delta.mpra_centeredz\"]\n",
    "\n",
    "    FASTA = config[section][\"fasta\"]\n",
    "\n",
    "    TILE_BED = config[section][\"tile_bed\"]\n",
    "    ENH_BED = config[section][\"enh_bed\"]\n",
    "\n",
    "    GREAT_GENE_TSV = config[section][\"great_nearest\"]\n",
    "    META_DATA = config[section][\"metadata\"]\n",
    "\n",
    "    SIG_ACTIVITY_Z = config[section][\"sig-wilcoxon\"]\n",
    "    VISTA = config[section][\"vista\"]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "enh = pd.read_csv(NOMS, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "# nominate candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## functions!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get min dist to genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-18T18:30:24.787961Z",
     "start_time": "2023-12-18T18:30:24.781374Z"
    }
   },
   "outputs": [],
   "source": [
    "def getMinDist(df):\n",
    "    \"\"\"return str of gene, min distance tuples to gene TSSs per enhancer. \n",
    "        Gene TSS annotations from GREAT. \n",
    "        note - most enhancers map to two genes, which means two TSSs\n",
    "    \"\"\"\n",
    "\n",
    "    genes = []\n",
    "\n",
    "    # per gene\n",
    "    for gene in df[\"gene\"].unique():\n",
    "        genes.append(gene)\n",
    "        print(gene)\n",
    "\n",
    "        # get the gene dataframe\n",
    "        g = df.loc[(df[\"gene\"] == gene)].copy()\n",
    "\n",
    "        # if upstream TSS, return min value\n",
    "        g = g.loc[g[\"dist\"] != \"NONE\"]\n",
    "        if len(g) > 0:\n",
    "            if int(g.dist.min()) > 1:\n",
    "                genes.append(g.dist.min())\n",
    "\n",
    "            # elif downstream TSS, return max negative value (min)\n",
    "            elif int(g.dist.min()) < 1:\n",
    "                genes.append(g.dist.max())\n",
    "                \n",
    "            elif list(set(g.dist)) == \"na\":\n",
    "                print(gene,  \"TO REMOVE\")\n",
    "                genes = genes[:-1]  # remove gene from list, no distance\n",
    "\n",
    "    if len(genes) == 0:  # handle cases where no gene is mapped to element.\n",
    "        genes = [\"None\"]\n",
    "\n",
    "    new_genes = []\n",
    "\n",
    "    for gene in genes:\n",
    "        new_genes.append(str(gene))\n",
    "\n",
    "    return \"_\".join(new_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### N significant labels per tile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-18T18:30:25.070812Z",
     "start_time": "2023-12-18T18:30:24.790295Z"
    }
   },
   "outputs": [],
   "source": [
    "def getSupport(enh_id, df):\n",
    "    \"\"\"count how many supports there is for differential activity\"\"\"\n",
    "    \n",
    "    # columns to keep\n",
    "    t_cols = [\"name\", 'tile.order', \"enh.id\", \"sig\", \"celltype_dif\", \"bs\"]\n",
    "    \n",
    "    # subset df to enh_id w cols, drop duplicates, copy, and fill na w False\n",
    "    t = df.loc[df[\"enh.id\"] == enh_id, t_cols].drop_duplicates().copy().fillna(False)\n",
    "    \n",
    "    # turn bool into int, into str\n",
    "    t[t_cols[-3:]]=t[t_cols[-3:]].astype(int)\n",
    "    \n",
    "    # str code the support col\n",
    "    t[\"support\"] =  t[\"sig\"] + t[\"celltype_dif\"] + t[\"bs\"]\n",
    "    \n",
    "    # turn int into str\n",
    "    t[t_cols[-3:]]=t[t_cols[-3:]].astype(str)\n",
    "\n",
    "    t[\"support_code\"] =  t[\"sig\"] + t[\"celltype_dif\"] + t[\"bs\"]    \n",
    "    \n",
    "    t[\"support_name\"] = None\n",
    "    \n",
    "    # name codes, so that they are legible\n",
    "    name_codes = {None:\"000\", \n",
    "                  \"cat-only\":\"010\", \n",
    "                  \"sig-only\":\"100\", \n",
    "                  \"bs-only\":'001',\n",
    "                  \"sig-n-cat\":\"110\", \n",
    "                  \"sig-n-bs\":\"101\", \n",
    "                  \"cat-n-bs\":\"011\", \n",
    "                  \"all\": \"111\"\n",
    "                 }\n",
    "    \n",
    "    for name, code in name_codes.items():\n",
    "        t.loc[t[\"support_code\"]==code, \"support_name\"] = name\n",
    "\n",
    "\n",
    "    return t[[\"name\", \"tile.order\", \"support\", \"support_code\", \"support_name\"\n",
    "             ]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Neighbor analysis - N significant labels from 5-tile window"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def supportingNeighbors(support):\n",
    "\n",
    "    \"\"\"Scan the nearest 5 windows and count support (number of significance tests the data passes)).\n",
    "    return the tiles with the most significance from supported neighborhoods, return set of tiles that support each other\n",
    "    \"\"\"\n",
    "    windowsize = 5  # define windowsize\n",
    "\n",
    "    support_list = list(support[\"support\"]) # list of the support for each tile (ordered) across enhancer \n",
    "\n",
    "    support_neighborhoods = []  # collect the tiles that have supporting neighbors\n",
    "    all_neighbors = []\n",
    "   \n",
    "    #  slide window across enhancer, finding supporting neighborhoods of activity\n",
    "    for i in np.arange((support.shape[0] - windowsize+1)):\n",
    "        \n",
    "        # get neighborhood slice\n",
    "        window_slice = support_list[i: i+windowsize]\n",
    "\n",
    "        # count the number of significant measurements for each tile in windown\n",
    "        sig_array, sig_array_count = np.unique(window_slice, return_counts=True)\n",
    "        \n",
    "        # test whether window has 3+ tiles with *ANY* level of significance. \n",
    "        if 0 not in sig_array or sig_array_count[0]<3:\n",
    "            \n",
    "            # find the tile with the most evidence of significance\n",
    "            m = window_slice.index(max(window_slice))\n",
    "            support_neighborhoods.append(i+m)  # add that tile to the list\n",
    "            #print(\"add\", i, m,  \"window\", window_slice,  \"tile\", (i+m))\n",
    "        \n",
    "            # get all significant neighbors when sig neighborhood\n",
    "            for n, slice in enumerate(window_slice):\n",
    "                if slice> 0:\n",
    "                    #print(\"add all neighbors\", n +i)\n",
    "                    all_neighbors.append((n+i))\n",
    "                    \n",
    "    \n",
    "            \n",
    "    #print(set(all_neighbors))\n",
    "\n",
    "    return set(support_neighborhoods), set(all_neighbors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def neighborActivity(longform_glut, longform_gaba, neighbor_support):\n",
    "    glut_s = longform_glut.loc[longform_glut[\"tile.order\"].isin(neighbor_support)].groupby('tile.order')['Mean z-score ratio'].mean().reset_index()\n",
    "    glut_s.rename(columns={\"Mean z-score ratio\":\"glut.mean\"}, inplace=True)\n",
    "    \n",
    "    gaba_s = longform_gaba.loc[longform_gaba[\"tile.order\"].isin(neighbor_support)].groupby('tile.order')['Mean z-score ratio'].mean().reset_index()\n",
    "    gaba_s.rename(columns={\"Mean z-score ratio\":\"gaba.mean\"}, inplace=True)\n",
    "\n",
    "    #print(pd.merge(glut_s, gaba_s ))\n",
    "    # merge the mean activity of the gaba and glut tiles\n",
    "    return pd.merge(glut_s, gaba_s )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fillInSigGaps(sig_neighbor_tiles):\n",
    "    \"\"\"create blocks of tiles that are contiguously significant. Accept 1 gap\"\"\"\n",
    "    \n",
    "    \n",
    "    gap_limit = 3 # anything this size or larger is excluded\n",
    "    counter, start = 0, 0\n",
    "    contig = []\n",
    "    for i in sig_neighbor_tiles:\n",
    "        \n",
    "        if i == 0:  # if the list starts with zero\n",
    "            continue\n",
    "        elif i == sig_neighbor_tiles[0]:\n",
    "            start, counter = i, i\n",
    "            \n",
    "        elif i == sig_neighbor_tiles[-1]:  # when we're at the end of the sig_tiles\n",
    "            print(\"last one\")\n",
    "            if i > (counter + gap_limit):\n",
    "             \n",
    "                run = [i for i in np.arange(start, counter+1)]\n",
    "            else:\n",
    "                run = [i for i in np.arange(start, i+1)]\n",
    "            contig.append(run)\n",
    "            \n",
    "        elif i == (counter + 1):  # continuous\n",
    "            counter +=1\n",
    "            run = [i for i in np.arange(start, counter+1)]\n",
    "            \n",
    "        elif i == (counter +gap_limit):  # gap of 1\n",
    "            counter +=2\n",
    "            run = [i for i in np.arange(start, counter+1)]\n",
    "            \n",
    "        elif i > (start+counter +2) & start !=0 : # gap > 1\n",
    "            run = [i for i in np.arange(start, counter+1)]\n",
    "            print(\"broken at\", i, \"range\", start, 'to', counter, run)\n",
    "            contig.append(run)\n",
    "            start, counter = i, i\n",
    "            \n",
    "    print(contig)\n",
    "        \n",
    "    return contig"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### filter for tiles with 3 significant labels and direction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-18T18:30:25.262441Z",
     "start_time": "2023-12-18T18:30:25.072983Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def getSigTileIdDirection(direction, df, n_criteria=3):\n",
    "    \"\"\"return tile ids in direction which \n",
    "    (1) have significant activity differences\n",
    "    (2) have categorical activity differences\n",
    "    (3) exceed the bootstrapped 95 CIs from delta distribution\n",
    "    \"\"\"\n",
    "    cols = [\"bs\",\n",
    "            'direction',\n",
    "            \"name\", \"sig\"]\n",
    "    \n",
    "    if n_criteria ==1:\n",
    "        sig_col = \"sig\"\n",
    "    elif n_criteria ==2:\n",
    "        sig_col1 = \"sig_n_dif\"\n",
    "        sig_col2 = \"celltype_dif\"\n",
    "        sig_col3 = \"bs\"\n",
    "        test = df.loc[\n",
    "                  (df[\"sig\"] == True)&\n",
    "                    ((df[sig_col2] == True)|\n",
    "                    (df[sig_col3] == True)),\n",
    "                  cols].drop_duplicates().copy()\n",
    "        sig_col = sig_col1 + sig_col2 + sig_col3\n",
    "    else:\n",
    "        sig_col = \"sig_n_dif_n_bs\"\n",
    "        test = df.loc[\n",
    "                  (df[sig_col] == True),\n",
    "                  cols].drop_duplicates().copy()\n",
    "   \n",
    "\n",
    "    print(sig_col) \n",
    "    \n",
    "    candidate_set = set(test.loc[test[\"direction\"] == direction, \"name\"])\n",
    "\n",
    "    return candidate_set"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### make longform enhancer dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-18T18:30:25.344215Z",
     "start_time": "2023-12-18T18:30:25.264973Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def getLongFormEnhInfo(df, mpra_celltype):\n",
    "    \"\"\" return dataframes for plotting gaba/glut longform tile activity and significance\n",
    "    \"\"\"\n",
    "    # do some data formatting, sorting`\n",
    "    df[\"tile.order\"] = df[\"tile.order\"].astype(int)\n",
    "    df = df.sort_values(by=\"tile.order\")\n",
    "\n",
    "    # get dataframe of tile.order x significant differences\n",
    "    sd = pd.melt(df, id_vars=\"tile.order\",\n",
    "                 value_vars=\"sig_n_dif_n_bs\",\n",
    "                 value_name=\"sig\"\n",
    "                 ).drop_duplicates().reset_index()\n",
    "    sd[\"y\"] = 3\n",
    "    sd[\"sig\"] = sd[\"sig\"].replace(False, None)\n",
    "    \n",
    "    # new variable\n",
    "    VAR_NAME = \"Mean z-score ratio\"\n",
    "\n",
    "    # get replicates of gaba, glut and melt into longform\n",
    "    if mpra_celltype == \"gaba\": \n",
    "        cols = [\"gaba_1\", \"gaba_2\", \"gaba_3\"]\n",
    "        \n",
    "    elif mpra_celltype ==\"glut\":\n",
    "        cols = [\"glut_1\", \"glut_2\", \"glut_3\"]\n",
    "\n",
    "    # longform GABA\n",
    "    longform = pd.melt(df, id_vars=\"tile.order\",\n",
    "                            value_vars=cols,\n",
    "                            value_name=VAR_NAME).drop_duplicates().reset_index()\n",
    "\n",
    "   \n",
    "    # add back significance column\n",
    "    longform = pd.merge(longform, sd, how=\"left\", left_on=\"tile.order\",\n",
    "                             right_on='tile.order').drop(columns=[\"index_x\", \"index_y\"])\n",
    "    \n",
    "\n",
    "    # add back missing tiles\n",
    "    longform = addMissingTiles(longform, VAR_NAME)\n",
    "    \n",
    "    # change tile.order datatype\n",
    "    longform[\"tile.order\"]= longform[\"tile.order\"].astype(int)\n",
    "    \n",
    "    # fill any missing data w zeros\n",
    "    longform[VAR_NAME]=longform[VAR_NAME].fillna(0)\n",
    "    \n",
    "    return longform, VAR_NAME"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### enh_id given tile_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-18T18:30:25.416496Z",
     "start_time": "2023-12-18T18:30:25.346448Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def getEnhid(tile_coor, df):\n",
    "    \"\"\"return enhancer coordinate for tile_id\"\"\"\n",
    "    enh = set(df.loc[df[\"coor\"] == tile_coor, 'enh.id'])\n",
    "\n",
    "    if len(enh) > 1:\n",
    "        print(\"more than one enh_id\", enh)\n",
    "\n",
    "    # return the first (and hopefully only) enh\n",
    "    if len(enh)>0:\n",
    "        return list(enh)[0]\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### add back missing tiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-18T18:30:25.473582Z",
     "start_time": "2023-12-18T18:30:25.418917Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def addMissingTiles(longform, var_name):\n",
    "    \"\"\"add back missing tile values for a dataframe\"\"\"\n",
    "\n",
    "    # make an empty dataframe\n",
    "    empty = pd.DataFrame(\n",
    "        np.arange(longform[\"tile.order\"].min(),\n",
    "                  longform[\"tile.order\"].max()+1),\n",
    "        columns=[\"tile.order\"]\n",
    "    )\n",
    "    # add back empty, drop old index\n",
    "    longform = pd.merge(empty, longform,\n",
    "                             how=\"left\")\n",
    "    # fill na\n",
    "    longform[var_name] = longform[var_name].fillna(0)\n",
    "\n",
    "    return longform"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### line plot for enhancer landscape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-18T18:30:25.682631Z",
     "start_time": "2023-12-18T18:30:25.623285Z"
    },
    "code_folding": [
     35,
     43
    ],
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def plot_lineplot(longform_gaba, longform_glut, var_name, out, enh_id,  enh_name, gene_name, support, plot_support_annot):\n",
    "\n",
    "    if len(set(longform_gaba['tile.order'])) > 20:\n",
    "        fig, ax = plt.subplots(figsize=(12, 4))\n",
    "    else:\n",
    "        fig, ax = plt.subplots(figsize=(8, 4))\n",
    "\n",
    "    # print max scores\n",
    "    max_gaba, max_glut = longform_gaba[var_name].max(),\\\n",
    "    longform_glut[var_name].max()\n",
    "\n",
    "    # min scores\n",
    "    min_gaba, min_glut = longform_gaba[var_name].min(),\\\n",
    "    longform_glut[var_name].min()\n",
    "\n",
    "    # get max of two for plotting purposes\n",
    "    max_of_max = max(max_gaba, max_glut)\n",
    "    min_of_min = min(min_gaba, min_glut)\n",
    "\n",
    "    if max_of_max < 3.2:\n",
    "        ymax = 3.2\n",
    "        ymin = (ymax*-1) - 1\n",
    "    else:\n",
    "        ymax = max_of_max + 1\n",
    "        ymin = (ymax*-1) - 1\n",
    "\n",
    "    print('max gaba', max_gaba, \"max_glut\", max_glut)\n",
    "\n",
    "    # plot longform gaba across enhancer\n",
    "    sns.pointplot(x=\"tile.order\", y=var_name, data=longform_gaba.fillna(0),\n",
    "                  label=\"gaba\",\n",
    "                  join=False,\n",
    "                  errorbar=(\"sd\"),\n",
    "                  color=\"orange\", ax=ax)\n",
    "    plt.setp(ax.collections, alpha=.3)  # alpha\n",
    "    x=\"tile.order\"\n",
    "    data=longform_glut.fillna(0)\n",
    "\n",
    "    # plot longform glut activity across enhancer\n",
    "    sns.pointplot(x=\"tile.order\", y=var_name, data=data,\n",
    "                  label=\"glut\",\n",
    "                  join=False,\n",
    "                  errorbar=(\"sd\"),\n",
    "                  color=\"blue\", \n",
    "                 # ax=ax\n",
    "                 )\n",
    "    plt.setp(ax.collections, alpha=.3)  # alpha\n",
    "\n",
    "    # where to place support for significance\n",
    "    support[\"y\"] = ymax + 0.5\n",
    "\n",
    "    x = \"tile.order\"\n",
    "    y = \"y\"\n",
    "    data = support.fillna(0).reset_index(drop=True)\n",
    "    \n",
    "\n",
    "    if plot_support_annot == \"support_name\":\n",
    "        hue = \"support_name\"  # \"support\"\n",
    "        hue_order = [\"all\", \"sig-n-bs\", \"sig-n-cat\",\n",
    "                     'cat-n-bs', \"cat-only\", \"sig-only\", \"bs-only\"]\n",
    "        markers = [\"*\", 'o', \"s\", \"p\", \".\", \".\", \".\"]\n",
    "        marker_title = \"sig criteria\"\n",
    "        palette=\"tab20b\"\n",
    "\n",
    "    else:\n",
    "        hue='support'\n",
    "        hue_order = [3, 2, 1]\n",
    "        markers = ['*', \"o\", \".\"]\n",
    "        marker_title = \"sig criteria N\"\n",
    "        palette=\"tab10\"\n",
    "    \n",
    "\n",
    "    ax2 = ax.twiny()\n",
    "    sns.pointplot(x=x, y=y, data=data, hue=hue,\n",
    "                    hue_order=hue_order,\n",
    "                    markers=markers,\n",
    "                    palette=palette,\n",
    "                    errorbar=None,\n",
    "                    join=False, ax=ax2,\n",
    "                    scale=1.5\n",
    "                 )\n",
    "    ax2.set_xticklabels(\"\")\n",
    "\n",
    "    # mark tiles w/ all three sig. \n",
    "    if True in set(longform_glut['sig']) or True in set(longform_gaba[\"sig\"]):\n",
    "        longform_glut[\"y\"] = ymax \n",
    "\n",
    "        x=\"tile.order\"\n",
    "        y=\"y\"\n",
    "        hue=\"sig\"\n",
    "        data=longform_glut[[x,y,hue]].drop_duplicates().fillna(False)\n",
    "    \n",
    "        sns.pointplot(x=x, y=y, hue=hue, data=data, \n",
    "                      markers=[' ', \"X\"],\n",
    "                      color=\"r\",\n",
    "                      errorbar=None,\n",
    "                      join=False, ax=ax\n",
    "                      )\n",
    "\n",
    "    # zero line\n",
    "    ax.axhline(0, c=\"grey\", ls=\"--\")\n",
    "\n",
    "    # ax title, labels\n",
    "    ax.set(title=f\"{gene_name}bp to TSS\\n{enh_id}\",\n",
    "           ylabel=\"mean z-score\",\n",
    "           ylim=(ymin, ymax + 1)\n",
    "           )\n",
    "\n",
    "    \n",
    "    # Set major ticks for x axis\n",
    "    ntiles = longform_gaba['tile.order'].max()\n",
    "    if ntiles >= 100:\n",
    "        segment = 10\n",
    "    else:\n",
    "        segment = 5\n",
    "\n",
    "    major_xticks = np.arange(0, ntiles, segment)\n",
    "    ax.set_xticks(major_xticks)\n",
    "\n",
    "    #ax2.set_yticklabels(\"\")\n",
    "    \n",
    "    # legend\n",
    "    ax.legend(title='key', frameon=False,\n",
    "              bbox_to_anchor=(0.75, -0.25))\n",
    "    ax2.legend(title=marker_title, bbox_to_anchor=(1, -0.25), frameon=False)\n",
    "\n",
    "    # save figure\n",
    "    plt.savefig(out, bbox_inches=\"tight\")\n",
    "\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### tile landscape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-18T18:30:56.364674Z",
     "start_time": "2023-12-18T18:30:25.947851Z"
    },
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def tileLandscape(candidate_list, df, top_n, re_maps, direction):\n",
    "\n",
    "    \"\"\"evaluate if there is support from neighboring tiles, plot landscape of the top_n, return candidate list\"\"\"\n",
    "    val = 0\n",
    "    already_run = []\n",
    "\n",
    "    remove_tiles = []\n",
    "    # get corresponding enhancer id\n",
    "    for tile_id in candidate_list:\n",
    "        enh_id = getEnhid(tile_id, df)\n",
    "        \n",
    "        #multiple tiles per enhancer\n",
    "        if enh_id not in already_run and enh_id is not None:\n",
    "            \n",
    "            # get enh_id for tile coordinate\n",
    "            enh_id = getEnhid(tile_id, enh)\n",
    "        \n",
    "            cols = [\"sig_n_dif_n_bs\", \"glut_1\", \"glut_2\", \"glut_3\", \"gaba_1\",\n",
    "                    \"gaba_2\", \"gaba_3\", \"coor\", \"enh.id\", \"enh.name\", \"tile.order\", \"gene\"]\n",
    "        \n",
    "            t = df.loc[df[\"enh.id\"] == enh_id].drop_duplicates().copy()\n",
    "        \n",
    "            # stratify by gene name, as one enhancer may be near two different genes\n",
    "            enh_name = list(set(t[\"enh.name\"]))[0]\n",
    "            gene = getMinDist(t)\n",
    "        \n",
    "            # get support (number of significant calls) per tile in enhancer\n",
    "            support = getSupport(enh_id, t)\n",
    "    \n",
    "            # scan neighbors for supporting significant differences in activity (at least 3/5 neighbor tiles must have support). \n",
    "            neighbor_support = supportingNeighbors(support)\n",
    "    \n",
    "            # if there are significant neighborhoods, then count this enhancer. \n",
    "            if len(neighbor_support) > 0:\n",
    "    \n",
    "                # count the enhancer, so that you don't run the same enhancer twice, even when there may be multiple significant tiles. \n",
    "                already_run.append(enh_id)\n",
    "                val+=1\n",
    "    \n",
    "                # plot\n",
    "                if val < top_n:\n",
    "                        \n",
    "                    t = t[cols[:-1]].fillna(0).drop_duplicates()  # subset the dataframe to just these columns\n",
    "                    \n",
    "                    # if the outpath does not exist, make it\n",
    "                    if os.path.exists(os.path.join(re_maps, direction)) is False:\n",
    "                        os.mkdir(os.path.join(re_maps, direction))\n",
    "                \n",
    "                   \n",
    "                    # put enhancer data into long form, plot tiles\n",
    "                    plot_support_annot = \"support_name\"\n",
    "                    \n",
    "                    # make an output file name for tile plot.\n",
    "                    out_line = os.path.join(re_maps, direction,\n",
    "                                            f\"{gene}.{enh_name}.line.{plot_support_annot}.pdf\")\n",
    "                \n",
    "                    # get long form tile activity across replicates for gaba, glut.\n",
    "                    longform_gaba, var_name = getLongFormEnhInfo(t, 'gaba')\n",
    "                    longform_glut, var_name = getLongFormEnhInfo(t, 'glut')\n",
    "            \n",
    "                    # measure neighbor activity\n",
    "                    tile_activity = neighborActivity(longform_glut, longform_gaba, neighbor_support)\n",
    "                    \n",
    "                    # plot line\n",
    "                    #plot_lineplot(longform_gaba, longform_glut, var_name, out_line, enh_id,  enh_name, gene, support, plot_support_annot)\n",
    "            \n",
    "                    # plot the name of the support\n",
    "                    #plot_support_annot = \"support\"\n",
    "                    \n",
    "                    # make an output file name for tile plot.\n",
    "                    #out_line = os.path.join(re_maps, direction,\n",
    "                                          #  f\"{gene}.{enh_name}.line.{plot_support_annot}.pdf\")\n",
    "                    \n",
    "                    # plot line\n",
    "                    # plot_lineplot(longform_gaba, longform_glut, var_name, out_line, enh_id,  enh_name, gene, support, plot_support_annot)\n",
    "            else:\n",
    "                print(\"remove this tile\", tile_id)\n",
    "                remove_tiles.append(tile_id)\n",
    "           \n",
    "    return remove_tiles\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "## Top 8 GABA candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-18T18:30:25.620759Z",
     "start_time": "2023-12-18T18:30:25.527883Z"
    },
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sig_n_difcelltype_difbs\n",
      "223\n"
     ]
    }
   ],
   "source": [
    "ACTIVE_CL = \"GABA\"\n",
    "INACTIVE_CL = \"GLUT\"\n",
    "direction = f\"positive {ACTIVE_CL}>negative {INACTIVE_CL}\"\n",
    "\n",
    "# get significant tiles that fall in direction\n",
    "SET = getSigTileIdDirection(direction, enh, 2)\n",
    "\n",
    "# subset dataframe and sort\n",
    "candidate_cols = [\"name\", ACTIVE_CL.lower(), 'delta.gaba-glut', \"tile.order\"]\n",
    "\n",
    "# order candidates by highest activity, greatest difference\n",
    "GABA_candidates = enh.loc[enh[\"name\"].isin(SET), candidate_cols].drop_duplicates().sort_values(\n",
    "    by=[ACTIVE_CL.lower(), 'delta.gaba-glut'], ascending=False)[\"name\"].to_list()\n",
    "\n",
    "\n",
    "print(len(GABA_candidates))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# hand-picked GLUT candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "candidate_genes = [\"OPRM1\", \"ELMO1\", \"NBPF1\", \"SLC2A1\", \"EPHB2\", \"SYNGAP1\", \"SLC6A1\", \"JAG2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1:16612624-16614198\n",
      "NBPF1\n",
      "NECAP2\n",
      "nan\n",
      "CROCC\n",
      "broken at 59 range 67 to 70 [67, 68, 69, 70]\n",
      "broken at 73 range 59 to 59 [59]\n",
      "last one\n",
      "[[67, 68, 69, 70], [59], []]\n",
      "chr1:16612624-16614198 67,68,69,70\n",
      "chr1:16612624-16614198 59\n",
      "chr1:16612624-16614198 \n",
      "chr1:16612624-16614198 25,26,27\n",
      "chr1:16644582-16645012\n",
      "CROCC\n",
      "NBPF1\n",
      "[]\n",
      "chr1:22709886-22710729\n",
      "EPHB2\n",
      "C1QB\n",
      "nan\n",
      "last one\n",
      "[[18, 19, 20, 21, 22, 23]]\n",
      "chr1:22709886-22710729 18,19,20,21,22,23\n",
      "chr1:42923935-42925027\n",
      "SLC2A1\n",
      "ZNF691\n",
      "nan\n",
      "broken at 72 range 66 to 70 [66, 67, 68, 69, 70]\n",
      "last one\n",
      "[[66, 67, 68, 69, 70], []]\n",
      "FILLIN [[66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78]]\n",
      "chr1:42923935-42925027 66,67,68,69,70,71,72,73,74,75,76,77,78\n",
      "chr1:42957757-42958920\n",
      "nan\n",
      "SLC2A1\n",
      "ZNF691\n",
      "last one\n",
      "[[]]\n",
      "chr1:42957757-42958920 \n",
      "chr14:105162368-105162798\n",
      "GPR132\n",
      "JAG2\n",
      "nan\n",
      "last one\n",
      "[[0, 1, 2, 3, 4, 5, 6, 7, 8]]\n",
      "chr14:105162368-105162798 0,1,2,3,4,5,6,7,8\n",
      "chr3:10972256-10972924\n",
      "SLC6A1\n",
      "SLC6A11\n",
      "broken at 4 range 1 to 2 [1, 2]\n",
      "last one\n",
      "[[1, 2], [4]]\n",
      "chr3:10972256-10972924 1,2\n",
      "chr3:10972256-10972924 4\n",
      "chr3:10992875-10993705\n",
      "nan\n",
      "HRH1\n",
      "SLC6A1\n",
      "[]\n",
      "chr3:10992875-10993705 12,13,14,15,16,17\n",
      "chr3:11000975-11001605\n",
      "HRH1\n",
      "SLC6A1\n",
      "nan\n",
      "last one\n",
      "[[0, 1, 2, 3, 4, 5, 6]]\n",
      "chr3:11000975-11001605 0,1,2,3,4,5,6\n",
      "chr3:11009275-11009705\n",
      "HRH1\n",
      "SLC6A1\n",
      "[]\n",
      "chr3:11009775-11010205\n",
      "HRH1\n",
      "SLC6A1\n",
      "[]\n",
      "chr3:11009775-11010205 1,2,3,4\n",
      "chr3:11010275-11010705\n",
      "HRH1\n",
      "SLC6A1\n",
      "[]\n",
      "chr3:11010775-11011205\n",
      "HRH1\n",
      "SLC6A1\n",
      "broken at 4 range 1 to 2 [1, 2]\n",
      "last one\n",
      "[[1, 2], [4, 5, 6]]\n",
      "chr3:11010775-11011205 1,2,3,4,5,6\n",
      "chr3:11014775-11015205\n",
      "HRH1\n",
      "SLC6A1\n",
      "nan\n",
      "last one\n",
      "[[0, 1, 2]]\n",
      "chr3:11014775-11015205 0,1,2\n",
      "chr3:11015275-11015705\n",
      "HRH1\n",
      "SLC6A1\n",
      "[]\n",
      "chr3:11016975-11017405\n",
      "HRH1\n",
      "SLC6A1\n",
      "[]\n",
      "chr3:11060885-11061460\n",
      "HRH1\n",
      "SLC6A1\n",
      "nan\n",
      "[]\n",
      "chr6:33417727-33418836\n",
      "CUTA\n",
      "PHF1\n",
      "nan\n",
      "SYNGAP1\n",
      "broken at 26 range 22 to 24 [22, 23, 24]\n",
      "broken at 34 range 26 to 28 [26, 27, 28]\n",
      "broken at 36 range 34 to 34 [34]\n",
      "broken at 40 range 36 to 38 [36, 37, 38]\n",
      "broken at 59 range 40 to 46 [40, 41, 42, 43, 44, 45, 46]\n",
      "broken at 61 range 59 to 59 [59]\n",
      "broken at 63 range 61 to 61 [61]\n",
      "last one\n",
      "[[22, 23, 24], [26, 27, 28], [34], [36, 37, 38], [40, 41, 42, 43, 44, 45, 46], [59], [61], [63]]\n",
      "chr6:33417727-33418836 22,23,24\n",
      "chr6:33417727-33418836 26,27,28\n",
      "chr6:33417727-33418836 34\n",
      "chr6:33417727-33418836 36,37,38\n",
      "chr6:33417727-33418836 40,41,42,43,44,45,46\n",
      "chr6:33417727-33418836 59\n",
      "chr6:33417727-33418836 61\n",
      "chr6:33417727-33418836 63\n",
      "chr6:33425409-33426610\n",
      "nan\n",
      "ZBTB9\n",
      "SYNGAP1\n",
      "last one\n",
      "[[33, 34, 35]]\n",
      "chr6:33425409-33426610 33,34,35\n",
      "chr6:33426709-33428145\n",
      "nan\n",
      "ZBTB9\n",
      "SYNGAP1\n",
      "last one\n",
      "[[34, 35, 36]]\n",
      "chr6:33426709-33428145 34,35,36\n",
      "chr6:33454246-33454866\n",
      "ZBTB9\n",
      "SYNGAP1\n",
      "nan\n",
      "BAK1\n",
      "last one\n",
      "[[0, 1, 2, 3]]\n",
      "chr6:33454246-33454866 0,1,2,3\n",
      "chr6:154127483-154128084\n",
      "OPRM1\n",
      "IPCEF1\n",
      "nan\n",
      "last one\n",
      "[[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]]\n",
      "chr6:154127483-154128084 0,1,2,3,4,5,6,7,8,9,10,11,12,13\n",
      "chr7:37031957-37032387\n",
      "AOAH\n",
      "ELMO1\n",
      "last one\n",
      "[[0, 1, 2, 3, 4, 5, 6, 7]]\n",
      "chr7:37031957-37032387 0,1,2,3,4,5,6,7\n"
     ]
    }
   ],
   "source": [
    "#tile_id = GLUT_HANDPICK[0]\n",
    "df = enh.copy()\n",
    "\n",
    "candidate_dict={}\n",
    "\n",
    "\"\"\"evaluate if there is support from neighboring tiles in an enhancer, plot landscape of the top_n, return candidate list\"\"\"\n",
    "\n",
    "candidate_enh_gene = df.loc[df[\"gene\"].isin(candidate_genes), [\"gene\", \"enh.id\", \"enh.name\"]].drop_duplicates()\n",
    "\n",
    "for enh_id in candidate_enh_gene[\"enh.id\"].unique():\n",
    "    print(enh_id)\n",
    "    \n",
    "    #multiple tiles per enhancer\n",
    "    #if enh_id not in already_run and enh_id is not None:\n",
    "    \n",
    "    cols = ['#chr', 'start.tile', 'end.tile',\n",
    "            \"sig_n_dif_n_bs\", \"glut_1\", \"glut_2\", \"glut_3\", \n",
    "            \"gaba_1\", \"gaba_2\", \"gaba_3\",\n",
    "            \"gaba\", \"glut\", \"coor\", \"enh.id\", \"enh.name\",\n",
    "            \"tile.order\", \"direction\"]\n",
    "    \n",
    "    t = df.loc[df[\"enh.id\"] == enh_id].drop_duplicates().copy()\n",
    "    \n",
    "    # stratify by gene name, as one enhancer may be near two different genes\n",
    "    enh_name = list(set(t[\"enh.name\"]))[0]\n",
    "    gene = getMinDist(t)\n",
    "    \n",
    "    # get support (number of significant calls) per tile in enhancer\n",
    "    support = getSupport(enh_id, t)\n",
    "    \n",
    "    # scan neighbors for supporting significant differences in activity (at least 3/5 neighbor tiles must have support). \n",
    "    neighbor_support, sig_neighbor_tiles = supportingNeighbors(support)\n",
    "    sig_neighbor_tiles =list(sig_neighbor_tiles)\n",
    "    t = t[cols].fillna(0).drop_duplicates()  # subset the dataframe to just these columns\n",
    "\n",
    "    # add annotations\n",
    "    t[\"gene\"] = gene\n",
    "    t[\"candidate\"] = False\n",
    "    t[\"candidate_list\"] = None\n",
    "    if neighbor_support!=None:\n",
    "\n",
    "        # if there are gaps between significant tiles, fill those gabs in!\n",
    "        filled_in_neighbors = fillInSigGaps(sig_neighbor_tiles)\n",
    "\n",
    "        # hard coded because this is a tricky tile w/ flanks missing data. \n",
    "        if enh_name ==\"enh.43\":  # nbpf1\n",
    "            filled_in_neighbors.append([25, 26, 27]) \n",
    "        elif enh_name ==\"enh.123\":  # slc2a1\n",
    "            filled_in_neighbors = [list(np.arange(66,79))]\n",
    "            print(\"FILLIN\", filled_in_neighbors)\n",
    "        elif enh_name =='enh.1159':\n",
    "            filled_in_neighbors = [list(np.arange(1,7))]\n",
    "        elif enh_name =='enh.1157':\n",
    "            filled_in_neighbors = [list(np.arange(1,5))]\n",
    "        elif enh_name ==\"enh.1154\":\n",
    "            filled_in_neighbors = [list(np.arange(12,18))]\n",
    "        # annotate dataframe with significant regions\n",
    "        for i in filled_in_neighbors:\n",
    "            t.loc[t[\"tile.order\"].isin(i), \"candidate\"] =True\n",
    "            t.loc[t[\"tile.order\"].isin(i), \"candidate_list\"] = \",\".join(str(n) for n in i)\n",
    "            print(enh_id, \",\".join(str(n) for n in i))\n",
    "    \n",
    "        candidate_dict[enh_id] = t\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([None, '25,26,27', '59', '67,68,69,70'], dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidates = pd.concat(candidate_dict.values())\n",
    "\n",
    "candidates.loc[candidates[\"enh.name\"]==\"enh.43\", \"candidate_list\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get direction, start, end of candidate regions summarized across significant tiles, with gaps filled in. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "candidates[\"direction\"] = candidates[\"direction\"].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "directions = candidates.loc[candidates[\"candidate\"]==True].groupby(\n",
    "    [\"enh.name\",\"gene\", 'enh.id', \"candidate_list\"])[\"direction\"].sum().reset_index().sort_values(by=\"enh.name\")\n",
    "\n",
    "starts = candidates.loc[candidates[\"candidate\"]==True].groupby(\n",
    "    [\"enh.name\",\"gene\", 'enh.id', \"candidate_list\"])[\"start.tile\"].min().reset_index().sort_values(by=\"enh.name\")\n",
    "\n",
    "ends = candidates.loc[candidates[\"candidate\"]==True].groupby(\n",
    "    [\"enh.name\",\"gene\", 'enh.id', \"candidate_list\"])[\"end.tile\"].max().reset_index().sort_values(by=\"enh.name\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### add summarized information back together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(26, 8)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summed = pd.merge(starts, ends) # combine min start, max end of sig tile runs\n",
    "summed = pd.merge(directions, summed) # add direction annotations for sig tile runs\n",
    "summed = pd.merge(summed, df[[\"enh.id\", \"#chr\"]], how=\"left\").drop_duplicates()# add chr annotations\n",
    "summed.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## reorg cols and write."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "summed_cols = [ '#chr', \n",
    "                'start.tile',\n",
    "                'end.tile',\n",
    "                'enh.name',\n",
    "                'gene',\n",
    "                'enh.id',\n",
    "                'candidate_list',\n",
    "                'direction',\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "candidates = pd.merge(candidates, summed[[\"enh.name\", \"candidate_list\"]], how=\"left\")\n",
    "candidates.drop_duplicates().to_csv(CANDIDATE_REGIONS, sep='\\t', index=False)\n",
    "summed[summed_cols].drop_duplicates().to_csv(SUMMED_CANDIDATE_REGIONS, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene                                             enh.name\n",
       "AOAH_-307643.0_ELMO1_416997.0                    enh.1628                                    [0,1,2,3,4,5,6,7]\n",
       "CUTA_455.0_PHF1_6901.0_nan_SYNGAP1_-1369.0       enh.1528    [63, 61, 59, 40,41,42,43,44,45,46, 36,37,38, 3...\n",
       "EPHB2_-391.0_C1QB_56832.0_nan                    enh.74                                    [18,19,20,21,22,23]\n",
       "GPR132_-97073.0_JAG2_6161.0_nan                  enh.486                                   [0,1,2,3,4,5,6,7,8]\n",
       "HRH1_-237629.0_SLC6A1_22186.0_nan                enh.1160                                              [0,1,2]\n",
       "HRH1_-241629.0_SLC6A1_18186.0                    enh.1159                                        [1,2,3,4,5,6]\n",
       "HRH1_-242629.0_SLC6A1_17186.0                    enh.1157                                            [1,2,3,4]\n",
       "HRH1_-251229.0_SLC6A1_8386.0_nan                 enh.1155                                      [0,1,2,3,4,5,6]\n",
       "NBPF1_728.0_NECAP2_172023.0_nan_CROCC_-307887.0  enh.43                            [25,26,27, 59, 67,68,69,70]\n",
       "OPRM1_117117.0_IPCEF1_202132.0_nan               enh.1503                    [0,1,2,3,4,5,6,7,8,9,10,11,12,13]\n",
       "SLC2A1_34291.0_ZNF691_77497.0_nan                enh.123              [66,67,68,69,70,71,72,73,74,75,76,77,78]\n",
       "SLC6A1_-19935.0_SLC6A11_156191.0                 enh.1153                                             [1,2, 4]\n",
       "ZBTB9_152.0_SYNGAP1_34311.0_nan_BAK1_125511.0    enh.1531                                            [0,1,2,3]\n",
       "nan_HRH1_-259129.0_SLC6A1_386.0                  enh.1154                                  [12,13,14,15,16,17]\n",
       "nan_ZBTB9_-26765.0_SYNGAP1_7044.0                enh.1530                                           [34,35,36]\n",
       "nan_ZBTB9_-28104.0_SYNGAP1_6014.0                enh.1529                                           [33,34,35]\n",
       "Name: candidate_list, dtype: object"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summed.groupby([\"gene\", \"enh.name\"])['candidate_list'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "enh_tiles = {\"enh.1154\":[12,13,14,15,16,17], \n",
    "            \"enh.1159\":[1,2,3,4,5,6], \n",
    "             \"enh.1157\":[1,2,3,4], \n",
    "             \"enh.74\":[18,19,20,21,22,23],\n",
    "             \"enh.43\":[25,26,27],\n",
    "             \"enh.1531\":[0,1,2,3,4,5],\n",
    "             \"enh.1503\":[0,1,2,3,4,5,6,7,8,9,10,11,12,13],\n",
    "             \"enh.1628\":[0,1,2,3,4,5,6,7]\n",
    "             \n",
    "            }\n",
    "\n",
    "GABA = [\"enh.1154\", \"enh.1159\", \"enh.1157\", \"enh.74\", \"enh.43\"]\n",
    "GLUT = [\"enh.1531\", \"enh.1603\", \"enh.1628\"]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# merged sequences. \n",
    "# highlight best single tile. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['coor',\n",
       " 'gene',\n",
       " 'dist',\n",
       " 'name',\n",
       " 'sequence',\n",
       " 'bkgd',\n",
       " '72h',\n",
       " 'ctrl',\n",
       " 'cl',\n",
       " 'top_bottom',\n",
       " 'strand',\n",
       " '#chr',\n",
       " 'start.tile',\n",
       " 'end.tile',\n",
       " 'enh.id',\n",
       " 'enh.name',\n",
       " 'tile.order',\n",
       " 'gaba_log2_mean',\n",
       " 'glut_log2_mean',\n",
       " 'delta',\n",
       " 'delta_rank',\n",
       " 'gaba-label',\n",
       " 'glut-label',\n",
       " 'celltype_dif',\n",
       " 'gaba',\n",
       " 'glut',\n",
       " 'delta.gaba-glut',\n",
       " 'gaba_1',\n",
       " 'gaba_2',\n",
       " 'gaba_3',\n",
       " 'glut_1',\n",
       " 'glut_2',\n",
       " 'glut_3',\n",
       " 'pval',\n",
       " 'fdr_bool',\n",
       " 'fdr',\n",
       " '-log10p_fdr',\n",
       " 'sig',\n",
       " 'sig_n_dif',\n",
       " 'bs',\n",
       " 'direction',\n",
       " 'sig_n_dif_n_bs',\n",
       " 'list1']"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(enh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getEnhDf(enh_name, tile_list, enh_df):\n",
    "    \"\"\" get enhancer only dataframe, sort by tile, return dataframe subset\"\"\"\n",
    "    cols = [\"sequence\", \"enh.name\", \"coor\", 'enh.id',\n",
    "            \"tile.order\", 'gaba',\n",
    "         'glut', 'delta.gaba-glut',\n",
    "            'celltype_dif', \"fdr_bool\", \"bs\", 'direction']\n",
    "    \n",
    "    sub_df = enh_df.loc[(enh_df[\"enh.name\"]==enh_name)&\n",
    "                    (enh_df[\"tile.order\"].isin(tile_list)),\n",
    "                    cols].drop_duplicates().sort_values(by=\"tile.order\").reset_index(drop=True)\n",
    "\n",
    "    return sub_df\n",
    "\n",
    "\n",
    "def mergeSeq(enh_name, tile_list, enh_df, length_merged):\n",
    "    \"\"\"merge sequences together for consecutive candidate tiles\"\"\"\n",
    "    \n",
    "    last_seq, full_seq = \"\", \"\"  # record the last sequence snippet and the full sequence snippet\n",
    "    \n",
    "    for n, i in enh_df.iterrows():\n",
    "    \n",
    "        seq = i['sequence'][15:-15]  ## remove adaptor sequences ( first and last 15 bp)\n",
    "        if last_seq == \"\":\n",
    "            last_seq, full_seq = seq, seq\n",
    "            \n",
    "        elif last_seq[20:] == seq[:-20]:  # check that the internal sequences match\n",
    "            full_seq+=seq[-20:] # append the last 20 bases\n",
    "            last_seq = seq\n",
    "            \n",
    "    if len(full_seq) == length_merged:\n",
    "        print(\"complete merge\")\n",
    "\n",
    "    return full_seq\n",
    "\n",
    "\n",
    "def labelSigTile(test):\n",
    "    \"\"\"label the best candidate tile per sequence\"\"\"\n",
    "    # sum the number of significant calls per tiles\n",
    "    sig = test[test.columns[-4:-1]].sum(axis=1).copy()  \n",
    "    \n",
    "    # get max significant counts in element\n",
    "    most_sig_calls = sig.max()  \n",
    "    \n",
    "    # find tile indexes with max significant calls \n",
    "    sig_indexes = sig.loc[sig== most_sig_calls].index \n",
    "    \n",
    "    # filter on direction, get coordinate. Require that direction is negative for one element an positive for another. \n",
    "    if test[\"enh.name\"].iloc[0] in GABA:\n",
    "        direction =\"positive GABA>negative GLUT\"\n",
    "    else:\n",
    "        direction =\"positive GLUT>negative GABA\"\n",
    "        \n",
    "    best_tile_coor = list(test.loc[(test[\"direction\"]==direction)&\n",
    "                              (test.index.isin(sig_indexes)), \"coor\"])\n",
    "    # annotate best_tile\n",
    "    test[\"best_tile\"] = \"\"\n",
    "    \n",
    "    if len(best_tile_coor)>1:\n",
    "        \"\"\"find biggest difference tile, label that oen the best\"\"\"\n",
    "        biggest_dif = test.loc[test[\"coor\"].isin(best_tile_coor), 'delta.gaba-glut'].abs().max() \n",
    "        \n",
    "        # label the most sig and biggest difference\n",
    "        test.loc[(test[\"coor\"].isin(best_tile_coor))&\n",
    "        (test['delta.gaba-glut'].abs()==biggest_dif),\n",
    "        \"best_tile\"] = \"best\"\n",
    "    else:\n",
    "        \"\"\"find the most significant tile, name that one the best\"\"\"\n",
    "        test.loc[test[\"coor\"].isin(best_tile_coor), \"best_tile\"] = \"best\"\n",
    "\n",
    "    return test\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "complete merge\n",
      "enh.1154\n",
      "complete merge\n",
      "enh.1159\n",
      "complete merge\n",
      "enh.1157\n",
      "enh.74\n",
      "enh.43\n",
      "complete merge\n",
      "labeled\n",
      "enh.1531\n",
      "enh.1503\n",
      "complete merge\n",
      "enh.1628\n"
     ]
    }
   ],
   "source": [
    "new_sum , new_tiles= {}, {}\n",
    "for key, vals in enh_tiles.items():\n",
    "    val_str=\",\".join([str(i) for i in vals])\n",
    "\n",
    "    # get the summarized data\n",
    "    summed_key = summed.loc[(summed[\"enh.name\"]==key)&\n",
    "              (summed['candidate_list']==val_str)].drop_duplicates()\n",
    "    length = 270 + ((len(vals)-1)*20) # minus one corrects for the lack of 20bp stride in the first tile\n",
    "    length_merged = summed_key[\"end.tile\"]-summed_key[\"start.tile\"]  \n",
    "\n",
    "    # make sure the tiles are continuous\n",
    "    #if length == length_merged.iloc[0]:\n",
    "\n",
    "    # get enh df\n",
    "    candidate_df = getEnhDf(key, vals, enh)\n",
    "\n",
    "    # get merged sequence\n",
    "    merged_sequence = mergeSeq(enh_name, vals, candidate_df, length)\n",
    "\n",
    "    # label best candidate\n",
    "    labeled = labelSigTile(candidate_df)\n",
    "\n",
    "    if key == \"enh.1531\":\n",
    "        labeled.loc[labeled[\"tile.order\"] == 4, 'best_tile'] = \"best\"\n",
    "        print(\"labeled\")\n",
    "        \n",
    "    new_tiles[key] = labeled\n",
    "\n",
    "    # add merged sequence\n",
    "    summed_key[\"sequence\"] = merged_sequence\n",
    "    summed_key[\"sequence_len\"] = length\n",
    "    new_sum[key] = summed_key\n",
    "    print(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### make best tile, all candidate tiles, merged tile files with sequences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_sumdf = pd.concat(new_sum.values())  # summaried candidate regions\n",
    "new_tiledf = pd.concat(new_tiles.values())  # all candidate tiles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### rename columns for legibility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_sumdf.columns =['enh.name',\n",
    "                     'gene',\n",
    "                     'enh.id',\n",
    "                     'candidate.list',\n",
    "                     'direction',\n",
    "                     'start.tile.merged',\n",
    "                     'end.tile.merged',\n",
    "                     '#chr',\n",
    "                     'sequence.no.adapters',\n",
    "                     'sequence.len']\n",
    "new_tiledf.columns = ['sequence.no.adapters',\n",
    "                     'enh.name',\n",
    "                     'tile.coor',\n",
    "                     'enh.id',\n",
    "                     'tile.order',\n",
    "                     'gaba.log2meanZ',\n",
    "                     'glut.log2meanZ',\n",
    "                     'delta.gaba-glut',\n",
    "                     'celltype.dif',\n",
    "                     'fdr.bool',\n",
    "                     'bs',\n",
    "                     'direction',\n",
    "                     'best.tile']\n",
    "new_tiledf[\"sequence.len\"] = new_tiledf[\"sequence.no.adapters\"].apply(lambda x: len(x))\n",
    "\n",
    "new_tiledf = new_tiledf[['enh.name',\n",
    "                     'tile.coor',\n",
    "                     'enh.id',\n",
    "                     'tile.order',\n",
    "                     'gaba.log2meanZ',\n",
    "                     'glut.log2meanZ',\n",
    "                     'delta.gaba-glut',\n",
    "                     'celltype.dif',\n",
    "                     'fdr.bool',\n",
    "                     'bs',\n",
    "                     'direction',\n",
    "                     'best.tile',\n",
    "                         \"sequence.len\", \n",
    "                         'sequence.no.adapters']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     AGGACCGGATCAACT\n",
       "1     AGGACCGGATCAACT\n",
       "2     AGGACCGGATCAACT\n",
       "3     AGGACCGGATCAACT\n",
       "4     AGGACCGGATCAACT\n",
       "5     AGGACCGGATCAACT\n",
       "0     AGGACCGGATCAACT\n",
       "1     AGGACCGGATCAACT\n",
       "2     AGGACCGGATCAACT\n",
       "3     AGGACCGGATCAACT\n",
       "4     AGGACCGGATCAACT\n",
       "5     AGGACCGGATCAACT\n",
       "0     AGGACCGGATCAACT\n",
       "1     AGGACCGGATCAACT\n",
       "2     AGGACCGGATCAACT\n",
       "3     AGGACCGGATCAACT\n",
       "0     AGGACCGGATCAACT\n",
       "1     AGGACCGGATCAACT\n",
       "2     AGGACCGGATCAACT\n",
       "3     AGGACCGGATCAACT\n",
       "4     AGGACCGGATCAACT\n",
       "5     AGGACCGGATCAACT\n",
       "0     AGGACCGGATCAACT\n",
       "1     AGGACCGGATCAACT\n",
       "2     AGGACCGGATCAACT\n",
       "0     AGGACCGGATCAACT\n",
       "1     AGGACCGGATCAACT\n",
       "2     AGGACCGGATCAACT\n",
       "3     AGGACCGGATCAACT\n",
       "4     AGGACCGGATCAACT\n",
       "5     AGGACCGGATCAACT\n",
       "0     AGGACCGGATCAACT\n",
       "1     AGGACCGGATCAACT\n",
       "2     AGGACCGGATCAACT\n",
       "3     AGGACCGGATCAACT\n",
       "4     AGGACCGGATCAACT\n",
       "5     AGGACCGGATCAACT\n",
       "6     AGGACCGGATCAACT\n",
       "7     AGGACCGGATCAACT\n",
       "8     AGGACCGGATCAACT\n",
       "9     AGGACCGGATCAACT\n",
       "10    AGGACCGGATCAACT\n",
       "11    AGGACCGGATCAACT\n",
       "12    AGGACCGGATCAACT\n",
       "13    AGGACCGGATCAACT\n",
       "0     AGGACCGGATCAACT\n",
       "1     AGGACCGGATCAACT\n",
       "2     AGGACCGGATCAACT\n",
       "3     AGGACCGGATCAACT\n",
       "4     AGGACCGGATCAACT\n",
       "5     AGGACCGGATCAACT\n",
       "6     AGGACCGGATCAACT\n",
       "7     AGGACCGGATCAACT\n",
       "Name: sequence.no.adapters, dtype: object"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_tiledf[\"sequence.no.adapters\"].apply(lambda x:x[:15])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove adaptors\n",
    "new_tiledf[\"sequence.no.adapters\"] = new_tiledf[\"sequence.no.adapters\"].apply(lambda x:x[15:-15])\n",
    "\n",
    "best = new_tiledf.loc[new_tiledf[\"best.tile\"]!=\"\"]  # all best candidate tiles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### visual inspection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### summarized regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>enh.name</th>\n",
       "      <th>gene</th>\n",
       "      <th>enh.id</th>\n",
       "      <th>candidate.list</th>\n",
       "      <th>direction</th>\n",
       "      <th>start.tile.merged</th>\n",
       "      <th>end.tile.merged</th>\n",
       "      <th>#chr</th>\n",
       "      <th>sequence.no.adapters</th>\n",
       "      <th>sequence.len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>enh.1154</td>\n",
       "      <td>nan_HRH1_-259129.0_SLC6A1_386.0</td>\n",
       "      <td>chr3:10992875-10993705</td>\n",
       "      <td>12,13,14,15,16,17</td>\n",
       "      <td>both positive GABA&gt;GLUTboth positive GABA&gt;GLUT...</td>\n",
       "      <td>10993335</td>\n",
       "      <td>10993705</td>\n",
       "      <td>chr3</td>\n",
       "      <td>TCCCAATCCTGGGGTCAGCTGAGCCGAGTTAGAGCTATGCCCCTCA...</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>enh.1159</td>\n",
       "      <td>HRH1_-241629.0_SLC6A1_18186.0</td>\n",
       "      <td>chr3:11010775-11011205</td>\n",
       "      <td>1,2,3,4,5,6</td>\n",
       "      <td>positive GABA&gt;negative GLUTboth positive GABA&gt;...</td>\n",
       "      <td>11010795</td>\n",
       "      <td>11011165</td>\n",
       "      <td>chr3</td>\n",
       "      <td>AGCACAGTCTCATGGAGGGAGTGGGGTGGGGATAGGGGGAGGGGAA...</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>enh.1157</td>\n",
       "      <td>HRH1_-242629.0_SLC6A1_17186.0</td>\n",
       "      <td>chr3:11009775-11010205</td>\n",
       "      <td>1,2,3,4</td>\n",
       "      <td>both positive GABA&gt;GLUTboth positive GABA&gt;GLUT...</td>\n",
       "      <td>11009795</td>\n",
       "      <td>11010125</td>\n",
       "      <td>chr3</td>\n",
       "      <td>AGAGCACCTTCTCCAGGGAGCTTTCCCTGACCTCCAGGACAGCGCG...</td>\n",
       "      <td>330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2178</th>\n",
       "      <td>enh.74</td>\n",
       "      <td>EPHB2_-391.0_C1QB_56832.0_nan</td>\n",
       "      <td>chr1:22709886-22710729</td>\n",
       "      <td>18,19,20,21,22,23</td>\n",
       "      <td>positive GABA&gt;negative GLUTpositive GABA&gt;negat...</td>\n",
       "      <td>22710178</td>\n",
       "      <td>22710532</td>\n",
       "      <td>chr1</td>\n",
       "      <td>GCCCCACCCCAGGCCCTCGGGTAGCGCCAAGGCCCAGTCTCCCACT...</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1630</th>\n",
       "      <td>enh.43</td>\n",
       "      <td>NBPF1_728.0_NECAP2_172023.0_nan_CROCC_-307887.0</td>\n",
       "      <td>chr1:16612624-16614198</td>\n",
       "      <td>25,26,27</td>\n",
       "      <td>both positive GABA&gt;GLUTpositive GABA&gt;negative ...</td>\n",
       "      <td>16612981</td>\n",
       "      <td>16613271</td>\n",
       "      <td>chr1</td>\n",
       "      <td>AGGTCCGCAGAGGGAACGGATTTCTGGCCTGGAGGGTGGGGTGCGG...</td>\n",
       "      <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1580</th>\n",
       "      <td>enh.1531</td>\n",
       "      <td>ZBTB9_152.0_SYNGAP1_34311.0_nan_BAK1_125511.0</td>\n",
       "      <td>chr6:33454246-33454866</td>\n",
       "      <td>0,1,2,3,4,5</td>\n",
       "      <td>both positive GLUT&gt;GABAboth positive GLUT&gt;GABA...</td>\n",
       "      <td>33454246</td>\n",
       "      <td>33454576</td>\n",
       "      <td>chr6</td>\n",
       "      <td>ATCTCGCCCCTCTCAGGACCTACCGGGGGCGGGTGCTACACTTAAT...</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>enh.1503</td>\n",
       "      <td>OPRM1_117117.0_IPCEF1_202132.0_nan</td>\n",
       "      <td>chr6:154127483-154128084</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12,13</td>\n",
       "      <td>both positive GLUT&gt;GABApositive GLUT&gt;negative ...</td>\n",
       "      <td>154127483</td>\n",
       "      <td>154128004</td>\n",
       "      <td>chr6</td>\n",
       "      <td>TCCCTGAAATGTTTGTCATCACACAGATACACGCTCAGGATAAGAA...</td>\n",
       "      <td>530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1612</th>\n",
       "      <td>enh.1628</td>\n",
       "      <td>AOAH_-307643.0_ELMO1_416997.0</td>\n",
       "      <td>chr7:37031957-37032387</td>\n",
       "      <td>0,1,2,3,4,5,6,7</td>\n",
       "      <td>both positive GLUT&gt;GABApositive GLUT&gt;negative ...</td>\n",
       "      <td>37031957</td>\n",
       "      <td>37032367</td>\n",
       "      <td>chr7</td>\n",
       "      <td>TGGGCTAAAATAAGCATGTAAGTGTGTAGGACTGATCATCATATTG...</td>\n",
       "      <td>410</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      enh.name                                             gene  \\\n",
       "108   enh.1154                  nan_HRH1_-259129.0_SLC6A1_386.0   \n",
       "188   enh.1159                    HRH1_-241629.0_SLC6A1_18186.0   \n",
       "170   enh.1157                    HRH1_-242629.0_SLC6A1_17186.0   \n",
       "2178    enh.74                    EPHB2_-391.0_C1QB_56832.0_nan   \n",
       "1630    enh.43  NBPF1_728.0_NECAP2_172023.0_nan_CROCC_-307887.0   \n",
       "1580  enh.1531    ZBTB9_152.0_SYNGAP1_34311.0_nan_BAK1_125511.0   \n",
       "379   enh.1503               OPRM1_117117.0_IPCEF1_202132.0_nan   \n",
       "1612  enh.1628                    AOAH_-307643.0_ELMO1_416997.0   \n",
       "\n",
       "                        enh.id                   candidate.list  \\\n",
       "108     chr3:10992875-10993705                12,13,14,15,16,17   \n",
       "188     chr3:11010775-11011205                      1,2,3,4,5,6   \n",
       "170     chr3:11009775-11010205                          1,2,3,4   \n",
       "2178    chr1:22709886-22710729                18,19,20,21,22,23   \n",
       "1630    chr1:16612624-16614198                         25,26,27   \n",
       "1580    chr6:33454246-33454866                      0,1,2,3,4,5   \n",
       "379   chr6:154127483-154128084  0,1,2,3,4,5,6,7,8,9,10,11,12,13   \n",
       "1612    chr7:37031957-37032387                  0,1,2,3,4,5,6,7   \n",
       "\n",
       "                                              direction  start.tile.merged  \\\n",
       "108   both positive GABA>GLUTboth positive GABA>GLUT...           10993335   \n",
       "188   positive GABA>negative GLUTboth positive GABA>...           11010795   \n",
       "170   both positive GABA>GLUTboth positive GABA>GLUT...           11009795   \n",
       "2178  positive GABA>negative GLUTpositive GABA>negat...           22710178   \n",
       "1630  both positive GABA>GLUTpositive GABA>negative ...           16612981   \n",
       "1580  both positive GLUT>GABAboth positive GLUT>GABA...           33454246   \n",
       "379   both positive GLUT>GABApositive GLUT>negative ...          154127483   \n",
       "1612  both positive GLUT>GABApositive GLUT>negative ...           37031957   \n",
       "\n",
       "      end.tile.merged  #chr  \\\n",
       "108          10993705  chr3   \n",
       "188          11011165  chr3   \n",
       "170          11010125  chr3   \n",
       "2178         22710532  chr1   \n",
       "1630         16613271  chr1   \n",
       "1580         33454576  chr6   \n",
       "379         154128004  chr6   \n",
       "1612         37032367  chr7   \n",
       "\n",
       "                                   sequence.no.adapters  sequence.len  \n",
       "108   TCCCAATCCTGGGGTCAGCTGAGCCGAGTTAGAGCTATGCCCCTCA...           370  \n",
       "188   AGCACAGTCTCATGGAGGGAGTGGGGTGGGGATAGGGGGAGGGGAA...           370  \n",
       "170   AGAGCACCTTCTCCAGGGAGCTTTCCCTGACCTCCAGGACAGCGCG...           330  \n",
       "2178  GCCCCACCCCAGGCCCTCGGGTAGCGCCAAGGCCCAGTCTCCCACT...           370  \n",
       "1630  AGGTCCGCAGAGGGAACGGATTTCTGGCCTGGAGGGTGGGGTGCGG...           310  \n",
       "1580  ATCTCGCCCCTCTCAGGACCTACCGGGGGCGGGTGCTACACTTAAT...           370  \n",
       "379   TCCCTGAAATGTTTGTCATCACACAGATACACGCTCAGGATAAGAA...           530  \n",
       "1612  TGGGCTAAAATAAGCATGTAAGTGTGTAGGACTGATCATCATATTG...           410  "
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_sumdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### all tiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>enh.name</th>\n",
       "      <th>tile.coor</th>\n",
       "      <th>enh.id</th>\n",
       "      <th>tile.order</th>\n",
       "      <th>gaba.log2meanZ</th>\n",
       "      <th>glut.log2meanZ</th>\n",
       "      <th>delta.gaba-glut</th>\n",
       "      <th>celltype.dif</th>\n",
       "      <th>fdr.bool</th>\n",
       "      <th>bs</th>\n",
       "      <th>direction</th>\n",
       "      <th>best.tile</th>\n",
       "      <th>sequence.len</th>\n",
       "      <th>sequence.no.adapters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>enh.1531</td>\n",
       "      <td>chr6:33454246-33454516</td>\n",
       "      <td>chr6:33454246-33454866</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.184457</td>\n",
       "      <td>0.766198</td>\n",
       "      <td>-0.581740</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>both positive GLUT&gt;GABA</td>\n",
       "      <td></td>\n",
       "      <td>240</td>\n",
       "      <td>GGGTGCTACACTTAATCACCGGGAAACGCAGCTTTCCGGGGCGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>enh.1531</td>\n",
       "      <td>chr6:33454266-33454536</td>\n",
       "      <td>chr6:33454246-33454866</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.938597</td>\n",
       "      <td>1.103601</td>\n",
       "      <td>-0.165004</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>both positive GLUT&gt;GABA</td>\n",
       "      <td></td>\n",
       "      <td>240</td>\n",
       "      <td>GGGAAACGCAGCTTTCCGGGGCGGCTGACGGCCCACCCCCTCCGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>enh.1531</td>\n",
       "      <td>chr6:33454286-33454556</td>\n",
       "      <td>chr6:33454246-33454866</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-0.029056</td>\n",
       "      <td>0.604383</td>\n",
       "      <td>-0.633439</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>positive GLUT&gt;negative GABA</td>\n",
       "      <td></td>\n",
       "      <td>240</td>\n",
       "      <td>GCGGCTGACGGCCCACCCCCTCCGACCGGGCTGTTCTCTCCCGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>enh.1531</td>\n",
       "      <td>chr6:33454306-33454576</td>\n",
       "      <td>chr6:33454246-33454866</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.813237</td>\n",
       "      <td>1.496245</td>\n",
       "      <td>-0.683008</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>both positive GLUT&gt;GABA</td>\n",
       "      <td></td>\n",
       "      <td>240</td>\n",
       "      <td>TCCGACCGGGCTGTTCTCTCCCGGGCCCTGCCTCTCGGTCCTCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>enh.1531</td>\n",
       "      <td>chr6:33454326-33454596</td>\n",
       "      <td>chr6:33454246-33454866</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.713814</td>\n",
       "      <td>1.646095</td>\n",
       "      <td>-0.932280</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>both positive GLUT&gt;GABA</td>\n",
       "      <td>best</td>\n",
       "      <td>240</td>\n",
       "      <td>CCGGGCCCTGCCTCTCGGTCCTCTGCAGCGGGGAGGACTTCAGCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>enh.1531</td>\n",
       "      <td>chr6:33454346-33454616</td>\n",
       "      <td>chr6:33454246-33454866</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.071822</td>\n",
       "      <td>1.999036</td>\n",
       "      <td>-0.927214</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>both positive GLUT&gt;GABA</td>\n",
       "      <td></td>\n",
       "      <td>240</td>\n",
       "      <td>CTCTGCAGCGGGGAGGACTTCAGCAGCTCCCCGGGACGCCCGTTGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   enh.name               tile.coor                  enh.id  tile.order  \\\n",
       "0  enh.1531  chr6:33454246-33454516  chr6:33454246-33454866         0.0   \n",
       "1  enh.1531  chr6:33454266-33454536  chr6:33454246-33454866         1.0   \n",
       "2  enh.1531  chr6:33454286-33454556  chr6:33454246-33454866         2.0   \n",
       "3  enh.1531  chr6:33454306-33454576  chr6:33454246-33454866         3.0   \n",
       "4  enh.1531  chr6:33454326-33454596  chr6:33454246-33454866         4.0   \n",
       "5  enh.1531  chr6:33454346-33454616  chr6:33454246-33454866         5.0   \n",
       "\n",
       "   gaba.log2meanZ  glut.log2meanZ  delta.gaba-glut celltype.dif fdr.bool  \\\n",
       "0        0.184457        0.766198        -0.581740         True    False   \n",
       "1        0.938597        1.103601        -0.165004         True    False   \n",
       "2       -0.029056        0.604383        -0.633439        False    False   \n",
       "3        0.813237        1.496245        -0.683008         True    False   \n",
       "4        0.713814        1.646095        -0.932280         True     True   \n",
       "5        1.071822        1.999036        -0.927214         True    False   \n",
       "\n",
       "      bs                    direction best.tile  sequence.len  \\\n",
       "0  False      both positive GLUT>GABA                     240   \n",
       "1  False      both positive GLUT>GABA                     240   \n",
       "2  False  positive GLUT>negative GABA                     240   \n",
       "3  False      both positive GLUT>GABA                     240   \n",
       "4  False      both positive GLUT>GABA      best           240   \n",
       "5  False      both positive GLUT>GABA                     240   \n",
       "\n",
       "                                sequence.no.adapters  \n",
       "0  GGGTGCTACACTTAATCACCGGGAAACGCAGCTTTCCGGGGCGGCT...  \n",
       "1  GGGAAACGCAGCTTTCCGGGGCGGCTGACGGCCCACCCCCTCCGAC...  \n",
       "2  GCGGCTGACGGCCCACCCCCTCCGACCGGGCTGTTCTCTCCCGGGC...  \n",
       "3  TCCGACCGGGCTGTTCTCTCCCGGGCCCTGCCTCTCGGTCCTCTGC...  \n",
       "4  CCGGGCCCTGCCTCTCGGTCCTCTGCAGCGGGGAGGACTTCAGCAG...  \n",
       "5  CTCTGCAGCGGGGAGGACTTCAGCAGCTCCCCGGGACGCCCGTTGC...  "
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_tiledf.loc[new_tiledf[\"enh.name\"]==\"enh.1531\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### best"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>enh.name</th>\n",
       "      <th>tile.coor</th>\n",
       "      <th>enh.id</th>\n",
       "      <th>tile.order</th>\n",
       "      <th>gaba.log2meanZ</th>\n",
       "      <th>glut.log2meanZ</th>\n",
       "      <th>delta.gaba-glut</th>\n",
       "      <th>celltype.dif</th>\n",
       "      <th>fdr.bool</th>\n",
       "      <th>bs</th>\n",
       "      <th>direction</th>\n",
       "      <th>best.tile</th>\n",
       "      <th>sequence.len</th>\n",
       "      <th>sequence.no.adapters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>enh.1154</td>\n",
       "      <td>chr3:10993435-10993705</td>\n",
       "      <td>chr3:10992875-10993705</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.415619</td>\n",
       "      <td>-0.355633</td>\n",
       "      <td>0.771251</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>positive GABA&gt;negative GLUT</td>\n",
       "      <td>best</td>\n",
       "      <td>240</td>\n",
       "      <td>GCTTAAAAGCACACCGGGGCTTGGGCTGAGTGCACCGGAGCTAAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>enh.1159</td>\n",
       "      <td>chr3:11010795-11011065</td>\n",
       "      <td>chr3:11010775-11011205</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.268034</td>\n",
       "      <td>-0.056142</td>\n",
       "      <td>1.324176</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>positive GABA&gt;negative GLUT</td>\n",
       "      <td>best</td>\n",
       "      <td>240</td>\n",
       "      <td>GATAGGGGGAGGGGAAGAGCAGGCCCTACTCCCAGACCCACTACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>enh.74</td>\n",
       "      <td>chr1:22710262-22710532</td>\n",
       "      <td>chr1:22709886-22710729</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.408834</td>\n",
       "      <td>-0.204801</td>\n",
       "      <td>1.613635</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>positive GABA&gt;negative GLUT</td>\n",
       "      <td>best</td>\n",
       "      <td>240</td>\n",
       "      <td>CACACACCCTTTTTTTTTTTTTTCCATTCATTGCAGTTCGGTCCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>enh.43</td>\n",
       "      <td>chr1:16613000-16613270</td>\n",
       "      <td>chr1:16612624-16614198</td>\n",
       "      <td>26.0</td>\n",
       "      <td>1.243811</td>\n",
       "      <td>-0.307345</td>\n",
       "      <td>1.551156</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>positive GABA&gt;negative GLUT</td>\n",
       "      <td>best</td>\n",
       "      <td>240</td>\n",
       "      <td>CAGTGTCCTCTACAGGATATAGGAGGACGTGCCCCCGAAGCTGCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>enh.1531</td>\n",
       "      <td>chr6:33454326-33454596</td>\n",
       "      <td>chr6:33454246-33454866</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.713814</td>\n",
       "      <td>1.646095</td>\n",
       "      <td>-0.932280</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>both positive GLUT&gt;GABA</td>\n",
       "      <td>best</td>\n",
       "      <td>240</td>\n",
       "      <td>CCGGGCCCTGCCTCTCGGTCCTCTGCAGCGGGGAGGACTTCAGCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>enh.1503</td>\n",
       "      <td>chr6:154127563-154127833</td>\n",
       "      <td>chr6:154127483-154128084</td>\n",
       "      <td>4.0</td>\n",
       "      <td>-0.837686</td>\n",
       "      <td>2.947972</td>\n",
       "      <td>-3.785658</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>positive GLUT&gt;negative GABA</td>\n",
       "      <td>best</td>\n",
       "      <td>240</td>\n",
       "      <td>TTGTGGTGGGCCGGCTACAACCCTCCCCACCCCTCGCTTTCACTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>enh.1628</td>\n",
       "      <td>chr7:37032077-37032347</td>\n",
       "      <td>chr7:37031957-37032387</td>\n",
       "      <td>6.0</td>\n",
       "      <td>-0.215767</td>\n",
       "      <td>2.605502</td>\n",
       "      <td>-2.821269</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>positive GLUT&gt;negative GABA</td>\n",
       "      <td>best</td>\n",
       "      <td>240</td>\n",
       "      <td>AGAGTGCACCCCTACTGATTGGCTTCCTTTGTATGTTCACGGTGAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   enh.name                 tile.coor                    enh.id  tile.order  \\\n",
       "5  enh.1154    chr3:10993435-10993705    chr3:10992875-10993705        17.0   \n",
       "0  enh.1159    chr3:11010795-11011065    chr3:11010775-11011205         1.0   \n",
       "5    enh.74    chr1:22710262-22710532    chr1:22709886-22710729        23.0   \n",
       "1    enh.43    chr1:16613000-16613270    chr1:16612624-16614198        26.0   \n",
       "4  enh.1531    chr6:33454326-33454596    chr6:33454246-33454866         4.0   \n",
       "4  enh.1503  chr6:154127563-154127833  chr6:154127483-154128084         4.0   \n",
       "6  enh.1628    chr7:37032077-37032347    chr7:37031957-37032387         6.0   \n",
       "\n",
       "   gaba.log2meanZ  glut.log2meanZ  delta.gaba-glut celltype.dif fdr.bool  \\\n",
       "5        0.415619       -0.355633         0.771251        False     True   \n",
       "0        1.268034       -0.056142         1.324176         True    False   \n",
       "5        1.408834       -0.204801         1.613635         True     True   \n",
       "1        1.243811       -0.307345         1.551156         True     True   \n",
       "4        0.713814        1.646095        -0.932280         True     True   \n",
       "4       -0.837686        2.947972        -3.785658         True     True   \n",
       "6       -0.215767        2.605502        -2.821269         True     True   \n",
       "\n",
       "      bs                    direction best.tile  sequence.len  \\\n",
       "5  False  positive GABA>negative GLUT      best           240   \n",
       "0   True  positive GABA>negative GLUT      best           240   \n",
       "5   True  positive GABA>negative GLUT      best           240   \n",
       "1   True  positive GABA>negative GLUT      best           240   \n",
       "4  False      both positive GLUT>GABA      best           240   \n",
       "4   True  positive GLUT>negative GABA      best           240   \n",
       "6   True  positive GLUT>negative GABA      best           240   \n",
       "\n",
       "                                sequence.no.adapters  \n",
       "5  GCTTAAAAGCACACCGGGGCTTGGGCTGAGTGCACCGGAGCTAAGC...  \n",
       "0  GATAGGGGGAGGGGAAGAGCAGGCCCTACTCCCAGACCCACTACTC...  \n",
       "5  CACACACCCTTTTTTTTTTTTTTCCATTCATTGCAGTTCGGTCCCT...  \n",
       "1  CAGTGTCCTCTACAGGATATAGGAGGACGTGCCCCCGAAGCTGCTC...  \n",
       "4  CCGGGCCCTGCCTCTCGGTCCTCTGCAGCGGGGAGGACTTCAGCAG...  \n",
       "4  TTGTGGTGGGCCGGCTACAACCCTCCCCACCCCTCGCTTTCACTAA...  \n",
       "6  AGAGTGCACCCCTACTGATTGGCTTCCTTTGTATGTTCACGGTGAC...  "
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "out = os.path.join(RE, \"SUMMARIZED.CANDIDATE.TILE.SEQUENCES.tsv\")\n",
    "new_sumdf.to_csv(out, sep=\"\\t\", index=False)\n",
    "\n",
    "out = os.path.join(RE, \"BEST.CANDIDATE.TILE.SEQUENCES.tsv\")\n",
    "best.to_csv(out, sep=\"\\t\", index=False)\n",
    "\n",
    "out = os.path.join(RE, \"ALL.CANDIDATE.TILE.SEQUENCES.tsv\")\n",
    "new_tiledf.to_csv(out, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mambaenv",
   "language": "python",
   "name": "mambaenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "328.8px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
